```{r phage}
fileP<-"L:/promec/Animesh/HeLa/ayu/new/"
data<-read.table(paste0(fileP,"krakencountcomb.txt"),sep = "\t",header = T,row.names = 1)
row.names(data)<-paste(data[,1],row.names(data))
data<-data[,-1]
summary(data)
hist(as.matrix((data)))
typeof(data)
y=as.matrix(log2(data*100+1))
apply(y,1,function(x) !all(x==0))
data[apply(data,1,function(x) !all(x==0))==T,]
y<-y[apply(y, 1, function(x) !all(x==0)),]
y<-y[,apply(y, 2, function(x) !all(x==0))]
summary(y)
hist(as.matrix(y[-1]))
plot(y)
library(pheatmap)
y[is.infinite(y)]<-0
pheatmap(y,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)
library(quantable)
y=robustscale(log2(data*100))
y=robustscale((y))
data[,-(which(colSums(data) == 0))] 
data[-(which(colSums(data) == 0)),] 
summary(y$data)
y$data[is.na(y$data)]<-0
names(y$data)=sub("X","",names(y$data))
hist(as.matrix((y$data)))
pheatmap(y$data,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)

grepl("phag",row.names( data ) )
phageD<-data[grepl( "phage" , row.names( data ) ),  ]
phageDn0<-phageD[,-(which(colSums(phageD) == 0))] 
#phageDn0<-phageDn0[-(which(rowSums(phageDn0) == 0)),] 
write.table(phageDn0,paste0(fileP,"phageDn0.txt"),sep="\t",quote=FALSE,col.names=TRUE,row.names=TRUE)
pheatmap(phageDn0,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)
colnames(phageDn0)=sub("X.","",colnames(phageDn0))
colnames(phageDn0)=sub(".QUALITY_PASSED..0","",colnames(phageDn0))
label<-read.table("L:/promec/Animesh/HeLa/ayu/new/class.txt",header=T,row.names=1,sep="\t")
colnames(label)
summary(label)
library(ggplot2)
ggsave(file=paste0(fileP,"phageD.svg"), plot=pheatmap(phageDn0,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6,annotation_col = label), width=6, height=6)
dev.off()
svgPHC<-pheatmap(phageDn0,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6,annotation_col = label)
ggsave(file=paste0(fileP,"phageD.svg"), plot=svgPHC, width=6, height=6)
```

```{r PCA}
fileP<-"L:/promec/Animesh/HeLa/ayu/new/"
#install.packages('readxl')
library('readxl')
#?read_excel
#install.packages("dplyr")
library(ggfortify)
for(i in 1:4){
  print(i)
  numSheet<-i
  data <- read_excel(paste0(fileP,"Phylum- for PCA.xlsx"), sheet = numSheet)
  data
  #row.names(data)=data[,1]
  data<-tibble::column_to_rownames(data, var = "median")
  #install.packages("ggfortify")
  data.pca<-prcomp(t(data),center=TRUE,scale.=TRUE) 
  data.pca.plot<-biplot(data.pca,col=c(1,8), cex=c(0.5, 0.4))
  ggsave(file=paste0(fileP,"sheet",numSheet,"WithPhylum.svg"), plot=biplot(data.pca,col=c(1,8), cex=c(0.5, 0.4)), width=6, height=6)
  
  #data.pca.plot<-autoplot(data.pca,label = TRUE, label.size = 2,loadings = TRUE, loadings.label = TRUE, loadings.label.size  =3,colour = rownames(data))#,colour=rownames(t(data)))#,main="dataset")
  data.pca.plot<-autoplot(data.pca,label = TRUE, label.size = 2,alpha=0.1)#,colour=rownames(t(data)))#,main="dataset")
  data.pca.plot <- data.pca.plot +  ggtitle(paste("Sheet",numSheet)) #+ geom_point(colour = "darkblue")
  print(data.pca.plot)
  #install.packages("svglite")
  ggsave(file=paste0(fileP,"sheet",numSheet,".svg"), plot=data.pca.plot, width=6, height=6)
}
```




```{r data}
pathD<-"L:/promec/Animesh/HeLa/Metagenome data analysis/Metagenome data analysis/"
typeF<-"phylum."
files<-list.files(path=pathD, pattern=typeF, full.names=TRUE, recursive=TRUE)
comb = data.frame(ID="_holder_",typeF=typeF,stringsAsFactors=F)#row.names(total)="bacteria"
for(i in files){
  name=gsub(pathD, "", i)
  print(name)
  i1<-read.table(i,sep='\t')
  i1<-t(i1)
  colnames(i1)=c("ID",name)
  i1df<-data.frame(i1,stringsAsFactors=F)
  comb=merge(comb,i1df,by="ID",all = TRUE)
}
comb<-comb[-1,]
comb<-comb[,-2]
write.table(comb,paste(pathD,toupper(typeF),"total",".txt",sep = ""),sep = "\t",quote = FALSE,col.names = TRUE,row.names = FALSE)
```

```{r scale}
y<-read.table(paste(pathD,toupper(typeF),"total",".txt",sep = ""),sep = "\t",header = T)
#install.packages('quantable')
library(quantable)
row.names(y)<-y[,1]
y<-y[,-1]
#y<-as.matrix(unlist(y))
#y[is.na(y)]<-0
summary(y)
typeof(y)
y=robustscale(y)
y$data[is.na(y$data)]<-0
names(y$data)=sub("X","",names(y$data))
pheatmap(y$data,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)
```

```{r data1}
#ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",sep='\t')#,col.names = F,row.names = F)
ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",header=T,sep="\t",row.names = 1)
summary(ayurP)
ayurP<-ayurP[2:20]
ayurP[is.na(ayurP)]<-0
pheatmap(ayurP,scale="column",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation")
```

```{r datacom}
#ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",sep='\t')#,col.names = F,row.names = F)
ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",header=T,sep="\t",row.names = 1)
ayurP<-ayurP[2:20]
summary(ayurP)

y<-read.table(paste(pathD,toupper(typeF),"total",".txt",sep = ""),sep = "\t",header = T)
library(quantable)
row.names(y)<-y[,1]
y<-y[,-1]
summary(y)

y=merge(ayurP,y,by=0,all = TRUE)
row.names(y)<-y[,1]
y<-y[,-1]

y=robustscale(y)
y$data[is.na(y$data)]<-0
names(y$data)=sub("X","",names(y$data))
pheatmap(y$data,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6,fontsize_col=6)

```


```{r data1}
#y[is.na(y)]<-0
y=robustscale(y)
names(y$data)=sub("X\\.","",names(y$data))
names(y$data)=sub("\\..*","",names(y$data))
summary(y)
pheatmap(y,scale="column",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation")
```


```{r data1}
n1<-read.table("Norway/mgm4541725.3_genus.csv",sep='\t')
colnames(n1) <-  as.character(unlist(n1[1,]))
n1<-n1[-1,]
n1<-t(n1)
#n1<-n1[]
i1<-read.table("L:/promec/Animesh/HeLa/Metagenome data analysis/Metagenome data analysis/USA/mgm4541707.3_genus.csv",sep='\t')
#colnames(i1) <-  as.character(unlist(i1[1,]))
i1<-i1[-1,]
i1<-t(i1)
inT<-merge(t(i1)[,2],t(n1)[,2])
inT<-merge(t(i1[2,]),t(n1[2,]))
colnames(i1)=c("ID","Value")
colnames(n1)=c("ID","nValue")
inT<-merge(i1,n1,by=0, all=TRUE)
names(i1)<-"ID"
totalG = data.frame(ID="genus place holder",Value=c(0),stringsAsFactors=F)#row.names(total)="bacteria"
i1df<-data.frame(i1,stringsAsFactors=F)
n1df<-data.frame(n1,stringsAsFactors=F)
totalG=merge(totalG,i1df,by="ID",all = TRUE)
totalG=merge(totalG,n1df,by="ID",all = TRUE)
totalonto=merge(totalG,i1,by="ID",all = TRUE)

for(i in dte){
    name=names(dte)[x]
    name=as.name(paste(name,"subsytem",sep = ""))
    dft <- data.frame(i$Subsystems[,1],i$Subsystems[,2],stringsAsFactors=F)
    names(dft)=c("ID",names(dte)[x])
    x=x+1
    totalonto=merge(totalonto,dft,by="ID",all = TRUE)
    write.table(dft,paste(name,".txt",sep = ""),sep = "\t",quote = FALSE,col.names = FALSE,row.names = FALSE)
    print(name)
}


#install.packages('readxl')
library('readxl')
pathD<-"L:/promec/Animesh/giz004_supplemental_files/"
inpF1<-file.path(pathD,"File-1.xlsx")
data1 <- read_excel(inpF1)
#plot(log(data$percentage.identity),log(data$e.value),col=factor(data$number.of.mismatches))
summary(data1)
boxplot(as.numeric(data1[11,]), col="orange")

```

```{r data2}
inpF2<-file.path(pathD,"File-2.xlsx")
data2 <- read_excel(inpF2)
summary(data2)
data2rn<-data2[,-10]
data2rn<-data2rn[-1,]
data2rn<-data2rn[,-1]
data2rn<-sapply(data2rn, as.numeric)
data2rn[is.na(data2rn)]=0
summary(data2rn)
summary(log2(data2rn+1))
princomp(data2rn)
biplot(prcomp(data2rn))
```
```{r data4}
#install.packages('readxl')
library('readxl')
inpF4<-file.path(pathD,"File-4.xlsx")
data4 <- read_excel(inpF4)
summary(data4)
data4rn<-data4[,-1]
row.names(data4rn)<-data4[,1]
plot(log2(data4rn))
biplot(prcomp(data4rn))
```

```{r venn}
install.packages("BiocManager")
BiocManager::install("VennDetail")
library(VennDetail)

```


```{r label}
inpL<-"~/Documents/new/class.txt"
label<-read.table(inpL,comment="D",header=T,sep="\t")
summary(label)
```

```{r install}
install.packages('gplots')
library(gplots) 
install.packages('MCMCglmm')
library(MCMCglmm)
install.packages('quantable')
library(quantable)
heatmap.2(as.matrix(log2(data[,3:10])))
log2data=data
log2data[is.na(log2data)]<-0
log2data<-(data[,3:21])
log2data<-log2(data[,3:21]+1)
rtnorm(n = 1, mean = mean(log2(data[!is.na(data),3:21])), sd = 1, lower = -Inf, upper = Inf)
data[is.na(data)]
plot(log2data)
names(data)  
#sub("(\\.*"([Ss][0-9]+)(\\.*)","\\2",names(data)) 
names(log2data)=sub("X\\.","",names(log2data))
names(log2data)=sub("\\..*","",names(log2data))
heatmap.2(as.matrix(log2data))
heatmap.2(as.matrix((log2data)),scale="row",col=redgreen, trace="none", tracecol="grey")
y=as.matrix((log2data))
y[is.na(y)]<-0
y=robustscale(y)
y=y$data
hr <- hclust(as.dist(1-cor(t(y), method="spearman")), method="complete") 
hc <- hclust(as.dist(1-cor(y, method="spearman")), method="complete")
heatmap(y, Rowv=as.dendrogram(hr), Colv=as.dendrogram(hc)) 
heatmap.2(y, Rowv=as.dendrogram(hr), Colv=as.dendrogram(hc),scale="row",col=redgreen, trace="none", tracecol="grey")
```


```{r install}
install.packages('matR')
library(matR)
```

```{r check}
??auth
auth.MGRAST('MgRastWebKeyGoesHereKEJ88')
biom_phylum <- biomRequest(metadata("mgp80869")$mgp80869, request="organism", hit_type="all", source="RDP", group_level="phylum", evalue=15,  wait=TRUE)
phylum_matrix <- as.matrix(biom_phylum)
#biomRequest(file="Z:\\USERS\\ayurveda\\mgm4804295.3.csv",group_level="level2",evalue=1,)
```
```{r matR}
#https://github.com/MG-RAST/matR/blob/master/demo/simple.R
zz0 <- transform (zz, t_Log)
columns (zz0, "host_common_name|samp_store_temp|material")
princomp (zz0, map=c(col="host_common_name", pch="samp_store_temp"), labels="$$pubmed_id", cex=2)
distx (zz0, groups="$$host_common_name")
pp <- (rowstats (zz0, groups="$$material") $ p.value < 0.05)
pp[is.na(pp)] <- FALSE
pp

####  that information can be used to make an informative heatmap.

image (zz0 [pp,], margins=c(5,10), cexRow=0.3)

####  for comparison, here is the same heatmap, but including all annotations.

image (zz0, margins=c(5,10), cexRow=0.3)
```

```{r data}
http://api.metagenomics.anl.gov/m5nr/taxonomy?filter=Bacteroidetes&filter_level=phylum&min_level=genus

#install.packages('jsonlite')
library(jsonlite)
S5 <- fromJSON("http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

#S5-mgm4804308.3
#http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
#s13_R2- https://api-ui.mg-rast.org/metagenome/72d6f57b296d676d343638363631342e33?verbosity=stats&detail=ontology&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

#http://api.mg-rast.org/api.html#metagenome
#http://api.metagenomics.anl.gov/1/annotation/sequence/mmgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

library(rjson)
S5 <- rjson::fromJSON(file="http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

dAyu<-read.table('http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk',sep='')

http://api.metagenomics.anl.gov/metadata/export/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/project/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/inbox?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/sample/mgm4804308.3?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

https://www.mg-rast.org/mgmain.html?mgpage=project&project=760ca003346d67703137303432 
mgp17042

https://www.mg-rast.org/mgmain.html?mgpage=pipeline

Name    class   fwd     rev
s13     Vatta   TGGAACAA        TGGAACAA
S20-2   Pita    TGGCTTCA        TGGCTTCA
S23     Pita    TGGTGGTA        TGGTGGTA
S27     Kapha   TTCACGCA        TTCACGCA
S28     Vatta   AACTCACC        AACTCACC
S2      Kapha   TCCGTCTA        TCCGTCTA
S30     Kapha   AAGAGATC        AAGAGATC
S31     Kapha   AAGGACAC        AAGGACAC
S33     Kapha   AATCCGTC        AATCCGTC
S35     Vatta   AATGTTGC        AATGTTGC
S36     Vatta   ACACGACC        ACACGACC
S37     Pita    ACAGATTC        ACAGATTC
S3      Vatta   TCTTCACA        TCTTCACA
S40     Kapha   AGATGTAC        AGATGTAC
S44     Vatta   AGCACCTC        AGCACCTC
S46     Kapha   AGCCATGC        AGCCATGC
S47     Vatta   AGGCTAAC        AGGCTAAC
S48     Vatta   ATAGCGAC        ATAGCGAC
S5      Pita    TGAAGAGA        TGAAGAGA


system("mkdir ~/R/")
system("mkdir ~/R/libs/")


system("echo 'R_LIBS_USER=\"~/R/library\"' >  $HOME/.Renviron")

install.packages('devtools',lib.loc="/home/notebook/R/library")
install.packages('devtools')



```{r data}
inpF<-"Z:\\USERS\\ayurveda\\mgm4804295.3.csv"
data<-read.table(inpF,comment="D",header=T,sep="\t")
plot(log(data$percentage.identity),log(data$e.value),col=factor(data$number.of.mismatches))
```

```{r install}
install.packages('matR')
library(matR)
```

```{r check}
??auth
auth.MGRAST('MgRastWebKeyGoesHereKEJ88')
biom_phylum <- biomRequest(metadata("mgp80869")$mgp80869, request="organism", hit_type="all", source="RDP", group_level="phylum", evalue=15,  wait=TRUE)
phylum_matrix <- as.matrix(biom_phylum)
#biomRequest(file="Z:\\USERS\\ayurveda\\mgm4804295.3.csv",group_level="level2",evalue=1,)
```
```{r matR}
#https://github.com/MG-RAST/matR/blob/master/demo/simple.R
zz0 <- transform (zz, t_Log)
columns (zz0, "host_common_name|samp_store_temp|material")
princomp (zz0, map=c(col="host_common_name", pch="samp_store_temp"), labels="$$pubmed_id", cex=2)
distx (zz0, groups="$$host_common_name")
pp <- (rowstats (zz0, groups="$$material") $ p.value < 0.05)
pp[is.na(pp)] <- FALSE
pp

####  that information can be used to make an informative heatmap.

image (zz0 [pp,], margins=c(5,10), cexRow=0.3)

####  for comparison, here is the same heatmap, but including all annotations.

image (zz0, margins=c(5,10), cexRow=0.3)
```

```{r data}
http://api.metagenomics.anl.gov/m5nr/taxonomy?filter=Bacteroidetes&filter_level=phylum&min_level=genus

#install.packages('jsonlite')
library(jsonlite)
S5 <- fromJSON("http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

#S5-mgm4804308.3
#http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
#s13_R2- https://api-ui.mg-rast.org/metagenome/72d6f57b296d676d343638363631342e33?verbosity=stats&detail=ontology&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

#http://api.mg-rast.org/api.html#metagenome
#http://api.metagenomics.anl.gov/1/annotation/sequence/mmgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

library(rjson)
S5 <- rjson::fromJSON(file="http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

dAyu<-read.table('http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk',sep='')

http://api.metagenomics.anl.gov/metadata/export/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/project/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/inbox?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/sample/mgm4804308.3?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

https://www.mg-rast.org/mgmain.html?mgpage=project&project=760ca003346d67703137303432
mgp17042

https://www.mg-rast.org/mgmain.html?mgpage=pipeline

Name    class   fwd     rev
s13     Vatta   TGGAACAA        TGGAACAA
S20-2   Pita    TGGCTTCA        TGGCTTCA
S23     Pita    TGGTGGTA        TGGTGGTA
S27     Kapha   TTCACGCA        TTCACGCA
S28     Vatta   AACTCACC        AACTCACC
S2      Kapha   TCCGTCTA        TCCGTCTA
S30     Kapha   AAGAGATC        AAGAGATC
S31     Kapha   AAGGACAC        AAGGACAC
S33     Kapha   AATCCGTC        AATCCGTC
S35     Vatta   AATGTTGC        AATGTTGC
S36     Vatta   ACACGACC        ACACGACC
S37     Pita    ACAGATTC        ACAGATTC
S3      Vatta   TCTTCACA        TCTTCACA
S40     Kapha   AGATGTAC        AGATGTAC
S44     Vatta   AGCACCTC        AGCACCTC
S46     Kapha   AGCCATGC        AGCCATGC
S47     Vatta   AGGCTAAC        AGGCTAAC
S48     Vatta   ATAGCGAC        ATAGCGAC
S5      Pita    TGAAGAGA        TGAAGAGA


system("mkdir ~/R/")
system("mkdir ~/R/libs/")


system("echo 'R_LIBS_USER=\"~/R/library\"' >  $HOME/.Renviron")

install.packages('devtools',lib.loc="/home/notebook/R/library")
install.packages('devtools')

```

```


