```{r MGA}
#tar cvzf  rep.tgz genomes/*/*.tsv stats/*.tsv  genomes/*/*.nwk genomes/annotations/dram/*.tsv
#followed by https://metagenome-atlas.shinyapps.io
```

```{r libraries,inlclude=FALSE}
#install.packages("plotly")
#install.packages("heatmaply")
#install.packages("dplyr") 
#install.packages("readr")
#install.packages("stringr")
#install.packages("tidyr")
#install.packages("tibble")
#install.packages("ggplot2")
#install.packages("ggbeeswarm")
#BiocManager::install("pheatmap")
#BiocManager::install("microbiome")
#install.packages("ape")
#install.packages("vegan")
#install.packages("useful")
#install.packages("kableExtra")
#BiocManager::install("ggtree")
#install.packages("purrr")
#install.packages("broom")
#BiocManager::install("ALDEx2")
#BiocManager::install("phytools")
library(plotly)
library(heatmaply)
library(dplyr) # dpyr masks select from plotly
library(readr)
library(stringr)
library(tidyr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(pheatmap)
library(microbiome)
library(ape)
library(vegan)
library(useful)
library(kableExtra)
library(ggtree)
library(purrr)
library(broom)
library(microbiome)
library(ggbeeswarm)
library(knitr)
library(ALDEx2)
library(phytools)
```

```{r data}
atlas_wd_folder = "F:/OneDrive - NTNU/PROMEC01/ayu/"
taxonomy_file = paste0(atlas_wd_folder,"genomes/taxonomy/gtdb_taxonomy.tsv")
tree_file = paste0(atlas_wd_folder,"genomes/tree/gtdbtk.bac120.nwk")
quality_file= paste0(atlas_wd_folder,"genomes/checkm/completeness.tsv")
counts_file= paste0(atlas_wd_folder,"genomes/counts/raw_counts_genomes.tsv")
abundance_file = paste0(atlas_wd_folder,"genomes/counts/median_coverage_genomes.tsv")
readstats_file= paste0(atlas_wd_folder,"stats/read_counts.tsv")
keggmodules_file = paste0(atlas_wd_folder,"genomes/annotations/dram/kegg_modules.tsv")
metanno_file =  paste0(atlas_wd_folder,"samples.tsv")
```

```{r tax}
Tax <- read_tsv(taxonomy_file,show_col_types = FALSE)
kable(Tax)
```

```{r lab_tax}
# create a short label for each species
Tax <- Tax %>%
  mutate(Label = ifelse(is.na(species) & is.na(genus), paste0(family, " ", user_genome), species)) %>%
  mutate(Label = ifelse(is.na(Label), paste0(genus, " ", user_genome), Label))
```

```{r tree}
T <- ggtree::read.tree(tree_file)
T <- full_join(T, rename(Tax,label=user_genome), by = "label")
```

```{r highlight_tree, fig.height = 10}
tree_plot <- ggtree(T, layout = 'circular') +
  geom_tiplab(aes(label=Label))
for (phy in unique(Tax$phylum)) {
  phy_tax= Tax %>% filter( phylum == phy)
  if (dim(phy_tax)[1] >1)  {
    phy_mrca <- findMRCA(T@phylo, phy_tax$user_genome)
    tree_plot <- tree_plot + geom_cladelabel(phy_mrca,
                                              label = phy,
                                              angle=0,
                                              align=TRUE,
                                              fontsize=20,
                                             offset = 0.5,
                                              offset.text=.1,
                                              barsize=5,
                                              )
  }
}
tree_plot
  #geom_highlight(Bacteroidota, fill = "steelblue", extend = 3)
```

```{r genome_quality}
genome_quality <- read_tsv(quality_file,show_col_types = FALSE) %>%
  mutate(Quality_Score = Completeness - (5*Contamination)) %>%
  left_join(Tax, by = c("Bin Id" = "user_genome")) %>%
  mutate(Name = Label) %>%
  dplyr::select(-Label)
```

```{r plot_contam_vs_quality}
plt <- ggplot(genome_quality, aes(x = Contamination, y = Completeness, color = phylum, 
                                  text =      paste("Bin Id:", `Bin Id`, "\nName",Name,"\nQuality score:", Quality_Score))) +
  geom_point() + xlab('Contamination [%]') + ylab('Completeness [%]') +
  theme_minimal()
ggplotly(plt,tooltip = c('text' ))
```

```{r mapping_rate, fig.width=1, fig.height=2}
Counts <- read_tsv(counts_file, show_col_types = FALSE)
read_stats <-read_tsv(readstats_file, show_col_types = FALSE) %>% filter(Step == "QC" )
d <- tibble( Sample= colnames(Counts)[-1],
            Mapped_Reads = colSums(Counts[,-1]),
             x=0
             ) %>%
    left_join(read_stats, on = "Sample") %>%
  mutate(Mapping_rate = Mapped_Reads / (Reads_pe*2+Reads_se) *100)
plt <- ggplot(d, aes(y = Mapping_rate, x=x,text=paste("Sample:",Sample))) +
  ylim(c(0, 100))+
  xlim(c(-0.1,0.1)) +
   geom_beeswarm() + 
  theme_minimal()
plt
#ggplotly(plt,tooltip = c('text','y' ))
```
## Relative abundance
For the relative abundance, we take the coverage over the genome, not the raw counts. This implicitly normalizes for genome size. The coverage is calculated as the median of the coverage values calculated in 1kb blocks.
```{r abudance}
D <- read_tsv(abundance_file, show_col_types = FALSE) %>%
  column_to_rownames(var = "...1") %>%
  as.data.frame()
kable(topleft(D, c= 10))
```

```{r relative_abund}
# calculate relative abundance
rel_ab <- sweep(D, 1, rowSums(D),`/`)
# get most abundant genomes
counts_per_genome <- data.frame(sums = colSums(rel_ab)) %>%
  rownames_to_column(var = "Sample") %>%
  left_join(Tax, by = c("Sample" = "user_genome")) %>%
  arrange(desc(sums))
ggplot(counts_per_genome %>%
         top_n(sums, n = 10), aes(x = reorder(Label, -sums), y = sums)) +
  geom_col() +
  labs(x = "", y = "Abundance [rel_ab]", title = "Most abundant genomes") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90))
```

### Typical bar chart

```{r barchart_tax}
level <- 'family'
grouping_columns= c("Sample",level)
grouping_columns <- lapply(grouping_columns, as.symbol)
grouped_data <- rel_ab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "user_genome") %>%
  pivot_longer(cols = -user_genome, names_to = "Sample", values_to = "rel_ab") %>%
  left_join(Tax, by = "user_genome") %>%
  group_by_(.dots = grouping_columns) %>%
  summarise(summarized_rel_ab = sum(rel_ab))
ggplot(grouped_data, aes_string(x = "Sample", y = "summarized_rel_ab", fill = level)) +
  geom_col() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90)) +
  #scale_fill_manual(values=rep(brewer.pal(5,"Paired"), times=5))
  scale_fill_manual(values = c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7708", "#DD0D07", "#770401", "#AA0704", "#DD0A07", "#770102", "#AA0405", "#DD0708"))
```

## Kegg modules produced by Dram

```{r load_modules}
kegg_modules= read_tsv(keggmodules_file,col_select = -1)
module_names= kegg_modules %>% dplyr::select(c('module','module_name')) %>% distinct()
kable(head(kegg_modules))
```

```{r module_presence}
step_coverage_threshold= 0.8
module_step_coverage_matrix = pivot_wider(kegg_modules,  
                                   id_cols = genome,
                                   names_from = module, 
                                   values_from = step_coverage
                                   ) %>%
                              column_to_rownames("genome") %>% as.matrix() 
module_step_coverage_matrix = module_step_coverage_matrix[, colSums(module_step_coverage_matrix) > 0]
hatmap<- pheatmap(module_step_coverage_matrix, show_colnames = F)
module_presence_matrix = 1 * (module_step_coverage_matrix>step_coverage_threshold)
module_presence_matrix = module_presence_matrix[, colSums(module_presence_matrix) > 0]
hatmap
```

```{r heatmap_interactive2}
d_tax= column_to_rownames(Tax,"user_genome")
annotations <- module_step_coverage_matrix
for (genome in rownames(annotations)) {
  for (module in colnames(annotations))   {
    annotations[genome,module] = paste0("Name: ", d_tax[genome,"Label"],
                       "\nPhylum: ",d_tax[genome,"phylum"],
                       "\nPathway: ", module_names[module,"module_name"]
                       )
  }
}
heatmaply(module_step_coverage_matrix,
          custom_hovertext = annotations,
          showticklabels = c(FALSE,FALSE)
          )
```

```{r module_abundance}
# Sum of rel_ab for all species where a module is presence is equel to the matrix multiplication
stopifnot(dim(rel_ab)[2] == dim(module_presence_matrix)[1]  )
module_rel_ab <- as.matrix(rel_ab) %*% module_presence_matrix
pheatmap(module_rel_ab, show_colnames = F)
```

```{r load_metadata}
# load metadata
metadata <- read_tsv(metanno_file,show_col_types = FALSE) %>% column_to_rownames("0")
#metadata = read.table("samples.tsv")#,row.names = 1)#,sep = "\t",header = T)
#colnames(metadata)<-metadata[1,]
#metadata<-metadata[-1,]
# Subset metadata ther is one sample not quantified
#metadata = metadata[rownames(rel_ab),]
kable(metadata)
```

```{r plot_metadata, fig.width=1, fig.height=1.5}
factor(metadata$BinGroup)
ggplot(metadata, aes(x = BinGroup,y = factor(metadata$BinGroup))) +
 geom_beeswarm(cex=5) +
  ylim(10,35)+
  theme_minimal()
```

```{r labels}
# create a short label for each genome
Tax <- read_tsv(taxonomy_file,show_col_types = FALSE) %>%
  mutate(Label = ifelse(is.na(species) & is.na(genus), paste0(family, " ", user_genome), species)) %>%
  mutate(Label = ifelse(is.na(Label), paste0(genus, " ", user_genome), Label))
```

## Relative abundance

For the relative abundance, we take the coverage over the genome, not the raw counts. This implicitly normalizes for genome size. The coverage is calculated as the median of the coverage values calculated in 1kb blocks.

```{r relab}
D <- read_tsv(abundance_file, show_col_types = FALSE) %>%
  column_to_rownames(var = "...1") %>%
  as.data.frame()
# calculate relative abundance
rel_ab <- sweep(D, 1, rowSums(D),`/`)
rownames(rel_ab)
```

### Bar chart wich group labels

```{r rel_ab}
level <- 'phylum'
grouping_columns= c("BinGroup")#,level,"Diet")
grouping_columns <- lapply(grouping_columns, as.symbol)
grouped_data <- rel_ab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "BinGroup") %>%
  pivot_longer(cols = -BinGroup, names_to = "Sample", values_to = "rel_ab") %>%
  left_join(Tax, by = "BinGroup") %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")) %>%
  group_by_(.dots = grouping_columns) %>%
  summarise(summarized_rel_ab = sum(rel_ab))
ggplot(grouped_data, aes_string(x = "Sample", y = "summarized_rel_ab", fill = level)) +
  geom_col() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.title.y = element_blank()) +
  facet_wrap(~Diet, scales = "free_x") +
  scale_fill_manual(values = c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788"))
```

## Compositional data analysis
In order to analyze the microbiome at the species or genome-level we use compositional data analysis (CoDa), see more on [Wikipedia](https://en.wikipedia.org/wiki/Compositional_data) and this article:
>Gloor, Gregory B., Jean M. Macklaim, Vera Pawlowsky-Glahn, and Juan J. Egozcue. 2017. “Microbiome Datasets Are  Compositional: And This Is Not Optional.” Frontiers in Microbiology 8 (November). Frontiers: 2224.
    doi: [10.3389/fmicb.2017.02224](https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224).
For this, we transform the abundances with the centric log-ratios (CLR) after imputing values for the zeros.

```{r}
# transforme counts with centered log ratio
data <- transform(rel_ab, transform = "clr")
write.table(data,paste(atlas_wd_folder,"data.rel.ab.transform.clr.kw.txt",sep="/"),sep="\t")
hist(data[,"MAG24"])#s5,23,20,37
```

### PCA of the Aitchison distance

```{r pca}
transformed_data <- prcomp(data,scores=TRUE)
pca_data <- transformed_data$x %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "BinGroup"))
ggplot(pca_data, aes(x = PC1, y = PC2,col=factor(metadata$BinGroup))) +
#ggplot(pca_data, aes(x = PC1, y = PC2,col=rownames(metadata))) +
  geom_point() +
  theme_minimal()
#+  scale_color_manual(values = c(chow = "#00BFC4", HF = "#F8766D"))
```

### Differential abundance analysis
As the counts are normalized in centred log-ratio the log FC becomes the difference. We use the alde2 package that uses a welch test to assess differential abundance in the two groups. See Gloor et al for more information.

```{r aldex2}
Counts <- read_tsv(counts_file, show_col_types = FALSE) %>% column_to_rownames("Sample") %>% as.matrix() #%>% t()
x <- aldex.clr(reads= Counts, conds= metadata$BinGroup,mc.samples=1000, denom="all", verbose=F)#,tests="kw"
d.kw<-aldex.kw(x)
write.table(d.kw,paste(atlas_wd_folder,"aldex.clr.kw.txt",sep="/"),sep="\t")
d.eff <- aldex.glm(d.kw, CI=T, verbose=FALSE)
aldex_stats <- data.frame(d.eff,d.kw)
# rename
aldex_stats <- aldex_stats %>%
  rename(
    chow= rab.win.chow ,
    HF= rab.win.HF,
    logFC= diff.btw,
    Effect = effect ,
    Pvalue= we.ep ,
    Pvalue_BH = we.eBH
    ) %>%
  rownames_to_column(var = 'Id') %>%
  left_join(Tax, by= c("Id"= "user_genome")) %>%
  mutate(logP = -log10(Pvalue))
```

#### heatmap of significant Genomes

```{r heatmap, fig.height=3, fig.width=5}
# filter to MAG abundances that were significantly different
sig_data <- data[, aldex_stats[aldex_stats$Pvalue_BH < 0.1, ]$Id] %>%
  as.data.frame() %>%
  t()
# make a dataframe to use to annotate the heatmap
annot_df <- data.frame(Sample = colnames(sig_data)) %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")) %>%
  column_to_rownames(var = "Sample") %>%
  dplyr::select(Diet)
# sort labels by sig_data order
heatmap_labels <- Tax %>%
  filter(user_genome %in% rownames(sig_data))
heatmap_labels <- heatmap_labels[order(match(heatmap_labels$user_genome, rownames(sig_data))), ]
pheatmap(sig_data, annotation_col = annot_df,
         labels_row = heatmap_labels$Label)
```

#### Volcano plot

```{r}
## non interactive plot
# ggplot(aldex_stats, aes(x = logFC, y = logP, alpha = logP)) +
#  geom_point(color = "#67000d") +
#  theme_minimal()

plt <- ggplot(aldex_stats, aes(x = logFC, y = logP, alpha = logP,
                         label = Label, label2 = Id)) +
  geom_point(color = "#67000d") +
  theme_minimal()

ggplotly(plt) #, tooltip = c("label", "label2"))
```



```{r}
genome_of_interest = 'MAG25'
ggplot(data %>%
         as.data.frame %>%
         rownames_to_column(var = "Sample") %>%
         left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")),
         aes_string(y = genome_of_interest, x = "Diet", fill = "Diet")) +
  geom_boxplot() +
  geom_beeswarm(cex=5) +
  theme_minimal() +
  scale_fill_manual(values = c(chow = "#00BFC4", HF = "#F8766D"))

kable(Tax %>%
  filter(user_genome == genome_of_interest))
```


# Functional Differences

```{r, load_modules}

kegg_modules= read_tsv(keggmodules_file,col_select = -1,show_col_types = FALSE)
module_names= kegg_modules %>% dplyr::select(c('module','module_name')) %>% distinct()
# convert to dataframe wiht row names
module_names=  data.frame(row.names= module_names$module, name= module_names$module_name)

```

```{r, module_presence}

step_coverage_threshold= 0.8

module_step_coverage_matrix = pivot_wider(kegg_modules,  
                                   id_cols = genome,
                                   names_from = module,
                                   values_from = step_coverage
                                   ) %>%
                              column_to_rownames("genome") %>% as.matrix()

module_step_coverage_matrix = module_step_coverage_matrix[, colSums(module_step_coverage_matrix) > 0]


hatmap<- pheatmap(module_step_coverage_matrix, show_colnames = F)

module_presence_matrix = 1 * (module_step_coverage_matrix>step_coverage_threshold)

module_presence_matrix = module_presence_matrix[, colSums(module_presence_matrix) > 0]

hatmap
```

```{r, module_abundance}

# Sum of rel_ab for all species where a module is presence is equel to the matrix multiplication
stopifnot(dim(rel_ab)[2] == dim(module_presence_matrix)[1]  )
module_rel_ab <- as.matrix(rel_ab) %*% module_presence_matrix

```



```{r,heatmap_interactive2}


library(heatmaply)
d_tax= column_to_rownames(Tax,"user_genome")



annotations <- module_rel_ab
for (sample in rownames(annotations))
{
  for (module in colnames(annotations))
  {
    annotations[sample,module] = paste0("Diet: ",metadata[sample,"Diet"],
                       "\nPathway: ", module_names[module,]
                       )
  }
}




heatmaply(module_rel_ab,
          custom_hovertext = annotations,
          showticklabels = c(FALSE,TRUE),
          row_side_colors = metadata$Diet,
          colors="RdBu"
          )


```


Hover over the heatmap to identify modules that are more abundant in the HF diet.



```{r, non_interactive_module_abundance}
# Non interactive plot

# make a dataframe to use to annotate the heatmap
annot_df <- data.frame(Sample = colnames(sig_data)) %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")) %>%
  column_to_rownames(var = "Sample") %>%
  dplyr::select(Diet)

# sort labels by sig_data order
heatmap_labels <- Tax %>%
  filter(user_genome %in% rownames(sig_data))
heatmap_labels <- heatmap_labels[order(match(heatmap_labels$user_genome, rownames(sig_data))), ]


pheatmap(module_rel_ab, show_colnames = F, annotation_row =dplyr::select(metadata,"Diet"))



```


```{r phage}
fileP<-"L:/promec/Animesh/HeLa/ayu/new/"
data<-read.table(paste0(fileP,"krakencountcomb.txt"),sep = "\t",header = T,row.names = 1)
row.names(data)<-paste(data[,1],row.names(data))
data<-data[,-1]
summary(data)
hist(as.matrix((data)))
typeof(data)
y=as.matrix(log2(data*100+1))
apply(y,1,function(x) !all(x==0))
data[apply(data,1,function(x) !all(x==0))==T,]
y<-y[apply(y, 1, function(x) !all(x==0)),]
y<-y[,apply(y, 2, function(x) !all(x==0))]
summary(y)
hist(as.matrix(y[-1]))
plot(y)
library(pheatmap)
y[is.infinite(y)]<-0
pheatmap(y,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)
library(quantable)
y=robustscale(log2(data*100))
y=robustscale((y))
data[,-(which(colSums(data) == 0))] 
data[-(which(colSums(data) == 0)),] 
summary(y$data)
y$data[is.na(y$data)]<-0
names(y$data)=sub("X","",names(y$data))
hist(as.matrix((y$data)))
pheatmap(y$data,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)

grepl("phag",row.names( data ) )
phageD<-data[grepl( "phage" , row.names( data ) ),  ]
phageDn0<-phageD[,-(which(colSums(phageD) == 0))] 
#phageDn0<-phageDn0[-(which(rowSums(phageDn0) == 0)),] 
write.table(phageDn0,paste0(fileP,"phageDn0.txt"),sep="\t",quote=FALSE,col.names=TRUE,row.names=TRUE)
pheatmap(phageDn0,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)
colnames(phageDn0)=sub("X.","",colnames(phageDn0))
colnames(phageDn0)=sub(".QUALITY_PASSED..0","",colnames(phageDn0))
label<-read.table("L:/promec/Animesh/HeLa/ayu/new/class.txt",header=T,row.names=1,sep="\t")
colnames(label)
summary(label)
library(ggplot2)
ggsave(file=paste0(fileP,"phageD.svg"), plot=pheatmap(phageDn0,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6,annotation_col = label), width=6, height=6)
dev.off()
svgPHC<-pheatmap(phageDn0,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6,annotation_col = label)
ggsave(file=paste0(fileP,"phageD.svg"), plot=svgPHC, width=6, height=6)
```

```{r PCA}
fileP<-"L:/promec/Animesh/HeLa/ayu/new/"
#install.packages('readxl')
library('readxl')
#?read_excel
#install.packages("dplyr")
library(ggfortify)
for(i in 1:4){
  print(i)
  numSheet<-i
  data <- read_excel(paste0(fileP,"Phylum- for PCA.xlsx"), sheet = numSheet)
  data
  #row.names(data)=data[,1]
  data<-tibble::column_to_rownames(data, var = "median")
  #install.packages("ggfortify")
  data.pca<-prcomp(t(data),center=TRUE,scale.=TRUE) 
  data.pca.plot<-biplot(data.pca,col=c(1,8), cex=c(0.5, 0.4))
  ggsave(file=paste0(fileP,"sheet",numSheet,"WithPhylum.svg"), plot=biplot(data.pca,col=c(1,8), cex=c(0.5, 0.4)), width=6, height=6)
  
  #data.pca.plot<-autoplot(data.pca,label = TRUE, label.size = 2,loadings = TRUE, loadings.label = TRUE, loadings.label.size  =3,colour = rownames(data))#,colour=rownames(t(data)))#,main="dataset")
  data.pca.plot<-autoplot(data.pca,label = TRUE, label.size = 2,alpha=0.1)#,colour=rownames(t(data)))#,main="dataset")
  data.pca.plot <- data.pca.plot +  ggtitle(paste("Sheet",numSheet)) #+ geom_point(colour = "darkblue")
  print(data.pca.plot)
  #install.packages("svglite")
  ggsave(file=paste0(fileP,"sheet",numSheet,".svg"), plot=data.pca.plot, width=6, height=6)
}
```




```{r data}
pathD<-"L:/promec/Animesh/HeLa/Metagenome data analysis/Metagenome data analysis/"
typeF<-"phylum."
files<-list.files(path=pathD, pattern=typeF, full.names=TRUE, recursive=TRUE)
comb = data.frame(ID="_holder_",typeF=typeF,stringsAsFactors=F)#row.names(total)="bacteria"
for(i in files){
  name=gsub(pathD, "", i)
  print(name)
  i1<-read.table(i,sep='\t')
  i1<-t(i1)
  colnames(i1)=c("ID",name)
  i1df<-data.frame(i1,stringsAsFactors=F)
  comb=merge(comb,i1df,by="ID",all = TRUE)
}
comb<-comb[-1,]
comb<-comb[,-2]
write.table(comb,paste(pathD,toupper(typeF),"total",".txt",sep = ""),sep = "\t",quote = FALSE,col.names = TRUE,row.names = FALSE)
```

```{r scale}
y<-read.table(paste(pathD,toupper(typeF),"total",".txt",sep = ""),sep = "\t",header = T)
#install.packages('quantable')
library(quantable)
row.names(y)<-y[,1]
y<-y[,-1]
#y<-as.matrix(unlist(y))
#y[is.na(y)]<-0
summary(y)
typeof(y)
y=robustscale(y)
y$data[is.na(y$data)]<-0
names(y$data)=sub("X","",names(y$data))
pheatmap(y$data,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6)
```

```{r data1}
#ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",sep='\t')#,col.names = F,row.names = F)
ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",header=T,sep="\t",row.names = 1)
summary(ayurP)
ayurP<-ayurP[2:20]
ayurP[is.na(ayurP)]<-0
pheatmap(ayurP,scale="column",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation")
```

```{r datacom}
#ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",sep='\t')#,col.names = F,row.names = F)
ayurP<-read.table("L:/promec/Animesh/HeLa/ayu/new/Total Phylum.txt",header=T,sep="\t",row.names = 1)
ayurP<-ayurP[2:20]
summary(ayurP)

y<-read.table(paste(pathD,toupper(typeF),"total",".txt",sep = ""),sep = "\t",header = T)
library(quantable)
row.names(y)<-y[,1]
y<-y[,-1]
summary(y)

y=merge(ayurP,y,by=0,all = TRUE)
row.names(y)<-y[,1]
y<-y[,-1]

y=robustscale(y)
y$data[is.na(y$data)]<-0
names(y$data)=sub("X","",names(y$data))
pheatmap(y$data,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6,fontsize_col=6)

```


```{r data1}
#y[is.na(y)]<-0
y=robustscale(y)
names(y$data)=sub("X\\.","",names(y$data))
names(y$data)=sub("\\..*","",names(y$data))
summary(y)
pheatmap(y,scale="column",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation")
```


```{r data1}
n1<-read.table("Norway/mgm4541725.3_genus.csv",sep='\t')
colnames(n1) <-  as.character(unlist(n1[1,]))
n1<-n1[-1,]
n1<-t(n1)
#n1<-n1[]
i1<-read.table("L:/promec/Animesh/HeLa/Metagenome data analysis/Metagenome data analysis/USA/mgm4541707.3_genus.csv",sep='\t')
#colnames(i1) <-  as.character(unlist(i1[1,]))
i1<-i1[-1,]
i1<-t(i1)
inT<-merge(t(i1)[,2],t(n1)[,2])
inT<-merge(t(i1[2,]),t(n1[2,]))
colnames(i1)=c("ID","Value")
colnames(n1)=c("ID","nValue")
inT<-merge(i1,n1,by=0, all=TRUE)
names(i1)<-"ID"
totalG = data.frame(ID="genus place holder",Value=c(0),stringsAsFactors=F)#row.names(total)="bacteria"
i1df<-data.frame(i1,stringsAsFactors=F)
n1df<-data.frame(n1,stringsAsFactors=F)
totalG=merge(totalG,i1df,by="ID",all = TRUE)
totalG=merge(totalG,n1df,by="ID",all = TRUE)
totalonto=merge(totalG,i1,by="ID",all = TRUE)

for(i in dte){
    name=names(dte)[x]
    name=as.name(paste(name,"subsytem",sep = ""))
    dft <- data.frame(i$Subsystems[,1],i$Subsystems[,2],stringsAsFactors=F)
    names(dft)=c("ID",names(dte)[x])
    x=x+1
    totalonto=merge(totalonto,dft,by="ID",all = TRUE)
    write.table(dft,paste(name,".txt",sep = ""),sep = "\t",quote = FALSE,col.names = FALSE,row.names = FALSE)
    print(name)
}


#install.packages('readxl')
library('readxl')
pathD<-"L:/promec/Animesh/giz004_supplemental_files/"
inpF1<-file.path(pathD,"File-1.xlsx")
data1 <- read_excel(inpF1)
#plot(log(data$percentage.identity),log(data$e.value),col=factor(data$number.of.mismatches))
summary(data1)
boxplot(as.numeric(data1[11,]), col="orange")

```

```{r data2}
inpF2<-file.path(pathD,"File-2.xlsx")
data2 <- read_excel(inpF2)
summary(data2)
data2rn<-data2[,-10]
data2rn<-data2rn[-1,]
data2rn<-data2rn[,-1]
data2rn<-sapply(data2rn, as.numeric)
data2rn[is.na(data2rn)]=0
summary(data2rn)
summary(log2(data2rn+1))
princomp(data2rn)
biplot(prcomp(data2rn))
```
```{r data4}
#install.packages('readxl')
library('readxl')
inpF4<-file.path(pathD,"File-4.xlsx")
data4 <- read_excel(inpF4)
summary(data4)
data4rn<-data4[,-1]
row.names(data4rn)<-data4[,1]
plot(log2(data4rn))
biplot(prcomp(data4rn))
```

```{r venn}
install.packages("BiocManager")
BiocManager::install("VennDetail")
library(VennDetail)

```


```{r label}
inpL<-"~/Documents/new/class.txt"
label<-read.table(inpL,comment="D",header=T,sep="\t")
summary(label)
```

```{r install}
install.packages('gplots')
library(gplots) 
install.packages('MCMCglmm')
library(MCMCglmm)
install.packages('quantable')
library(quantable)
heatmap.2(as.matrix(log2(data[,3:10])))
log2data=data
log2data[is.na(log2data)]<-0
log2data<-(data[,3:21])
log2data<-log2(data[,3:21]+1)
rtnorm(n = 1, mean = mean(log2(data[!is.na(data),3:21])), sd = 1, lower = -Inf, upper = Inf)
data[is.na(data)]
plot(log2data)
names(data)  
#sub("(\\.*"([Ss][0-9]+)(\\.*)","\\2",names(data)) 
names(log2data)=sub("X\\.","",names(log2data))
names(log2data)=sub("\\..*","",names(log2data))
heatmap.2(as.matrix(log2data))
heatmap.2(as.matrix((log2data)),scale="row",col=redgreen, trace="none", tracecol="grey")
y=as.matrix((log2data))
y[is.na(y)]<-0
y=robustscale(y)
y=y$data
hr <- hclust(as.dist(1-cor(t(y), method="spearman")), method="complete") 
hc <- hclust(as.dist(1-cor(y, method="spearman")), method="complete")
heatmap(y, Rowv=as.dendrogram(hr), Colv=as.dendrogram(hc)) 
heatmap.2(y, Rowv=as.dendrogram(hr), Colv=as.dendrogram(hc),scale="row",col=redgreen, trace="none", tracecol="grey")
```


```{r install}
install.packages('matR')
library(matR)
```

```{r check}
??auth
auth.MGRAST('MgRastWebKeyGoesHereKEJ88')
biom_phylum <- biomRequest(metadata("mgp80869")$mgp80869, request="organism", hit_type="all", source="RDP", group_level="phylum", evalue=15,  wait=TRUE)
phylum_matrix <- as.matrix(biom_phylum)
#biomRequest(file="Z:\\USERS\\ayurveda\\mgm4804295.3.csv",group_level="level2",evalue=1,)
```
```{r matR}
#https://github.com/MG-RAST/matR/blob/master/demo/simple.R
zz0 <- transform (zz, t_Log)
columns (zz0, "host_common_name|samp_store_temp|material")
princomp (zz0, map=c(col="host_common_name", pch="samp_store_temp"), labels="$$pubmed_id", cex=2)
distx (zz0, groups="$$host_common_name")
pp <- (rowstats (zz0, groups="$$material") $ p.value < 0.05)
pp[is.na(pp)] <- FALSE
pp

####  that information can be used to make an informative heatmap.

image (zz0 [pp,], margins=c(5,10), cexRow=0.3)

####  for comparison, here is the same heatmap, but including all annotations.

image (zz0, margins=c(5,10), cexRow=0.3)
```

```{r data}
http://api.metagenomics.anl.gov/m5nr/taxonomy?filter=Bacteroidetes&filter_level=phylum&min_level=genus

#install.packages('jsonlite')
library(jsonlite)
S5 <- fromJSON("http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

#S5-mgm4804308.3
#http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
#s13_R2- https://api-ui.mg-rast.org/metagenome/72d6f57b296d676d343638363631342e33?verbosity=stats&detail=ontology&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

#http://api.mg-rast.org/api.html#metagenome
#http://api.metagenomics.anl.gov/1/annotation/sequence/mmgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

library(rjson)
S5 <- rjson::fromJSON(file="http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

dAyu<-read.table('http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk',sep='')

http://api.metagenomics.anl.gov/metadata/export/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/project/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/inbox?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/sample/mgm4804308.3?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

https://www.mg-rast.org/mgmain.html?mgpage=project&project=760ca003346d67703137303432 
mgp17042

https://www.mg-rast.org/mgmain.html?mgpage=pipeline

Name    class   fwd     rev
s13     Vatta   TGGAACAA        TGGAACAA
S20-2   Pita    TGGCTTCA        TGGCTTCA
S23     Pita    TGGTGGTA        TGGTGGTA
S27     Kapha   TTCACGCA        TTCACGCA
S28     Vatta   AACTCACC        AACTCACC
S2      Kapha   TCCGTCTA        TCCGTCTA
S30     Kapha   AAGAGATC        AAGAGATC
S31     Kapha   AAGGACAC        AAGGACAC
S33     Kapha   AATCCGTC        AATCCGTC
S35     Vatta   AATGTTGC        AATGTTGC
S36     Vatta   ACACGACC        ACACGACC
S37     Pita    ACAGATTC        ACAGATTC
S3      Vatta   TCTTCACA        TCTTCACA
S40     Kapha   AGATGTAC        AGATGTAC
S44     Vatta   AGCACCTC        AGCACCTC
S46     Kapha   AGCCATGC        AGCCATGC
S47     Vatta   AGGCTAAC        AGGCTAAC
S48     Vatta   ATAGCGAC        ATAGCGAC
S5      Pita    TGAAGAGA        TGAAGAGA


system("mkdir ~/R/")
system("mkdir ~/R/libs/")


system("echo 'R_LIBS_USER=\"~/R/library\"' >  $HOME/.Renviron")

install.packages('devtools',lib.loc="/home/notebook/R/library")
install.packages('devtools')



```{r data}
inpF<-"Z:\\USERS\\ayurveda\\mgm4804295.3.csv"
data<-read.table(inpF,comment="D",header=T,sep="\t")
plot(log(data$percentage.identity),log(data$e.value),col=factor(data$number.of.mismatches))
```

```{r install}
install.packages('matR')
library(matR)
```

```{r check}
??auth
auth.MGRAST('MgRastWebKeyGoesHereKEJ88')
biom_phylum <- biomRequest(metadata("mgp80869")$mgp80869, request="organism", hit_type="all", source="RDP", group_level="phylum", evalue=15,  wait=TRUE)
phylum_matrix <- as.matrix(biom_phylum)
#biomRequest(file="Z:\\USERS\\ayurveda\\mgm4804295.3.csv",group_level="level2",evalue=1,)
```
```{r matR}
#https://github.com/MG-RAST/matR/blob/master/demo/simple.R
zz0 <- transform (zz, t_Log)
columns (zz0, "host_common_name|samp_store_temp|material")
princomp (zz0, map=c(col="host_common_name", pch="samp_store_temp"), labels="$$pubmed_id", cex=2)
distx (zz0, groups="$$host_common_name")
pp <- (rowstats (zz0, groups="$$material") $ p.value < 0.05)
pp[is.na(pp)] <- FALSE
pp

####  that information can be used to make an informative heatmap.

image (zz0 [pp,], margins=c(5,10), cexRow=0.3)

####  for comparison, here is the same heatmap, but including all annotations.

image (zz0, margins=c(5,10), cexRow=0.3)
```

```{r data}
http://api.metagenomics.anl.gov/m5nr/taxonomy?filter=Bacteroidetes&filter_level=phylum&min_level=genus

#install.packages('jsonlite')
library(jsonlite)
S5 <- fromJSON("http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

#S5-mgm4804308.3
#http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
#s13_R2- https://api-ui.mg-rast.org/metagenome/72d6f57b296d676d343638363631342e33?verbosity=stats&detail=ontology&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

#http://api.mg-rast.org/api.html#metagenome
#http://api.metagenomics.anl.gov/1/annotation/sequence/mmgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

library(rjson)
S5 <- rjson::fromJSON(file="http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk")
head(S5)

dAyu<-read.table('http://api.metagenomics.anl.gov/1/annotation/sequence/mgm4804308.3?evalue=10&type=organism&source=SwissProt&auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk',sep='')

http://api.metagenomics.anl.gov/metadata/export/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/project/mgp17042?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/inbox?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk
http://api.metagenomics.anl.gov/sample/mgm4804308.3?auth=kbm2R6jxx9LSnLVNmtawtFBVtVNbk

https://www.mg-rast.org/mgmain.html?mgpage=project&project=760ca003346d67703137303432
mgp17042

https://www.mg-rast.org/mgmain.html?mgpage=pipeline

Name    class   fwd     rev
s13     Vatta   TGGAACAA        TGGAACAA
S20-2   Pita    TGGCTTCA        TGGCTTCA
S23     Pita    TGGTGGTA        TGGTGGTA
S27     Kapha   TTCACGCA        TTCACGCA
S28     Vatta   AACTCACC        AACTCACC
S2      Kapha   TCCGTCTA        TCCGTCTA
S30     Kapha   AAGAGATC        AAGAGATC
S31     Kapha   AAGGACAC        AAGGACAC
S33     Kapha   AATCCGTC        AATCCGTC
S35     Vatta   AATGTTGC        AATGTTGC
S36     Vatta   ACACGACC        ACACGACC
S37     Pita    ACAGATTC        ACAGATTC
S3      Vatta   TCTTCACA        TCTTCACA
S40     Kapha   AGATGTAC        AGATGTAC
S44     Vatta   AGCACCTC        AGCACCTC
S46     Kapha   AGCCATGC        AGCCATGC
S47     Vatta   AGGCTAAC        AGGCTAAC
S48     Vatta   ATAGCGAC        ATAGCGAC
S5      Pita    TGAAGAGA        TGAAGAGA


system("mkdir ~/R/")
system("mkdir ~/R/libs/")


system("echo 'R_LIBS_USER=\"~/R/library\"' >  $HOME/.Renviron")

install.packages('devtools',lib.loc="/home/notebook/R/library")
install.packages('devtools')

```

```


