<<<<<<< HEAD
```{r data}
inpF <-"~/promec/Animesh/mouseSILAC/dePepSS1LFQ1/proteinGroups.txt"
```

```{r windows}
#inpF <-"L:/Animesh/Abba/OutlierData.txt"
inpF <-"L:/Animesh/Jimita/txt48/proteinGroups.txt"
```


```{r summary}
data <- read.delim(inpF, row.names = 1, sep = "\t", header = T)
summary(data)
dataNorm = data[, grep("LFQ.intensity.", names(data))]
dataNorm = log2(dataNorm+1)
dataNorm[dataNorm == 0] <- NA
dataNorm[dataNorm == 0] <- runif(length(dataNorm[dataNorm == 0]))
length(dataNorm[dataNorm == 0])
#dataNorm = log2(data[, grep("LFQ.intensity.L.", names(data))])
#dataUnNorm=log2(data[,grep("Ratio.H.L.[0-9]+", names(data))])
#dataNorm=log2(data[,grep("Ratio.H.L.normalized.", names(data))])
plot(dataNorm[!is.na(dataNorm)])
boxplot(dataNorm)
hist(dataNorm[is.finite(as.matrix(dataNorm))])
summary(dataNorm)
```

```{r miss}
#source("https://bioconductor.org/biocLite.R")
#biocLite("missRows")
library("missRows")
?missRows
```

```{r power}
#biocLite("PowerExplorer")
library("PowerExplorer")
?PowerExplorerStorage
```

```{r wikiP}
biocLite("rWikiPathways")
library("rWikiPathways")
?rWikiPathways
```

```{r diff}
biocLite("SDAMS")
library("SDAMS")
?SDAMS
```

```{r sig}
biocLite("signet")
library("signet")
?signet
browseVignettes("signet")
```

```{r monokate}
biocLite("singscore")
library("singscore")
?singscore
```

```{r monokate}
biocLite("singscore")
library("singscore")
?singscore
```

```{r patientGroup}
biocLite("drawProteins")
library("drawProteins")
?drawProteins
```

```{r patientGroup}
biocLite("DEComplexDisease")
library("DEComplexDisease")
?DEComplexDisease
```

```{r SNProtein}
biocLite("DominoEffect")
library("DominoEffect")
?DominoEffect
```

```{r enrichplot}
biocLite("enrichplot")
library("enrichplot")
?enrichplot
```

```{r proteoGenome}
library("BiocInstaller")
biocLite("PGA")
?PGA
```

```{r trend}
biocLite("Trendy")
library("Trendy")
?Trendy
```


```{r distribution}
#install.packages('softImpute')
#require('softImpute')
data=data[, grep("LFQ.intensity.", names(data))]
dataNormComp = log2(data[complete.cases(log2(data)),])
summary(dataNormComp)
boxplot(dataNormComp)
biplot(princomp(dataNormComp))
#write.csv(dataNormImpCom,paste(inpF,".csv"))
```


```{r impute}
install.packages('softImpute')
require('softImpute')
#install.packages('randomForest')
?mice
#dataNorm=data[,grep("^X", names(data))]
#dataNormImp = mice(dataNorm, method = "norm.predict")
dataNormImp = mice(dataNorm, method = "pmm", seed=10, post=0.5)
#library(randomForest)
#dataNorm[dataNorm==-Inf]=NA
#dataNormImp = mice(dataNorm, method = "rf")
#dataNormImp = mice(dataNorm, method = "norm.nob")
dataNormImpCom <- complete(dataNormImp,1)
summary(dataNormImpCom)
boxplot(dataNormImpCom)
biplot(princomp(dataNormImpCom))
write.csv(dataNormImpCom,paste(inpF,".csv"))
```

```{r impute}
#install.packages('mice')
library(mice)
#install.packages('randomForest')
?mice
#dataNorm=data[,grep("^X", names(data))]
#dataNormImp = mice(dataNorm, method = "norm.predict")
dataNormImp = mice(dataNorm, method = "pmm", seed=10, post=0.5)
#library(randomForest)
#dataNorm[dataNorm==-Inf]=NA
#dataNormImp = mice(dataNorm, method = "rf")
#dataNormImp = mice(dataNorm, method = "norm.nob")
dataNormImpCom <- complete(dataNormImp,1)
summary(dataNormImpCom)
boxplot(dataNormImpCom)
biplot(princomp(dataNormImpCom))
write.csv(dataNormImpCom,paste(inpF,".csv"))
```

```{r enrichR}
install.packages("devtools")
library(devtools)
install_github("wjawaid/enrichR")
library(enrichR)
dbs <- listEnrichrDbs()
summary(dbs)
dbs <- c("GO_Cellular_Component_2017", "KEGG_2016")
enriched <- enrichr(c("NFYA", "SET", "PDS5A", "DDX23", "CWC15", "SUPT6H", "SNRNP27", "NUP153", "PANK4", "THOC5", "TRMT112"), dbs)
write.csv(enriched,"t.csv")
```


```{r remBatchEffect}

install.packages("DataExplorer")
library(DataExplorer)
create_report()
plot_missing()

source("https://bioconductor.org/biocLite.R")
biocLite("clusterSeq")
browseVignettes("clusterSeq")
```

```{r iRF}

install.packages("iRF")
library(DataExplorer)
create_report()
plot_missing()

source("https://bioconductor.org/biocLite.R")
biocLite("clusterSeq")
browseVignettes("clusterSeq")
```

```{r remBatchEffect}
#http://users.humboldt.edu/finneyb/gfortran-windows-20140629.exe # windows
#install.packages("doMC", repos="http://R-Forge.R-project.org") # windows
#sudo apt-get install libssl-dev libxml2-dev libcurl4-gnutls-dev # linux
#install.packages(c('mclust', 'mvtnorm', 'sva', 'Rcpp', 'RcppArmadillo', 'rARPACK', 'parallel', 'foreach', 'doParallel', 'ggplot2', 'RColorBrewer', 'doMC'))
#install.packages('devtools')
#install_github("syspremed/exploBATCH",verbose=TRUE)
#require(devtools)
#install_github("gnyamundanda/exploBATCHbreast") # install_github requires devtools package
library("exploBATCH")
#factors = strsplit(names(dataNorm)[grep("LFQ.intensity.L.", names(dataNorm))], "[._]")
#factors = strsplit(names(dataNormImpCom)[grep("Ratio.H.L.normalized", names(dataNorm))], "[._]")
dataNormImpCom=dataNorm
factors = strsplit(names(dataNormImpCom),"[._]")
#factorB = sapply(factors, '[[', 4)
factorB = sapply(factors, '[[', 3)
factorB<-as.numeric(t(factorB))
dtmpB<-t(as.matrix(dataNormImpCom))
#write.table(dtmpB,"$HOME/scripts/exploBATCHresults/dtmpB.txt")
#write.table(factorB,"factorB.txt")
#dtmpB<-read.table("dtmpB.txt")
#factorB<-read.table("factorB.txt")
detach("package:randomForest", unload = TRUE)
?expBATCH
#expBATCH(D = dtmpB, batchCL = factorB, Conf = NA, mindim = 2, maxdim = 6, method = "ppcca", SDselect = 1)
expBATCH(dtmpB,factorB,SDselect = 9)
dim(Breast)
dim(batchBreast)
dim((dtmpB))
rownames(dtmpB)
rownames(Breast)
colnames(dtmpB)
colnames(Breast)
```

```{r exploBATCH-test}
#http://users.humboldt.edu/finneyb/gfortran-windows-20140629.exe
#install.packages("devtools")
library(devtools)
require(exploBATCHbreast)
#nstall_github("syspremed/exploBATCH",verbose=TRUE)
library("exploBATCH")
#nstall.packages(c('mclust', 'mvtnorm', 'sva', 'Rcpp', 'RcppArmadillo', 'rARPACK', 'parallel', 'foreach', 'doParallel', 'ggplot2', 'RColorBrewer', 'doMC', 'compiler'))
#nstall.packages("doMC", repos="http://R-Forge.R-project.org")
install.packages("Rcpp")
data(Breast)
data(batchBreast)
expBATCH(D, batchCL, Conf, mindim, maxdim, method, scale, SDselect)
?expBATCH
install_github("gnyamundanda/exploBATCHbreast") # install_github requires devtools package
expBATCH(D = Breast, batchCL = batchBreast, Conf = NA, mindim = 2, maxdim = 9, method = "ppcca", SDselect = 2)
```



```{r image}
#sudo apt-get install libmagick++-dev 
#install.packages("magick")
?image_read
plot(sin(1:100))
library("magick")
image_resize(image_read("~/scripts/exploBATCHresults/assessComBat/BatchEffect.png"),250)
```

```{r OCR}
#sudo apt-get install libtesseract-dev libleptonica-dev tesseract-ocr
#install.packages("tesseract")
#library("tesseract")
cat(image_ocr(image_read("~/scripts/exploBATCHresults/ppccaBeforeCorrection/ppccaBeforeCorrection.png")))
```

```{r bioC}
source("https://bioconductor.org/biocLite.R")
biocLite("sva")
library("sva")
```

```{r comp}
install.packages('mice')
library(mice)
install.packages('randomForest')
library(randomForest)
dataNorm[dataNorm==-Inf]=NA
dataNormImp=mice(dataNorm, method="rf")
dataNormImpCom <- complete(dataNormImp,1)
summary(dataNormImpCom)
#barplot(dataNormImpCom)
plot(dataNormImp$imp$LFQ.intensity.607_1_)
plot(princomp(dataNormImpCom))
biplot(prcomp(dataNormImpCom,scale = F))
biplot(princomp(dataNormImpCom))
```


```{r condition}
factors=strsplit(names(data)[grep("Ratio.H.L.normalized.", names(data))],"[._]")
factorS=sapply(factors, '[[', 5) 
factorC=sapply(factors, '[[', 6) 
```

```{r SVA}
#https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btu375
#source("http://www.bioconductor.org/biocLite.R")
#biocLite(c("Biobase","sva","bladderbatch","snpStats"))
  library(sva)
  library(devtools)
  library(Biobase)
  library(bladderbatch)
  library(snpStats)

data(bladderdata)
pheno = pData(bladderEset)
edata = exprs(bladderEset)

mod = model.matrix(~as.factor(cancer) + as.factor(batch),data=pheno)
fit = lm.fit(mod,t(edata))
hist(fit$coefficients[2,],col=2,breaks=100)
table(pheno$cancer,pheno$batch)

batch = pheno$batch
modcombat = model.matrix(~1, data=pheno)
modcancer = model.matrix(~cancer, data=pheno)
combat_edata = ComBat(dat=edata, batch=batch, mod=modcombat, par.prior=TRUE, prior.plots=FALSE)

combat_fit = lm.fit(modcancer,t(combat_edata))
hist(combat_fit$coefficients[2,],col=2,breaks=100)

plot(fit$coefficients[2,],combat_fit$coefficients[2,],col=2,
      xlab="Linear Model",ylab="Combat",xlim=c(-5,5),ylim=c(-5,5))
abline(c(0,1),col=1,lwd=3)

mod = model.matrix(~cancer,data=pheno)
mod0 = model.matrix(~1, data=pheno)
sva1 = sva(edata,mod,mod0,n.sv=2)

summary(lm(sva1$sv ~ pheno$batch))

boxplot(sva1$sv[,2] ~ pheno$batch)
points(sva1$sv[,2] ~ jitter(as.numeric(pheno$batch)),col=as.numeric(pheno$batch))

modsv = cbind(mod,sva1$sv)
fitsv = lm.fit(modsv,t(edata))

par(mfrow=c(1,2))
plot(fitsv$coefficients[2,],combat_fit$coefficients[2,],col=2,
      xlab="SVA",ylab="Combat",xlim=c(-5,5),ylim=c(-5,5))
abline(c(0,1),col=1,lwd=3)
plot(fitsv$coefficients[2,], fit$coefficients[2,],col=2,
      xlab="SVA",ylab="linear model",xlim=c(-5,5),ylim=c(-5,5))
abline(c(0,1),col=1,lwd=3)

data(for.exercise)
controls <- rownames(subject.support)[subject.support$cc==0]
use <- seq(1, ncol(snps.10), 10)
ctl.10 <- snps.10[controls,use]

xxmat <- xxt(ctl.10, correct.for.missing=FALSE)
evv <- eigen(xxmat, symmetric=TRUE)
pcs <- evv$vectors[,1:5]

pop <- subject.support[controls,"stratum"]
plot(pcs[,1],pcs[,2],col=as.numeric(pop),
      xlab="PC1",ylab="PC2")
legend(0,0.15,legend=levels(pop),pch=19,col=1:2)
```




```{r ANOVA}
#https://stats.stackexchange.com/questions/89021/how-to-get-only-desirable-comparisons-from-post-hoc
#plot(log2(data[,grep("Ratio.H.L.normalized.", names(data))]))
#dataNorm=log2(data[,grep("Ratio.H.L.normalized.", names(data))])
anova(lm(as.numeric(dataNorm[2,])~factorC*factorS))
summary(dataNorm)
#summary(anova(lm(as.numeric(dataNorm[2,])~factorC*factorS)))
aov((as.numeric(dataNorm[2,])~factorC*factorS))
TukeyHSD(aov((as.numeric(dataNorm[2,])~factorC*factorS)))
tcold=tc
tc=apply(dataNorm,1,function(x){
#tc=apply(dataNorm, 1, function(x) 
  tryCatch(TukeyHSD(aov(x~factorC*factorS),"factorC:factorS", ordered = TRUE),error=function(x){return(rep(1,15))})})
wval=t(sapply(names(tc),function(x){tryCatch(tc[[x]]$`factorC:factorS`[46:60],error=function(x){return(rep(1,15))})}))
write.table(wval,outF,sep="\t")
tc$`F7A0B0;P04370-9;F6ZIA4;P04370-7`
dataNorm[grep("F7A",row.names(data)),]
#try(TukeyHSD(aov((x~factorC*factorS)))))
tcold$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`
plot(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`)
```

```{r write-output}
tc
write.table(t(tc),sep = "\t")
outF = paste0(inpF,"anovaTukey.txt")
class(tc)
names(tc)
do.call(rbind, lapply(names(tc), function(x) data.frame(c(ID=x, tc$x$`factorC:factorS`))))
lapply(names(tc), function(x) write.table(t(t(tcold[[x]]$`factorC:factorS`)[4,]),outF,sep = "\t"))
write.table(t(sapply(tc, 
                     function(x){tryCatch(x$`factorC:factorS`)})),sep="\t")
#function(x){tryCatch(x$`factorC:factorS`,error=function(x){return(NULL)})})),outF,sep="\t")
write.table(t(t(tc$x$`factorC:factorS`)[4,]),sep = "\t")
write.table(t(t(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`$`factorC:factorS`)[4,]),sep = "\t")
dump(tc, file=outF)
```

```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```


```{r interpolate}
data<-read.table("c:/Users/animeshs/Downloads/Sunshine hours.txt",sep='\t',header=T)
summary(data$SONNENSCHEINDAUER)
dp=approx(data$SONNENSCHEINDAUER)
summary(dp)
plot(dp$SONNENSCHEINDAUER,data$SONNENSCHEINDAUER)
install.packages('zoo')
library('zoo')
dp<-rollmean(data,1)
dp<-na.approx(data)
summary(dp)
write.table(dp,"c:/Users/animeshs/Downloads/SunshineDP.txt",sep='\t')
```

```{r iPondData}
iPond=read.delim("f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVA.txt",row.names = 48,header=T)
#[,190:213]
label=t(read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Kristian/Groups.txt",row.names = 1,header=T))
```

```{r corr3ction}
hist(iPond$M..Two.way.ANOVA.p.value.Interaction)
hist(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"))
hist((p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH")))
hist(log2(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH")))
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhInt.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhTr.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhCond.txt", sep="\t")

```


```{r data}
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Articles in prep/AID/ID/AIDberitPD21SequestScoreTtestRcomp.txt",sep="\t",row.names=1, header = T) #AIDome
#data <- read.delim("F:/promec/Results/Ani/Mirta/DataActinNormalizedHeatMapCtrlwPCNA.txt",row.names=19,sep="\t",header = T) #BER
#data <- read.delim("F:/promec/Davi/Mei/2016-05_PancreaticCancer/QE/lfq2pep/LFQvalues.txt",row.names=90,sep="\t",header = T)
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15P.txt",fill=T, row.names=86, sep = "\t",header = T)
#df <- read.delim("https://gist.githubusercontent.com/anonymous/0bc36ec5f46757de7c2c/raw/517ef70ab6a68e600f57308e045c2b4669a7abfc/example.txt", header=TRUE, row.names=1, sep="\t")
#df<-df[df$Potential.contaminant!='+',] 
data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15Pr.txt",sep = "\t",header = T)


summary(data)
t.test(data[1,3:5],data[1,6:8],data[1,9:11])
t.test(data[1,3:5],data[1,6:8])

```

```{r lmfit}
source("https://bioconductor.org/biocLite.R")
biocLite("golubEsets")
library(golubEsets)
data()
data(golubTrain)
golubTrain
golub<-data.frame(golubTrain@exprs)
labelgb <- factor(c(rep("ALL",27),rep("AML",11)))
names(golub) <- labelgb
golub2<-log2(golub)
plot(golub2)
biocLite('limma')
library('limma')
fit3 = lmFit(golub2)
e3 = eBayes(fit3)
tab3<-topTable(e3, number=nrow(golub))
head(tab3)
fit3@.Data[[9]]
fit4 = lmFit(golub2, design)
```


```{r shuffle}
ecoli <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples.txt", header=TRUE, na.strings='', sep="\t")
for  (cnt in 1:6){
  cnt
}
  ecoli_shuffle <- ecoli_shuffle[sample(nrow(ecoli)),]
write.csv(file="Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples_shuffled_rn.csv",x=ecoli_shuffle,row.names = F)
```
```{r rawread}
library("MSnbase")
```


```{r prot}
install.packages("PTXQC")
library("PTXQC")
help(package = "PTXQC")
vignette("PTXQC-InputData", package = "PTXQC")
txt_folder = "F:/promec/Elite/LARS/2016/desember/Lymphoma/Results"
r = createReport(txt_folder)
cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```

```{r prot}
require(PTXQC)

## the next require() is needed to prevent a spurious error in certain R versions (mighyou be a bug in R or a package)
## error message is:
##    Error in Scales$new : could not find function "loadMethod"
require(methods)

## specify a path to a MaxQuant txt folder
## Note: This folder needs to be complete (see 'vignette("PTXQC-InputData", package = "PTXQC")')
if (1) {
  ## we will use an example dataset from PRIDE (dataset 2 of the PTXQC publication)
  local_zip = tempfile(fileext=".zip")
  download.file("ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2015/11/PXD003133/txt_20min.zip", destfile = local_zip)
  unzip(local_zip, exdir = tempdir()) ## extracts content
  txt_folder = paste0(tempdir(),"/txt_20min")
} else {
  ## if you have local MaxQuant output, just use it
  txt_folder = "c:/Proteomics/MouseLiver/combined/txt"
}

r = createReport(txt_folder)

cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```


```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```

=======
```{r data}
inpF <-"/home/animeshs/Downloads/proteinGroups.txt"
data <- read.delim(inpF,row.names=1,sep="\t",header = T)
summary(data)
#plot(data)
```

```{r pkgList}
mice
combat
wgcna

biocLite("sva")
library("sva")

```

```{r exploBATCHinstall}
#http://users.humboldt.edu/finneyb/gfortran-windows-20140629.exe
install.packages('httr')
install.packages('devtools')
require(devtools)
install_github("syspremed/exploBATCH",verbose=TRUE)
```

```{r exploBATCHuseCase}
library("exploBATCH")
?expBATCH
require(exploBATCHbreast)
data(Breast)
data(batchBreast)
expBATCH(D=Breast,batchCL=batchBreast,Conf=NA,mindim=2,maxdim=9,method="ppcca",SDselect=2)
```


```{r condition}
factors=strsplit(names(data)[grep("Ratio.H.L.normalized.", names(data))],"[._]")
factorS=sapply(factors, '[[', 5) 
factorC=sapply(factors, '[[', 6) 
```

```{r ANOVA}
#https://stats.stackexchange.com/questions/89021/how-to-get-only-desirable-comparisons-from-post-hoc
#plot(log2(data[,grep("Ratio.H.L.normalized.", names(data))]))
dataNorm=log2(data[,grep("Ratio.H.L.normalized.", names(data))])
anova(lm(as.numeric(dataNorm[2,])~factorC*factorS))
summary(dataNorm)
#summary(anova(lm(as.numeric(dataNorm[2,])~factorC*factorS)))
aov((as.numeric(dataNorm[2,])~factorC*factorS))
TukeyHSD(aov((as.numeric(dataNorm[2,])~factorC*factorS)))
tcold=tc
tc=apply(dataNorm,1,function(x){
#tc=apply(dataNorm, 1, function(x) 
  tryCatch(TukeyHSD(aov(x~factorC*factorS),"factorC:factorS", ordered = TRUE),error=function(x){return(rep(1,15))})})
wval=t(sapply(names(tc),function(x){tryCatch(tc[[x]]$`factorC:factorS`[46:60],error=function(x){return(rep(1,15))})}))
write.table(wval,outF,sep="\t")
tc$`F7A0B0;P04370-9;F6ZIA4;P04370-7`
dataNorm[grep("F7A",row.names(data)),]
#try(TukeyHSD(aov((x~factorC*factorS)))))
tcold$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`
plot(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`)
```

```{r write-output}
tc
write.table(t(tc),sep = "\t")
outF = paste0(inpF,"anovaTukey.txt")
class(tc)
names(tc)
do.call(rbind, lapply(names(tc), function(x) data.frame(c(ID=x, tc$x$`factorC:factorS`))))
lapply(names(tc), function(x) write.table(t(t(tcold[[x]]$`factorC:factorS`)[4,]),outF,sep = "\t"))
write.table(t(sapply(tc, 
                     function(x){tryCatch(x$`factorC:factorS`)})),sep="\t")
#function(x){tryCatch(x$`factorC:factorS`,error=function(x){return(NULL)})})),outF,sep="\t")
write.table(t(t(tc$x$`factorC:factorS`)[4,]),sep = "\t")
write.table(t(t(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`$`factorC:factorS`)[4,]),sep = "\t")
dump(tc, file=outF)
```

```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
---
title: "batch-effect-analysis"
author: "Ani"
date: "Monday, Feb 15, 2016"
output: html_document
---
<http://rmarkdown.rstudio.com>.

=======
```{r interpolate}
data<-read.table("c:/Users/animeshs/Downloads/Sunshine hours.txt",sep='\t',header=T)
summary(data$SONNENSCHEINDAUER)
dp=approx(data$SONNENSCHEINDAUER)
summary(dp)
plot(dp$SONNENSCHEINDAUER,data$SONNENSCHEINDAUER)
install.packages('zoo')
library('zoo')
dp<-rollmean(data,1)
dp<-na.approx(data)
summary(dp)
write.table(dp,"c:/Users/animeshs/Downloads/SunshineDP.txt",sep='\t')
```

```{r iPondData}
iPond=read.delim("f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVA.txt",row.names = 48,header=T)
#[,190:213]
label=t(read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Kristian/Groups.txt",row.names = 1,header=T))
```

```{r corr3ction}
hist(iPond$M..Two.way.ANOVA.p.value.Interaction)
hist(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"))
hist((p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH")))
hist(log2(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH")))
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhInt.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhTr.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhCond.txt", sep="\t")

```


```{r data}
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Articles in prep/AID/ID/AIDberitPD21SequestScoreTtestRcomp.txt",sep="\t",row.names=1, header = T) #AIDome
#data <- read.delim("F:/promec/Results/Ani/Mirta/DataActinNormalizedHeatMapCtrlwPCNA.txt",row.names=19,sep="\t",header = T) #BER
#data <- read.delim("F:/promec/Davi/Mei/2016-05_PancreaticCancer/QE/lfq2pep/LFQvalues.txt",row.names=90,sep="\t",header = T)
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15P.txt",fill=T, row.names=86, sep = "\t",header = T)
#df <- read.delim("https://gist.githubusercontent.com/anonymous/0bc36ec5f46757de7c2c/raw/517ef70ab6a68e600f57308e045c2b4669a7abfc/example.txt", header=TRUE, row.names=1, sep="\t")
#df<-df[df$Potential.contaminant!='+',] 
data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15Pr.txt",sep = "\t",header = T)


summary(data)
t.test(data[1,3:5],data[1,6:8],data[1,9:11])
t.test(data[1,3:5],data[1,6:8])

```

```{r lmfit}
source("https://bioconductor.org/biocLite.R")
biocLite("golubEsets")
library(golubEsets)
data()
data(golubTrain)
golubTrain
golub<-data.frame(golubTrain@exprs)
labelgb <- factor(c(rep("ALL",27),rep("AML",11)))
names(golub) <- labelgb
golub2<-log2(golub)
plot(golub2)
biocLite('limma')
library('limma')
fit3 = lmFit(golub2)
e3 = eBayes(fit3)
tab3<-topTable(e3, number=nrow(golub))
head(tab3)
fit3@.Data[[9]]
fit4 = lmFit(golub2, design)
```


```{r shuffle}
ecoli <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples.txt", header=TRUE, na.strings='', sep="\t")
for  (cnt in 1:6){
  cnt
}
  ecoli_shuffle <- ecoli_shuffle[sample(nrow(ecoli)),]
write.csv(file="Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples_shuffled_rn.csv",x=ecoli_shuffle,row.names = F)
```
```{r rawread}
library("MSnbase")
```


```{r prot}
install.packages("PTXQC")
library("PTXQC")
help(package = "PTXQC")
vignette("PTXQC-InputData", package = "PTXQC")
txt_folder = "F:/promec/Elite/LARS/2016/desember/Lymphoma/Results"
r = createReport(txt_folder)
cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```

```{r prot}
require(PTXQC)

## the next require() is needed to prevent a spurious error in certain R versions (mighyou be a bug in R or a package)
## error message is:
##    Error in Scales$new : could not find function "loadMethod"
require(methods)

## specify a path to a MaxQuant txt folder
## Note: This folder needs to be complete (see 'vignette("PTXQC-InputData", package = "PTXQC")')
if (1) {
  ## we will use an example dataset from PRIDE (dataset 2 of the PTXQC publication)
  local_zip = tempfile(fileext=".zip")
  download.file("ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2015/11/PXD003133/txt_20min.zip", destfile = local_zip)
  unzip(local_zip, exdir = tempdir()) ## extracts content
  txt_folder = paste0(tempdir(),"/txt_20min")
} else {
  ## if you have local MaxQuant output, just use it
  txt_folder = "c:/Proteomics/MouseLiver/combined/txt"
}

r = createReport(txt_folder)

cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```


```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
---
title: "batch-effect-analysis"
author: "Ani"
date: "Monday, Feb 15, 2016"
output: html_document
---
<http://rmarkdown.rstudio.com>.

>>>>>>> ebfc112f27001e13d78ab738e1c0e1a49f30d70e
=======
```{r data}
inpF <-"F:/promec/Animesh/Jimita/txt48/proteinGroups.txt"
data <- read.delim(inpF,row.names=1,sep="\t",header = T)
summary(data)
dataNorm=log2(data[,grep("LFQ*", names(data))])
dataUnNorm=log2(data[,grep("Ratio.H.L.[0-9]+", names(data))])
plot(dataNorm[is.na(dataN)])
hist(dataNorm)
boxplot(dataNorm)
boxplot(dataUnNorm)
dataUnNormZS=dataUnNorm-mean(dataUnNorm)
class(dataUnNorm)
biplot(princomp((dataNorm)))
```

```{r comp}
install.packages('mice')
library(mice)
install.packages('randomForest')
library(randomForest)
dataNorm[dataNorm==-Inf]=NA
dataNormImp=mice(dataNorm, method="rf")
dataNormImpCom <- complete(dataNormImp,1)
summary(dataNormImpCom)
#barplot(dataNormImpCom)
plot(dataNormImp$imp$LFQ.intensity.607_1_)
plot(princomp(dataNormImpCom))
biplot(prcomp(dataNormImpCom,scale = F))
biplot(princomp(dataNormImpCom))
```


```{r condition}
factors=strsplit(names(data)[grep("Ratio.H.L.normalized.", names(data))],"[._]")
factorS=sapply(factors, '[[', 5) 
factorC=sapply(factors, '[[', 6) 
```

```{r SVA}
#https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btu375
#source("http://www.bioconductor.org/biocLite.R")
#biocLite(c("Biobase","sva","bladderbatch","snpStats"))
  library(sva)
  library(devtools)
  library(Biobase)
  library(bladderbatch)
  library(snpStats)

data(bladderdata)
pheno = pData(bladderEset)
edata = exprs(bladderEset)

mod = model.matrix(~as.factor(cancer) + as.factor(batch),data=pheno)
fit = lm.fit(mod,t(edata))
hist(fit$coefficients[2,],col=2,breaks=100)
table(pheno$cancer,pheno$batch)

batch = pheno$batch
modcombat = model.matrix(~1, data=pheno)
modcancer = model.matrix(~cancer, data=pheno)
combat_edata = ComBat(dat=edata, batch=batch, mod=modcombat, par.prior=TRUE, prior.plots=FALSE)

combat_fit = lm.fit(modcancer,t(combat_edata))
hist(combat_fit$coefficients[2,],col=2,breaks=100)

plot(fit$coefficients[2,],combat_fit$coefficients[2,],col=2,
      xlab="Linear Model",ylab="Combat",xlim=c(-5,5),ylim=c(-5,5))
abline(c(0,1),col=1,lwd=3)

mod = model.matrix(~cancer,data=pheno)
mod0 = model.matrix(~1, data=pheno)
sva1 = sva(edata,mod,mod0,n.sv=2)

summary(lm(sva1$sv ~ pheno$batch))

boxplot(sva1$sv[,2] ~ pheno$batch)
points(sva1$sv[,2] ~ jitter(as.numeric(pheno$batch)),col=as.numeric(pheno$batch))

modsv = cbind(mod,sva1$sv)
fitsv = lm.fit(modsv,t(edata))

par(mfrow=c(1,2))
plot(fitsv$coefficients[2,],combat_fit$coefficients[2,],col=2,
      xlab="SVA",ylab="Combat",xlim=c(-5,5),ylim=c(-5,5))
abline(c(0,1),col=1,lwd=3)
plot(fitsv$coefficients[2,], fit$coefficients[2,],col=2,
      xlab="SVA",ylab="linear model",xlim=c(-5,5),ylim=c(-5,5))
abline(c(0,1),col=1,lwd=3)

data(for.exercise)
controls <- rownames(subject.support)[subject.support$cc==0]
use <- seq(1, ncol(snps.10), 10)
ctl.10 <- snps.10[controls,use]

xxmat <- xxt(ctl.10, correct.for.missing=FALSE)
evv <- eigen(xxmat, symmetric=TRUE)
pcs <- evv$vectors[,1:5]

pop <- subject.support[controls,"stratum"]
plot(pcs[,1],pcs[,2],col=as.numeric(pop),
      xlab="PC1",ylab="PC2")
legend(0,0.15,legend=levels(pop),pch=19,col=1:2)
```




```{r ANOVA}
#https://stats.stackexchange.com/questions/89021/how-to-get-only-desirable-comparisons-from-post-hoc
#plot(log2(data[,grep("Ratio.H.L.normalized.", names(data))]))
#dataNorm=log2(data[,grep("Ratio.H.L.normalized.", names(data))])
anova(lm(as.numeric(dataNorm[2,])~factorC*factorS))
summary(dataNorm)
#summary(anova(lm(as.numeric(dataNorm[2,])~factorC*factorS)))
aov((as.numeric(dataNorm[2,])~factorC*factorS))
TukeyHSD(aov((as.numeric(dataNorm[2,])~factorC*factorS)))
tcold=tc
tc=apply(dataNorm,1,function(x){
#tc=apply(dataNorm, 1, function(x) 
  tryCatch(TukeyHSD(aov(x~factorC*factorS),"factorC:factorS", ordered = TRUE),error=function(x){return(rep(1,15))})})
wval=t(sapply(names(tc),function(x){tryCatch(tc[[x]]$`factorC:factorS`[46:60],error=function(x){return(rep(1,15))})}))
write.table(wval,outF,sep="\t")
tc$`F7A0B0;P04370-9;F6ZIA4;P04370-7`
dataNorm[grep("F7A",row.names(data)),]
#try(TukeyHSD(aov((x~factorC*factorS)))))
tcold$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`
plot(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`)
```

```{r write-output}
tc
write.table(t(tc),sep = "\t")
outF = paste0(inpF,"anovaTukey.txt")
class(tc)
names(tc)
do.call(rbind, lapply(names(tc), function(x) data.frame(c(ID=x, tc$x$`factorC:factorS`))))
lapply(names(tc), function(x) write.table(t(t(tcold[[x]]$`factorC:factorS`)[4,]),outF,sep = "\t"))
write.table(t(sapply(tc, 
                     function(x){tryCatch(x$`factorC:factorS`)})),sep="\t")
#function(x){tryCatch(x$`factorC:factorS`,error=function(x){return(NULL)})})),outF,sep="\t")
write.table(t(t(tc$x$`factorC:factorS`)[4,]),sep = "\t")
write.table(t(t(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`$`factorC:factorS`)[4,]),sep = "\t")
dump(tc, file=outF)
```

```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
---
title: "batch-effect-analysis"
author: "Ani"
date: "Monday, Feb 15, 2016"
output: html_document
---
<http://rmarkdown.rstudio.com>.

=======
```{r interpolate}
data<-read.table("c:/Users/animeshs/Downloads/Sunshine hours.txt",sep='\t',header=T)
summary(data$SONNENSCHEINDAUER)
dp=approx(data$SONNENSCHEINDAUER)
summary(dp)
plot(dp$SONNENSCHEINDAUER,data$SONNENSCHEINDAUER)
install.packages('zoo')
library('zoo')
dp<-rollmean(data,1)
dp<-na.approx(data)
summary(dp)
write.table(dp,"c:/Users/animeshs/Downloads/SunshineDP.txt",sep='\t')
```

```{r iPondData}
iPond=read.delim("f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVA.txt",row.names = 48,header=T)
#[,190:213]
label=t(read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Kristian/Groups.txt",row.names = 1,header=T))
```

```{r corr3ction}
hist(iPond$M..Two.way.ANOVA.p.value.Interaction)
hist(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"))
hist((p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH")))
hist(log2(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH")))
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhInt.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhTr.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhCond.txt", sep="\t")

```


```{r data}
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Articles in prep/AID/ID/AIDberitPD21SequestScoreTtestRcomp.txt",sep="\t",row.names=1, header = T) #AIDome
#data <- read.delim("F:/promec/Results/Ani/Mirta/DataActinNormalizedHeatMapCtrlwPCNA.txt",row.names=19,sep="\t",header = T) #BER
#data <- read.delim("F:/promec/Davi/Mei/2016-05_PancreaticCancer/QE/lfq2pep/LFQvalues.txt",row.names=90,sep="\t",header = T)
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15P.txt",fill=T, row.names=86, sep = "\t",header = T)
#df <- read.delim("https://gist.githubusercontent.com/anonymous/0bc36ec5f46757de7c2c/raw/517ef70ab6a68e600f57308e045c2b4669a7abfc/example.txt", header=TRUE, row.names=1, sep="\t")
#df<-df[df$Potential.contaminant!='+',] 
data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15Pr.txt",sep = "\t",header = T)


summary(data)
t.test(data[1,3:5],data[1,6:8],data[1,9:11])
t.test(data[1,3:5],data[1,6:8])

```

```{r lmfit}
source("https://bioconductor.org/biocLite.R")
biocLite("golubEsets")
library(golubEsets)
data()
data(golubTrain)
golubTrain
golub<-data.frame(golubTrain@exprs)
labelgb <- factor(c(rep("ALL",27),rep("AML",11)))
names(golub) <- labelgb
golub2<-log2(golub)
plot(golub2)
biocLite('limma')
library('limma')
fit3 = lmFit(golub2)
e3 = eBayes(fit3)
tab3<-topTable(e3, number=nrow(golub))
head(tab3)
fit3@.Data[[9]]
fit4 = lmFit(golub2, design)
```


```{r shuffle}
ecoli <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples.txt", header=TRUE, na.strings='', sep="\t")
for  (cnt in 1:6){
  cnt
}
  ecoli_shuffle <- ecoli_shuffle[sample(nrow(ecoli)),]
write.csv(file="Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples_shuffled_rn.csv",x=ecoli_shuffle,row.names = F)
```
```{r rawread}
library("MSnbase")
```


```{r prot}
install.packages("PTXQC")
library("PTXQC")
help(package = "PTXQC")
vignette("PTXQC-InputData", package = "PTXQC")
txt_folder = "F:/promec/Elite/LARS/2016/desember/Lymphoma/Results"
r = createReport(txt_folder)
cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```

```{r prot}
require(PTXQC)

## the next require() is needed to prevent a spurious error in certain R versions (mighyou be a bug in R or a package)
## error message is:
##    Error in Scales$new : could not find function "loadMethod"
require(methods)

## specify a path to a MaxQuant txt folder
## Note: This folder needs to be complete (see 'vignette("PTXQC-InputData", package = "PTXQC")')
if (1) {
  ## we will use an example dataset from PRIDE (dataset 2 of the PTXQC publication)
  local_zip = tempfile(fileext=".zip")
  download.file("ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2015/11/PXD003133/txt_20min.zip", destfile = local_zip)
  unzip(local_zip, exdir = tempdir()) ## extracts content
  txt_folder = paste0(tempdir(),"/txt_20min")
} else {
  ## if you have local MaxQuant output, just use it
  txt_folder = "c:/Proteomics/MouseLiver/combined/txt"
}

r = createReport(txt_folder)

cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```


```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
---
title: "batch-effect-analysis"
author: "Ani"
date: "Monday, Feb 15, 2016"
output: html_document
---
<http://rmarkdown.rstudio.com>.

>>>>>>> ebfc112f27001e13d78ab738e1c0e1a49f30d70e
>>>>>>> 7c553c78a017dea6dd835926cd41c986ed3c51c2:batchEffectAnalysis.Rmd
>>>>>>> 4a85c5776a7184459faa8f7536f7456de1421d62
