```{r setup, include=FALSE}
#plot
par(mfrow=c(1,2))
#library/packages
.libPaths( c( .libPaths(), "C:/Users/animeshs/GD/R-4.0.1/library") )
.libPaths()
#[1] "C:/Users/animeshs/GD/R-4.0.1/library"
#install.packages("lattice", repos="http://cran.r-project.org", lib="~/local/R_libs/")
#.libPaths( c( .libPaths(), "C:/Users/animeshs/GD/R_libs") )
#install.packages("writexl")
#install.packages("readxl")
#install.packages("BiocManager")
#BiocManager::install("limma")
#install.packages("matrixStats")
#directory/home
setwd("C:/Users/animeshs/GD/scripts")
getwd()
chkrVector<-c(22.39459,20.48316,21.87155,NA,20.34495)
chkrDF<-as.data.frame(chkrVector)
plot(chkrDF)
hist(as.matrix(chkrDF))
chkrDF[,1]
apply(chkrDF, 2, function(x) t.test(as.numeric(x[c(1:3)]),as.numeric(x[4:5]),na.rm=T,var.equal=T)$p.value)
t.test(as.numeric(chkrVector[c(1:3)]),as.numeric(chkrVector[4:5]),na.rm=T,var.equal=T)$p.value
t.test(as.numeric(c(22.39459,20.48316,21.87155)),as.numeric(c(NA,20.34495)),na.rm=T,var.equal=T)$p.value
t.test(c(22.39459,20.48316,21.87155),c(NA,20.34495),na.rm=T,var.equal=T)$p.value
t.test(c(22.39459,20.48316,21.87155),c(NA,17.97),na.rm=T,var.equal=T)$p.value
t.test(c(21.4316,21.77155),c(NA,17.7),na.rm=T,var.equal=T)$p.value
t.test(c(21.4316,21.77155),c(17.6,17.7),na.rm=T,var.equal=F)$p.value
```

```{r data, echo=FALSE}
#inpD<-"L:\\promec\\HF\\Lars\\2020\\AUGUST\\siri\\combined\\txt_noPHO\\"
inpD <-paste0(getwd(),"/txt_PHOSTY/")
inpF<-paste0(inpD,"proteinGroups.txt")
data <- read.table(inpF,stringsAsFactors = FALSE, header = TRUE, quote = "", comment.char = "", sep = "\t")
#read.delim(inpF,sep="\t",header = T,row.names = 1)
#clean
#data = data[!data$Reverse=="+",]
#data = data[!data$Potential.contaminant=="+",]
#data = data[!data$Only.identified.by.site=="+",]
row.names(data)<-paste(row.names(data),data$Gene.names,data$Protein.IDs,data$Protein.names,data$Score,data$Peptide.counts..unique.,sep=";;")
summary(data)
dim(data)
hist(as.matrix(log2(data[,grep("Intensity",colnames(data))])))
summary(log2(data[,grep("Intensity",colnames(data))]))
selection<-"LFQ.intensity"
LFQ<-as.matrix(data[,grep(selection,colnames(data))])
protNum<-1:ncol(LFQ)
colnames(LFQ)=paste(protNum,sub(selection,"",colnames(LFQ)),protNum,sep="_")
dim(LFQ)
log2LFQ<-log2(LFQ)
log2LFQ[log2LFQ==-Inf]=NA
log2LFQ[log2LFQ==0]=NA
summary(log2LFQ)
hist(log2LFQ)
```

```{r excel, echo = FALSE}
inpW<-paste0(inpD,"log2LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(log2LFQ,rownames(data))),inpW)
```

```{r label}
#inpL<-paste0(inpD,"Groups.txt")
#factor(label$annotation)
#inpL<-paste0(inpD,"Groups.xlsx")
#label<-read_xlsx(inpL)
#colnames(label)=sub("\\-[0-9]+","",colnames(label))
#colnames(label)
#table(label)
label<-colnames(log2LFQ)
summary(label)
rownames(label)
```

```{r control}
ctr<-as.matrix(log2LFQ[,grep("_1_",colnames(log2LFQ))])
ctr[is.na(ctr)]<-0
hist(as.matrix(ctr))
limma::vennDiagram(ctr>0)
```

```{r S}
S<-as.matrix(log2LFQ[,grep("_2_",colnames(log2LFQ))])
S[is.na(S)]<-1
hist(as.matrix(S))
limma::vennDiagram(S>1)
```

```{r MDR}
MDR<-as.matrix(log2LFQ[,grep("_3_",colnames(log2LFQ))])
MDR[is.na(MDR)]<-1
hist(as.matrix(MDR))
limma::vennDiagram(MDR>1)
```

```{r combination}
comb<-as.matrix(log2LFQ[,grep("_4_",colnames(log2LFQ))])
comb[is.na(comb)]<-1
hist(as.matrix(comb))
limma::vennDiagram(comb>1)
```

```{r median}
ctrM<-as.matrix(log2LFQ[,grep("_1_",colnames(log2LFQ))])
#ctrM<-apply(ctrM,1,function(x) if(sum(is.na(x[c(1:3)]))==0){median(x[c(1:3)],na.rm=T)} else{0})
ctrM<-apply(ctrM,1,function(x) if(sum(is.na(x[c(1:2)]))==0){median(x[c(1:2)],na.rm=T)} else{0})
sum(ctrM>0)
Sm<-as.matrix(log2LFQ[,grep("_2_",colnames(log2LFQ))])
Sm<-apply(Sm,1,function(x) if(sum(is.na(x[c(1:3)]))==0){median(x[c(1:3)],na.rm=T)} else{0})
sum(Sm>0)
MDRm<-as.matrix(log2LFQ[,grep("_3_",colnames(log2LFQ))])
MDRm<-apply(MDRm,1,function(x) if(sum(is.na(x[c(1:3)]))==0){median(x[c(1:3)],na.rm=T)} else{0})
sum(MDRm>0)
combM<-as.matrix(log2LFQ[,grep("_4_",colnames(log2LFQ))])
combM<-apply(combM,1,function(x) if(sum(is.na(x[c(1:3)]))==0){median(x[c(1:3)],na.rm=T)} else{0})
sum(combM>0)
hdaM=cbind(ctrM,Sm,MDRm,combM)
hist(hdaM)
Uniprot<-sapply(strsplit(data[,1],";"), `[`, 1)
uniprot<-sapply(strsplit(Uniprot,"-",fixed=TRUE),`[`,1)
row.names(hdaM)<-uniprot
#colnames(hdaM)<-paste(sapply(cn, "[", 2),sapply(cn, "[", 4))
gene<-toupper(as.character(data$Gene.names))
gen<-as.character(sapply(strsplit(gene,";"), `[`, 1))
protein<-as.character(data$Protein.IDs)
summary(hdaM)
#limma::vennDiagram(hdaM[,1:5]>0)
limma::vennDiagram(hdaM>0)
inpW<-paste0(inpF,"sel-LFQ-onevar-x1-control.xlsx")
writexl::write_xlsx(as.data.frame(cbind(uniprot,hdaM,hda,data[,7])),inpW)
```

```{r S_WSRT, echo = FALSE}
ctr<-as.matrix(log2LFQ[,grep("_1_",colnames(log2LFQ))])
ctr[is.na(ctr)]<-0
hist(ctr)
S<-as.matrix(log2LFQ[,grep("_2_",colnames(log2LFQ))])
S[is.na(S)]<-0
hist(S)
dataSellog2grpWSRT<-(S-ctr)
hist(dataSellog2grpWSRT)
#dataSellog2grpWSRT<-dataSellog2grpWSRT[,-ncol(dataSellog2grpWSRT)]
summary(dataSellog2grpWSRT)
dataSellog2grpWSRT[dataSellog2grpWSRT==0]<-NA
hist(dataSellog2grpWSRT)
summary(dataSellog2grpWSRT)
min(dataSellog2grpWSRT,na.rm=T)
max(dataSellog2grpWSRT,na.rm=T)
dim(dataSellog2grpWSRT)
options(nwarnings = 1000000)
minNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else{min(x,na.rm=T)}
  )
summary(warnings())
hist(minNA)
maxNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else{max(x,na.rm=T)}
  )
summary(warnings())
hist(maxNA)
hist(maxNA-minNA)
pValNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else if(mode(x)==median(x,na.rm=T)){0}
    else{wilcox.test(as.numeric(x),na.rm=T)$p.value}
  )
summary(warnings())
hist(pValNA)
S_dataSellog2grpWSRT<-cbind(uniprot,dataSellog2grpWSRT,pValNA,data[,7],rownames(dataSellog2grpWSRT))
#writexl::write_xlsx(as.data.frame(S_dataSellog2grpWSRT),paste0(inpF,"sel-LFQ-S-control.WSRT.xlsx"))
write.csv(as.data.frame(S_dataSellog2grpWSRT),row.names=F,paste0(inpF,"sel-LFQ-S-control.WSRT.csv"))
```

```{r S_wsrtMatlab, echo = FALSE}
inpM<-"L:\\promec\\HF\\Lars\\2020\\AUGUST\\siri\\combined\\txt_PHOSTY\\proteinGroups.txt WSRTpVal.csv"
hdr<-read.csv(inpM,skip=1,header=F,nrows=1,as.is=T)
resMatlab<-read.csv(inpM, skip = 2, header = F)
colnames(resMatlab)<-hdr
summary(resMatlab)
hist(as.matrix(resMatlab[,5]))#-dataSellog2grpWSRT[,1]))#,(resMatlab[,7]+dataSellog2grpWSRT[,1]))
hist(pValNA)
hist(as.matrix(resMatlab[,6])-pValNA)
plot(as.matrix(resMatlab[,6]),pValNA)
```


```{r MDR_WSRT, echo = FALSE}
ctr<-as.matrix(log2LFQ[,grep("_1_",colnames(log2LFQ))])
ctr[is.na(ctr)]<-0
hist(ctr)
M<-as.matrix(log2LFQ[,grep("_3_",colnames(log2LFQ))])
M[is.na(M)]<-0
hist(M)
dataSellog2grpWSRT<-(M-ctr)
hist(dataSellog2grpWSRT)
#dataSellog2grpWSRT<-dataSellog2grpWSRT[,-ncol(dataSellog2grpWSRT)]
summary(dataSellog2grpWSRT)
dataSellog2grpWSRT[dataSellog2grpWSRT==0]<-NA
hist(dataSellog2grpWSRT)
summary(dataSellog2grpWSRT)
min(dataSellog2grpWSRT,na.rm=T)
max(dataSellog2grpWSRT,na.rm=T)
dim(dataSellog2grpWSRT)
options(nwarnings = 1000000)
minNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else{min(x,na.rm=T)}
  )
summary(warnings())
hist(minNA)
maxNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else{max(x,na.rm=T)}
  )
summary(warnings())
hist(maxNA)
hist(maxNA-minNA)
pValNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else if(mode(x)==median(x,na.rm=T)){0}
    else{wilcox.test(as.numeric(x),na.rm=T)$p.value}
  )
summary(warnings())
hist(pValNA)
inpW<-paste0(inpF,"sel-LFQ-MDR-control.WSRT.xlsx")
MDR_dataSellog2grpWSRT<-cbind(uniprot,dataSellog2grpWSRT,pValNA,data[,7],rownames(dataSellog2grpWSRT))
#writexl::write_xlsx(as.data.frame(MDR_dataSellog2grpWSRT),inpW)
write.csv(as.data.frame(MDR_dataSellog2grpWSRT),row.names=F,paste0(inpF,"sel-LFQ-MDR-control.WSRT.csv"))
```

```{r comb_WSRT, echo = FALSE}
ctr<-as.matrix(log2LFQ[,grep("_1_",colnames(log2LFQ))])
ctr[is.na(ctr)]<-0
hist(ctr)
C<-as.matrix(log2LFQ[,grep("_4_",colnames(log2LFQ))])
C[is.na(C)]<-0
hist(C)
dataSellog2grpWSRT<-(C-ctr)
hist(dataSellog2grpWSRT)
#dataSellog2grpWSRT<-dataSellog2grpWSRT[,-ncol(dataSellog2grpWSRT)]
summary(dataSellog2grpWSRT)
dataSellog2grpWSRT[dataSellog2grpWSRT==0]<-NA
hist(dataSellog2grpWSRT)
summary(dataSellog2grpWSRT)
min(dataSellog2grpWSRT,na.rm=T)
max(dataSellog2grpWSRT,na.rm=T)
dim(dataSellog2grpWSRT)
options(nwarnings = 1000000)
minNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else{min(x,na.rm=T)}
  )
summary(warnings())
hist(minNA)
maxNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else{max(x,na.rm=T)}
  )
summary(warnings())
hist(maxNA)
hist(maxNA-minNA)
pValNA = apply(
  dataSellog2grpWSRT, 1, function(x)
    if(sum(!is.na(x))<2){NA}
    else if(mode(x)==median(x,na.rm=T)){0}
    else{wilcox.test(as.numeric(x),na.rm=T)$p.value}
  )
summary(warnings())
hist(pValNA)
inpW<-paste0(inpF,"sel-LFQ-combi-control.WSRT.xlsx")
combi_dataSellog2grpWSRT<-cbind(uniprot,dataSellog2grpWSRT,pValNA,data[,7],rownames(dataSellog2grpWSRT))
#writexl::write_xlsx(as.data.frame(combi_dataSellog2grpWSRT),inpW)
write.csv(as.data.frame(combi_dataSellog2grpWSRT),row.names=F,paste0(inpF,"sel-LFQ-combi-control.WSRT.csv"))
```

```{r sel-WSRTs}
selThr<-0.3
dataS<-cbind(S_dataSellog2grpWSRT,MDR_dataSellog2grpWSRT,combi_dataSellog2grpWSRT)
dataScom<-dataS[(dataS[,5]<selThr) & (dataS[,12]<selThr) & (dataS[,19]<selThr),]
dataSel<-dataS[,c(5,12,19)]
dataSel[dataSel==0]<-1
dataSel[is.na(dataSel)]<-1
colnames(dataSel)<-c("S","MDR","comb")
hist(as.numeric(dataSel))
limma::vennDiagram(dataSel[,c(1:3)]<selThr)
gene<-sapply(strsplit(row.names(dataSel),";;"), `[`, 2)
#gene<-sapply(strsplit(gene,";"), `[`, 1)
dataSel<-cbind(gene,dataSel)#[,ncol(dataSel)+1]<-
writexl::write_xlsx(as.data.frame(dataS),paste0(inpF,"sel-WSRT-onevar.xlsx"))
write.csv(as.data.frame(dataS),row.names=FALSE,paste0(inpF,"sel-WSRT-onevar.csv"))
```

```{r SOS-map}
#https://www.asmscience.org/content/journal/ecosalplus/10.1128/ecosalplus.5.4.3
##Table1
sosD<-read.csv(paste0(inpD,"sosList.txt"),header=T,sep="\t")
#gene<-sapply(strsplit(sosD[,1],";"), `[`, 1)
sosD<-unique(data.frame(gene=unlist(strsplit(as.character(sosD$LexA.dependent.genes),";"))))
#sosD<-cbind(gene,sosD)
#table1<-merge(x=sosD,y=dataSel,by="gene",all=TRUE)
rownames(sosD)<-sosD$gene
sosD$gene<-0.25
colnames(sosD)<-c("sosD")
sosD
```

```{r SOSlike-map}
#https://www.asmscience.org/content/journal/ecosalplus/10.1128/ecosalplus.5.4.3
##Table2
sosLD<-read.csv(paste0(inpD,"sosLikeList.txt"),header=T,sep="\t")
#gene<-sapply(strsplit(sosLD[,1],";"), `[`, 1)
sosLD<-unique(data.frame(gene=unlist(strsplit(as.character(sosLD$Genes.with.potential.SOS.boxes.a),";"))))
#sosLD<-sosLD[-grep("ORF_within",sosLD),]
sosLD<-as.data.frame(sosLD[!sosLD$gene=="ORF_within", ])
rownames(sosLD)<-sosLD$`sosLD[!sosLD$gene == "ORF_within", ]`
sosLD$`sosLD[!sosLD$gene == "ORF_within", ]`<-0.25
colnames(sosLD)<-c("sosLD")
sosLD
```

```{r SdataSel-map}
dataSelmapS<-unique(data.frame(gene=unlist(strsplit(dataSel[dataSel[,2]=="0.25",1],";"))))
rownames(dataSelmapS)<-dataSelmapS$gene
dataSelmapS$gene<-0.25
colnames(dataSelmapS)<-c("S")
dataSelmapS
```

```{r MDRdataSel-map}
dataSelmapMDR<-unique(data.frame(gene=unlist(strsplit(dataSel[dataSel[,3]=="0.25",1],";"))))
rownames(dataSelmapMDR)<-dataSelmapMDR$gene
dataSelmapMDR$gene<-0.25
colnames(dataSelmapMDR)<-c("MDR")
dataSelmapMDR
```

```{r combDataSel-map}
dataSelmapComb<-unique(data.frame(gene=unlist(strsplit(dataSel[dataSel[,4]=="0.25",1],";"))))
rownames(dataSelmapComb)<-dataSelmapComb$gene
dataSelmapComb$gene<-0.25
colnames(dataSelmapComb)<-c("comb")
dataSelmapComb
```

```{r overlapSOSboth}
overlapSOS<-merge(sosD,sosLD,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS<-merge(overlapSOS,dataSelmapS,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS<-merge(overlapSOS,dataSelmapMDR,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS<-merge(overlapSOS,dataSelmapComb,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS[overlapSOS==0]<-1
overlapSOS[is.na(overlapSOS)]<-1
limma::vennDiagram(overlapSOS<selThr)
writexl::write_xlsx(as.data.frame(cbind(row.names(overlapSOS),overlapSOS)),paste0(inpF,"sosMapBoth.xlsx"))
write.csv(as.data.frame(cbind(row.names(overlapSOS),overlapSOS)),row.names=FALSE,paste0(inpF,"sosMapBoth.csv"))
```

```{r overlapSOS}
overlapSOS<-sosD
overlapSOS<-merge(overlapSOS,dataSelmapS,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS<-merge(overlapSOS,dataSelmapMDR,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS<-merge(overlapSOS,dataSelmapComb,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS[overlapSOS==0]<-1
overlapSOS[is.na(overlapSOS)]<-1
limma::vennDiagram(overlapSOS<selThr)
writexl::write_xlsx(as.data.frame(cbind(row.names(overlapSOS),overlapSOS)),paste0(inpF,"sosMap.xlsx"))
write.csv(as.data.frame(cbind(row.names(overlapSOS),overlapSOS)),row.names=FALSE,paste0(inpF,"sosMap.csv"))
```

```{r overlapSOSlike}
overlapSOS<-sosLD
overlapSOS<-merge(overlapSOS,dataSelmapS,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS<-merge(overlapSOS,dataSelmapMDR,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS<-merge(overlapSOS,dataSelmapComb,all=T,by="row.names")
rownames(overlapSOS)<-overlapSOS$Row.names
overlapSOS<-overlapSOS[,-1]
overlapSOS[overlapSOS==0]<-1
overlapSOS[is.na(overlapSOS)]<-1
limma::vennDiagram(overlapSOS<selThr)
writexl::write_xlsx(as.data.frame(cbind(row.names(overlapSOS),overlapSOS)),paste0(inpF,"sosLikeMap.xlsx"))
write.csv(as.data.frame(cbind(row.names(overlapSOS),overlapSOS)),row.names=FALSE,paste0(inpF,"sosLikeMap.csv"))
```

```{r S_tTest, echo = FALSE}
dataSellog2grpTtest<-cbind(S,ctr)
dataSellog2grpTtest[dataSellog2grpTtest==0]<-NA
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-1
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)
pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=F)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<=max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>=min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<=max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>=min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=F)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
dataSellog2grpTtestNum<-apply(dataSellog2grpTtest, 2,as.numeric)
logFCmedianGrp1 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c(sCol:mCol)],na.rm=T)
logFCmedianGrp2 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c((mCol+1):eCol)],na.rm=T)
logFCmedianGrp1[is.nan(logFCmedianGrp1)]=0
logFCmedianGrp2[is.nan(logFCmedianGrp2)]=0
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hdr<-gsub("[^[:alnum:] ]", "",inpF)
hist(log2FCmedianFC,main = hdr)
rowName<-paste(sapply(strsplit(paste(sapply(strsplit(rownames(dataSellog2grpTtest), ";"), "[", 1)), "-"), "[", 1))
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,selection,sCol,eCol,hdr,"wTestBHgrpOut25pQ.xlsx"))
#limma::vennDiagram(cbind(as.matrix(dataSellog2grpTtest[,c(sCol:mCol)]),as.matrix(dataSellog2grpTtest[,c((mCol+1):eCol)])))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
```


```{r M_tTest, echo = FALSE}
dataSellog2grpTtest<-cbind(MDR,ctr)
dataSellog2grpTtest[dataSellog2grpTtest==0]<-NA
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-1
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)
pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=F)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<=max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>=min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<=max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>=min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=F)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
dataSellog2grpTtestNum<-apply(dataSellog2grpTtest, 2,as.numeric)
logFCmedianGrp1 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c(sCol:mCol)],na.rm=T)
logFCmedianGrp2 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c((mCol+1):eCol)],na.rm=T)
logFCmedianGrp1[is.nan(logFCmedianGrp1)]=0
logFCmedianGrp2[is.nan(logFCmedianGrp2)]=0
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hdr<-gsub("[^[:alnum:] ]", "",inpF)
hist(log2FCmedianFC,main = hdr)
rowName<-paste(sapply(strsplit(paste(sapply(strsplit(rownames(dataSellog2grpTtest), ";"), "[", 1)), "-"), "[", 1))
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,selection,sCol,eCol,hdr,"wTestBHgrpOut25pM.xlsx"))
#limma::vennDiagram(cbind(as.matrix(dataSellog2grpTtest[,c(sCol:mCol)]),as.matrix(dataSellog2grpTtest[,c((mCol+1):eCol)])))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
```


```{r C_tTest, echo = FALSE}
dataSellog2grpTtest<-cbind(comb,ctr)
dataSellog2grpTtest[dataSellog2grpTtest==0]<-NA
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-1
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)
pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=F)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<=max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>=min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<=max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(1.5*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>=min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=F)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
dataSellog2grpTtestNum<-apply(dataSellog2grpTtest, 2,as.numeric)
logFCmedianGrp1 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c(sCol:mCol)],na.rm=T)
logFCmedianGrp2 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c((mCol+1):eCol)],na.rm=T)
logFCmedianGrp1[is.nan(logFCmedianGrp1)]=0
logFCmedianGrp2[is.nan(logFCmedianGrp2)]=0
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hdr<-gsub("[^[:alnum:] ]", "",inpF)
hist(log2FCmedianFC,main = hdr)
rowName<-paste(sapply(strsplit(paste(sapply(strsplit(rownames(dataSellog2grpTtest), ";"), "[", 1)), "-"), "[", 1))
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,selection,sCol,eCol,hdr,"wTestBHgrpOut25pC.xlsx"))
#limma::vennDiagram(cbind(as.matrix(dataSellog2grpTtest[,c(sCol:mCol)]),as.matrix(dataSellog2grpTtest[,c((mCol+1):eCol)])))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
```


```{r NC154/63/4comb}
#https://github.com/gowerc/diffdf
#install.packages("devtools")
#devtools::install_github("gowerc/diffdf")
diffdf::diffdf(dataNC154plasmawTest, dataNC154plasmawTest)
dataNC_154_163<-diffdf::diffdf(dataNC154plasmawTest, dataNC163plasmawTest)
plot(dataNC_154_163$VarDiff_Log2MedianChange)
dataNC_164_163<-diffdf::diffdf(dataNC164plasmawTest, dataNC163plasmawTest)
plot(dataNC_164_163$VarDiff_Log2MedianChange)
dataNC_154_164<-diffdf::diffdf(dataNC164plasmawTest, dataNC154plasmawTest)
plot(dataNC_154_164$VarDiff_Log2MedianChange)
```

```{r NC154/63/4BH}
dataV<-data.frame(cbind(NC154=dataNC154plasmawTest$CorrectedPValueBH,NC163=dataNC163plasmawTest$CorrectedPValueBH,NC164=dataNC164plasmawTest$CorrectedPValueBH))
row.names(dataV)<-row.names(data)
summary(dataV<0.1)
dataV[is.na(dataV)]<-1
limma::vennDiagram(dataV<0.1)
writexl::write_xlsx(cbind(row.names(dataV),dataV),paste0(inpF,selection,"wTestBHvennNC154_63_4.xlsx"))
```

```{r NC154/63/4PV}
dataV<-data.frame(cbind(NC154=dataNC154plasmawTest$welchTtestPval,NC163=dataNC163plasmawTest$welchTtestPval,NC164=dataNC164plasmawTest$welchTtestPval))
row.names(dataV)<-row.names(data)
summary(dataV<0.1)
dataV[is.na(dataV)]<-1
limma::vennDiagram(dataV<0.1)
writexl::write_xlsx(cbind(row.names(dataV),dataV),paste0(inpF,selection,"wTestPVvennNC154_63_4.xlsx"))
```

```{r NC154/63/4LFQ}
dataV<-data.frame(cbind(NC154=dataNC154plasmawTest$FoldChanglog2median,NC163=dataNC163plasmawTest$FoldChanglog2median,NC164=dataNC164plasmawTest$FoldChanglog2median))
row.names(dataV)<-row.names(data)
summary(dataV>1)
dataV[is.na(dataV)]<-0
limma::vennDiagram(dataV>1)
limma::vennDiagram(dataV<1)
writexl::write_xlsx(cbind(row.names(dataV),dataV),paste0(inpF,selection,"FoldChangeNC154_63_4.xlsx"))
```

```{r NC154/63/4HC}
dataV<-data.frame(cbind(NC154=dataNC154plasmawTest$Log2MedianChange,NC163=dataNC163plasmawTest$Log2MedianChange,NC164=dataNC164plasmawTest$Log2MedianChange))
row.names(dataV)<-row.names(data)
#dataV[is.na(dataV)]<-0
limma::vennDiagram(dataV>0)
limma::vennDiagram(dataV<0)
```

```{r DEP}
#http://www.matrixscience.com/blog/using-the-quantitation-summary-to-create-reports-and-charts.html
#BiocManager::install("DEP")
library("DEP")
# tweaks to data frame required by DEP
data$Genes = data$Protein.IDs
#data$Majority.protein.IDs = data$Protein.IDs
colnames(data)[colnames(data) == 'Peptide.counts..all.'] <- 'Peptides'
colnames(data)[colnames(data) == 'Peptide.counts..unique.'] <- 'Unique.peptides'
data$Razor...unique.peptide = data$Unique.peptides
data$ID = data$Fasta.headers
colnames(data)
dim(data)
data$Potential.contaminant
data <- filter(data,Potential.contaminant!="+")
dim(data) 
data_unique <- make_unique(data,"Genes", "Proteins", delim = ";")
LFQ_columns <- grep("LFQ.", colnames(data))
experimental_design = read.table(paste0(inpD,"/annodataPC2.txt"),stringsAsFactors = FALSE, header = TRUE, quote = "", comment.char = "", sep = "\t")
experimental_design$label<-paste0("LFQ.intensity.",experimental_design$ID)
experimental_design$condition<-experimental_design$annotation
experimental_design$replicate<-experimental_design$axisLabels
rownames(experimental_design)<-experimental_design$label
summary(experimental_design)
data_se <- make_se(data, LFQ_columns, experimental_design)
summary(data_se)
plot_frequency(data_se)
data_filt <- filter_missval(data_se, thr = 0)
plot_numbers(data_filt)
data_norm <- normalize_vsn(data_filt)
plot_normalization(data_filt, data_norm)
plot_detect(data_filt)
data_imp <- impute(data_norm, fun = "MinProb", q = 0.01)
plot_imputation(data_norm, data_imp)
data_diff <- test_diff(data_imp, type = "control", control = "plasma")
dep <- add_rejections(data_diff, alpha = 0.05, lfc = log2(1.5))
plot_pca(dep, x = 1, y = 2, n = 500, point_size = 4)
plot_cor(dep, significant = TRUE, lower = 0.8, upper = 1, pal = "Reds")
plot_heatmap(dep, type = "centered", kmeans = TRUE, k = 6, col_limit = 4, show_row_names = FALSE, indicate = c("condition", "replicate"))
plot_heatmap(dep, type = "contrast", kmeans = TRUE, k = 6, col_limit = 10, show_row_names = FALSE)
plot_volcano(dep, contrast = "beadVsplasma", label_size = 2, add_names = TRUE)
```

```{r tests}
plot(as.matrix(log2(data)))
cor(as.matrix(log2(data)),nan.rm=T)
cor(data[,colSums(is.na(log2(data)))==0]) 
aovt=aov(as.double(data[1,])~as.character(label[3,]))
summary(aovt)
TukeyHSD(aovt)
```

```{r siggenes}
source("https://bioconductor.org/biocLite.R")
biocLite("siggenes")
library("siggenes")
data("golub")
sam.out <- sam(golub, golub.cl, rand = 123, gene.names = golub.gnames[,3])
samwsrt.out <- sam(golub, golub.cl, method = wilc.stat, rand = 123, gene.names = golub.gnames[,3])
summary(samwsrt.out)
summary(sam.out)
print(sam.out, seq(1.5, 2.4, 0.1))
plot(sam.out, 3.3)
identify(sam.out)
sum.sam.out <- summary(sam.out, 3.3)
sum.sam.out
print(sum.sam.out, varNames = "Proteins")
list.siggenes(sam.out, 3.3)
findDelta(sam.out, fdr = 0.05)
findDelta(sam.out, genes = 200)

```


```{r GF}
biocLite("genefilter")
library("genefilter")
m=matrix(runif(80), 20)
rowttests(m, factor(c("M", "M", "F", "F")))

find.out <- find.a0(golub, golub.cl, rand = 123)
print(find.out, 0.99)
plot(find.out)
ebam(find.out)
ebam.out <- ebam(golub, golub.cl, a0 = 0, rand = 123)
ebam.out <- ebam(golub, golub.cl, method = wilc.ebam, rand =123)
summary(ebam.out, 0.99997)
print(ebam.out, seq(0.91, 0.99, 0.01))
plot(ebam.out, 0.95)

plot(as.matrix(log2(data)))
cor(as.matrix(log2(data)),nan.rm=T)
cor(data[,colSums(is.na(log2(data)))==0]) 
aovt=aov(as.double(data[1,])~as.character(label[3,]))
summary(aovt)
TukeyHSD(aovt)
```

```{r data}
inpF <-"L:/Animesh/Zahoor/CoMMpass_IA8b.txt"
dataG <- read.delim(inpF,row.names=1,sep="\t",header = T)
#attach(data)
summary(dataG)
factorS=names(data)[grep("MMRF", names(data))]
plot(sqrt(as.matrix(dataG)))
devtools::install_github("rstudio/rstudioapi")
library('rstudioapi')
rstudioapi::getThemeInfo()
library(devtools)
install_github('jdrudolph/perseusr')
BiocInstaller::biocLite("checkmate")
library(WGCNA)
allowWGCNAThreads()
memory.limit(size = 256000000)
```

```{r clust}
biocLite("multiClust")
biocLite("dendextend")
library("dendextend")
library("multiClust")
?multiClust
```

```{r coxph}
inpF <-"L:/Animesh/Zahoor/CoMMpass_IA8b.txt"
dataG <- read.delim(inpF,row.names=1,sep="\t",header = T)
summary(dataG)
factorS = names(data)[grep("MMRF", names(data))]
library(survival)
?coxph
inpC <- "L:/Animesh/Zahoor/Compass_Survival.txt"
dataC <- read.delim(inpC, row.names = 1, sep = "\t", header = T)
dataNorm = log2(dataG[, grep("MMRF", names(dataG))] + 1)
summary(dataNorm$MMRF_1021)
hist(dataNorm$MMRF_1021)
factor(dataC$censpfs)
summary(factor(dataC$censpfs))
factorC = rownames(data)[grep("MMRF", names(data))]
resEfsByDis <- coxph(Surv(dataC$ttcpfs, dataC$censpfs) ~ ., data.frame(t(dataNorm)))
```

```{r condition}
expRt <- rep(0:10, each=5)
facR <- factor(expRt) 
Subject <- rep(1:5,time=11)
expR <- rnorm(55, mean=40, sd=5)
df <- data.frame(Subject, expR, expRt)
pairwise.t.test(expR, expRt,p.adj = "none")
#https://stackoverflow.com/questions/27834627/how-to-produce-an-anova-matrix-with-r
```


```{r condition}
library(survival)
#library(cmprsk)
inpC <-"L:/Animesh/Zahoor/Compass_Survival.txt"
dataC <- read.delim(inpC,row.names=1,sep="\t",header = T)
factor(dataC$censpfs)
summary(factor(dataC$censpfs))
factorC=rownames(data)[grep("MMRF", names(data))]
#plot(data)
resEfsByDis <- coxph(Surv(dataC$ttcpfs, dataC$censpfs) ~ . , data.frame(t(dataNorm)))
                       type      = "kaplan-meier",
                       error     = "greenwood",
                       conf.type = "log-log")
install.packages('rms')
library(rms)
rms::survplot(resEfsByDis)
summary(resEfsByDis)
#https://rpubs.com/kaz_yos/cmprsk2
resEfsCoxByDis <- coxph(formula = Surv(dataC$ttcpfs, dataC$censpfs != 1) ~ data=dataC,ties    = c("efron","breslow","exact")[1])
summary(resEfsCoxByDis)
```

```{r condition}
factors=strsplit(names(data)[grep("MMRF", names(data))],"[._]")
factorS=sapply(factors, '[[', 5)
factorS=gsub('[[:digit:]]+', '', factorS)
factorC=sapply(factors, '[[', 7) 
```

```{r TT}
#plot(log2(data[,grep("Ratio.H.L.normalized.", names(data))]))
dataNorm=log2(dataG[,grep("Ratio.H.L.normalized.", names(data))])
dataNorm=log2(dataG[,grep("MMRF", names(dataG))]+1)
summary(dataNorm$MMRF_1021)
hist(dataNorm$MMRF_1021)
#summary(anova(lm(as.numeric(dataNorm[2,])~factorC*factorS)))
#aov((as.numeric(dataNorm[2,])~factorC*factorS))
#TukeyHSD(aov((as.numeric(dataNorm[2,])~factorC*factorS)))
tc=1
tcold=tc
tc=apply(dataNorm,1,function(x){
#tc=apply(dataNorm, 1, function(x) 
  tryCatch(TukeyHSD(aov(x~factorC*factorS),"factorC:factorS", ordered = TRUE),error=function(x){return(rep(1,15))})})
wval=t(sapply(names(tc),function(x){tryCatch(tc[[x]]$`factorC:factorS`[46:60],error=function(x){return(rep(1,15))})}))
tc$`E5RIU6;A0A087WZZ9;A0A024QZP7;P06493;P06493-2;K7ELV5;Q00526;F8VXD2;F8VYH9;F8VZZ0;F8VTV8;F8VWX7;F8VZ51;K7EJ83;H0YAZ9;Q9BVE2;E7EUK8;E5RGN0;A0A087X209;A0A087WZU2;F5H6Z0;P11802-2;P11802;Q96Q40-2;Q96Q40-4;Q96Q40-3;O94921-3;Q96Q40-5;Q96Q40;O94921-2;O94921;Q07002;Q07002-2;Q00536;Q00536-3;Q07002-3;Q00537-2;Q00537;Q00536-2;Q14004-2;Q14004`
dataNorm[grep("F7A",row.names(data)),]
#try(TukeyHSD(aov((x~factorC*factorS)))))
tcold$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`
plot(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`)
?aov
```

```{r ANOVA}
#https://stats.stackexchange.com/questions/89021/how-to-get-only-desirable-comparisons-from-post-hoc
#plot(log2(data[,grep("Ratio.H.L.normalized.", names(data))]))
dataNorm=log2(data[,grep("Ratio.H.L.normalized.", names(data))])
summary(dataNorm)
#summary(anova(lm(as.numeric(dataNorm[2,])~factorC*factorS)))
#aov((as.numeric(dataNorm[2,])~factorC*factorS))
#TukeyHSD(aov((as.numeric(dataNorm[2,])~factorC*factorS)))
tc=1
tcold=tc
tc=apply(dataNorm,1,function(x){
#tc=apply(dataNorm, 1, function(x) 
  tryCatch(TukeyHSD(aov(x~factorC*factorS),"factorC:factorS", ordered = TRUE),error=function(x){return(rep(1,15))})})
wval=t(sapply(names(tc),function(x){tryCatch(tc[[x]]$`factorC:factorS`[46:60],error=function(x){return(rep(1,15))})}))
tc$`E5RIU6;A0A087WZZ9;A0A024QZP7;P06493;P06493-2;K7ELV5;Q00526;F8VXD2;F8VYH9;F8VZZ0;F8VTV8;F8VWX7;F8VZ51;K7EJ83;H0YAZ9;Q9BVE2;E7EUK8;E5RGN0;A0A087X209;A0A087WZU2;F5H6Z0;P11802-2;P11802;Q96Q40-2;Q96Q40-4;Q96Q40-3;O94921-3;Q96Q40-5;Q96Q40;O94921-2;O94921;Q07002;Q07002-2;Q00536;Q00536-3;Q07002-3;Q00537-2;Q00537;Q00536-2;Q14004-2;Q14004`
dataNorm[grep("F7A",row.names(data)),]
#try(TukeyHSD(aov((x~factorC*factorS)))))
tcold$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`
plot(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`)
```


```{r write-output}
write.table(wval,outF,sep="\t")
tc
write.table(t(tc),sep = "\t")
outF = paste0(inpF,"anovaTukey.txt")
class(tc)
names(tc)
do.call(rbind, lapply(names(tc), function(x) data.frame(c(ID=x, tc$x$`factorC:factorS`))))
lapply(names(tc), function(x) write.table(t(t(tcold[[x]]$`factorC:factorS`)[4,]),outF,sep = "\t"))
write.table(t(sapply(tc, 
                     function(x){tryCatch(x$`factorC:factorS`)})),sep="\t")
#function(x){tryCatch(x$`factorC:factorS`,error=function(x){return(NULL)})})),outF,sep="\t")
write.table(t(t(tc$x$`factorC:factorS`)[4,]),sep = "\t")
write.table(t(t(tc$`A0A087WNP6;Q4VAA2-2;Q4VAA2;A0A087WRM0;F8WGL9;A0A087WS49`$`factorC:factorS`)[4,]),sep = "\t")
dump(tc, file=outF)
```

```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
---
title: "batch-effect-analysis"
author: "Ani"
date: "Monday, Feb 15, 2016"
output: html_document
---
<http://rmarkdown.rstudio.com>.

=======
```{r interpolate}
data<-read.table("c:/Users/animeshs/Downloads/Sunshine hours.txt",sep='\t',header=T)
summary(data$SONNENSCHEINDAUER)
dp=approx(data$SONNENSCHEINDAUER)
summary(dp)
plot(dp$SONNENSCHEINDAUER,data$SONNENSCHEINDAUER)
install.packages('zoo')
library('zoo')
dp<-rollmean(data,1)
dp<-na.approx(data)
summary(dp)
write.table(dp,"c:/Users/animeshs/Downloads/SunshineDP.txt",sep='\t')
```

```{r iPondData}
iPond=read.delim("f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVA.txt",row.names = 48,header=T)
#[,190:213]
label=t(read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Kristian/Groups.txt",row.names = 1,header=T))
```

```{r corr3ction}
hist(iPond$M..Two.way.ANOVA.p.value.Interaction)
hist(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"))
hist((p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH")))
hist(log2(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH")))
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Interaction,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhInt.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Treatment,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhTr.txt", sep="\t")
write.table(p.adjust(iPond$M..Two.way.ANOVA.p.value.Cond,method = "BH"), "f:/promec/Elite/Karine/170216_IP/dePep/Log2LFQproteinGroupsBHtTests2wayANOVAbhCond.txt", sep="\t")

```


```{r data}
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Articles in prep/AID/ID/AIDberitPD21SequestScoreTtestRcomp.txt",sep="\t",row.names=1, header = T) #AIDome
#data <- read.delim("F:/promec/Results/Ani/Mirta/DataActinNormalizedHeatMapCtrlwPCNA.txt",row.names=19,sep="\t",header = T) #BER
#data <- read.delim("F:/promec/Davi/Mei/2016-05_PancreaticCancer/QE/lfq2pep/LFQvalues.txt",row.names=90,sep="\t",header = T)
#data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15P.txt",fill=T, row.names=86, sep = "\t",header = T)
#df <- read.delim("https://gist.githubusercontent.com/anonymous/0bc36ec5f46757de7c2c/raw/517ef70ab6a68e600f57308e045c2b4669a7abfc/example.txt", header=TRUE, row.names=1, sep="\t")
#df<-df[df$Potential.contaminant!='+',] 
data <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/iPOND/batch1and2lfqaimpnovaTtest90C15Pr.txt",sep = "\t",header = T)


summary(data)
t.test(data[1,3:5],data[1,6:8],data[1,9:11])
t.test(data[1,3:5],data[1,6:8])

```

```{r lmfit}
source("https://bioconductor.org/biocLite.R")
biocLite("golubEsets")
library(golubEsets)
data()
data(golubTrain)
golubTrain
golub<-data.frame(golubTrain@exprs)
labelgb <- factor(c(rep("ALL",27),rep("AML",11)))
names(golub) <- labelgb
golub2<-log2(golub)
plot(golub2)
biocLite('limma')
library('limma')
fit3 = lmFit(golub2)
e3 = eBayes(fit3)
tab3<-topTable(e3, number=nrow(golub))
head(tab3)
fit3@.Data[[9]]
fit4 = lmFit(golub2, design)
```


```{r shuffle}
ecoli <- read.delim("Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples.txt", header=TRUE, na.strings='', sep="\t")
for  (cnt in 1:6){
  cnt
}
  ecoli_shuffle <- ecoli_shuffle[sample(nrow(ecoli)),]
write.csv(file="Y:/felles/PROTEOMICS and XRAY/Ani/Christina/Strainssamples_shuffled_rn.csv",x=ecoli_shuffle,row.names = F)
```
```{r rawread}
library("MSnbase")
```


```{r prot}
install.packages("PTXQC")
library("PTXQC")
help(package = "PTXQC")
vignette("PTXQC-InputData", package = "PTXQC")
txt_folder = "F:/promec/Elite/LARS/2016/desember/Lymphoma/Results"
r = createReport(txt_folder)
cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```

```{r prot}
require(PTXQC)

## the next require() is needed to prevent a spurious error in certain R versions (mighyou be a bug in R or a package)
## error message is:
##    Error in Scales$new : could not find function "loadMethod"
require(methods)

## specify a path to a MaxQuant txt folder
## Note: This folder needs to be complete (see 'vignette("PTXQC-InputData", package = "PTXQC")')
if (1) {
  ## we will use an example dataset from PRIDE (dataset 2 of the PTXQC publication)
  local_zip = tempfile(fileext=".zip")
  download.file("ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2015/11/PXD003133/txt_20min.zip", destfile = local_zip)
  unzip(local_zip, exdir = tempdir()) ## extracts content
  txt_folder = paste0(tempdir(),"/txt_20min")
} else {
  ## if you have local MaxQuant output, just use it
  txt_folder = "c:/Proteomics/MouseLiver/combined/txt"
}

r = createReport(txt_folder)

cat(paste0("\nReport generated as '", r$report_file, "'\n\n"))

```


```{r prot}
source("https://bioconductor.org/biocLite.R")
biocLite("PGA")
library("PGA")
biocLite("XML")
biocLite("colorspace")
install.packages("Rcpp")
library("PGA")
dat_file <- system.file("extdata/input", "mascot.dat",package = "PGA")
parserGear(file = dat_file, db = dbfile, decoyPrefix = "#REV#",xmx = 1,thread = 8,outdir = "parser_outdir_mascot")

```

```{r ir}
i=0.9
n=1
y1<-0
y2<-0

for(x1 in seq(0.1, 10, 0.5)){
  for(x2 in seq(0.1, 10, 0.5)){
    y1<-c(((1-i)*n+x1)/((1-i)*n+x2),y1)
    y2<-c(x1/x2,y2)
  }
}
plot(log2(y1),log2(y2))
plot((y1),(y2))

```


```{r MSstats}
source("http://www.bioconductor.org/biocLite.R")
biocLite('GOpro')
library(GOpro)
biocLite('GOSim')
library(GOSim)
genes=c("207","208","596","901","780","3169","9518","2852","26353","8614","7494")
getGOInfo(genes)
biocLite("RTopper")
library(RTopper)
biocLite("PGA")
biocLite("rTANDEM")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
vcffile <- system.file("extdata/input", "PGA.vcf",package="PGA")
bedfile <- system.file("extdata/input", "junctions.bed",package="PGA")
gtffile <- system.file("extdata/input", "transcripts.gtf",package="PGA")
annotation <- system.file("extdata", "annotation",package="PGA")
outfile_path<-"db/"
outfile_name<-"test"
library(BSgenome.Hsapiens.UCSC.hg19)
dbfile <- dbCreator(gtfFile=gtffile,vcfFile=vcffile,bedFile=bedfile,
                    annotation_path=annotation,outfile_name=outfile_name,
                    genome=Hsapiens,outdir=outfile_path)
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jdk1.8.0_45\\jre\\')
parserGear("F:/promec/Gygi/b1906_293T_proteinID_01A_QE3_122212.pep.xml", db = dbfile, decoyPrefix="#REV#",xmx = 1,thread = 8,outdir = "parser_outdir")
```

```{r MSstats}
#https://bioconductor.org/packages/3.3/bioc/vignettes/MSstats/inst/doc/MSstats-manual.pdf
library('MSstats')
QuantData<-dataProcess(SRMRawData)
head(QuantData$ProcessedData)
## based on multiple comparisons (T1 vs T3; T1 vs T7; T1 vs T9)
comparison1<-matrix(c(-1,0,1,0,0,0,0,0,0,0),nrow=1)
comparison2<-matrix(c(-1,0,0,0,0,0,1,0,0,0),nrow=1)
comparison3<-matrix(c(-1,0,0,0,0,0,0,0,1,0),nrow=1)
comparison<-rbind(comparison1,comparison2, comparison3)
row.names(comparison)<-c("T3-T1","T7-T1","T9-T1")
testResultMultiComparisons<-groupComparison(contrast.matrix=comparison,data=QuantData)
## Calculate sample size for future experiments:
#(1) Minimal number of biological replicates per condition
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=TRUE,
desiredFC=c(1.25,1.75),FDR=0.05,power=0.8)
#(2) Power calculation
designSampleSize(data=testResultMultiComparisons$fittedmodel,numSample=2,
desiredFC=c(1.25,1.75),FDR=0.05,power=TRUE)```
```

```{r SAM}
#https://github.com/MikeJSeo/SAM
install.packages(c("samr", "matrixStats", "GSA", "shiny", "shinyFiles", "openxlsx"))
source("http://bioconductor.org/biocLite.R")
biocLite("impute")
library(shiny)
library(shinyFiles)
runGitHub("SAM", "MikeJSeo")
```



```{r pkgs}
install.packages('BEST') #http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3489534/
library('BEST') #https://sourceforge.net/projects/mcmc-jags
install.packages('mice') #http://www.helsinki.fi/~rosenstr/papers/ImputationNotes.pdf
library('mice') #http://www.r-bloggers.com/imputing-missing-data-with-r-mice-package/
install.packages('RColorBrewer')
library(RColorBrewer)
source("http://bioconductor.org/biocLite.R")
biocLite("preprocessCore")
library('preprocessCore')
#install.packages("gplots")
# library(gplots) 
#library(ggplot2)
```

```{r JAGS}
y1 <- c(5.77, 5.33, 4.59, 4.33, 3.66, 4.48)
y2 <- c(3.88, 3.55, 3.29, 2.59, 2.33, 3.59)
priors <- list(muM = 6, muSD = 2)
out <- BESTmcmc(y1, y2, priors=priors, parallel=FALSE) # where def prior sigmaMode = sd(y), sigmaSD = sd(y)*5, nuMean = 30, nuSD = 30), y = c(y1, y2).
plot(out)
plot(y1,y2)
summary(t.test(y1,y2))
hist(y2)
hist(datasel)
```

```{r Cluster}
colscl <- colorRampPalette(c("darkblue", "chocolate"), space = "rgb")(100)
heatmap(na.omit(as.matrix(data)),col=colscl)
```

```{r ColumnSel}
datasel<-data[, grep("^Tumor_*", colnames(data))]
summary(datasel)
t.test(datasel)
datanonan=na.omit(as.matrix(datasel))
summary(datanonan)
heatmap(na.omit(as.matrix(datasel)),col=colscl)
```

```{r NA}
#impute(data, method = "mixed",randna = fData(naset)$randna,mar = "knn", mnar = "min")
md.pattern(datasel)
dataimp <- mice(datasel,m=length(colnames(datasel))/2,maxit=length(colnames(datasel)),meth='pmm',seed=1)
dataimp <- complete(dataimp,length(colnames(datasel))/2)
heatmap(as.matrix(dataimp),col=colscl)
```

```{r Scale}
datascale=t(scale(t(dataimp))) # takes away the polynomial in YFP data
heatmap(as.matrix(datascale),col=colscl)
```

```{r DistCor}
dataasmat<-data.matrix(datascale)
cor(data)
dissimilarityc <- 1 - cor(dataasmat,method = "s")
distancec <- as.dist(dissimilarityc)
col.clus <- hclust(distancec, "aver")
dissimilarityr <- 1 - cor(t(dataasmat),method = "s")
distancer <- as.dist(dissimilarityr)
row.clus <- hclust(distancer, "aver")
heatmap(as.matrix(datascale),col=colscl)
cr <- rainbow(nrow(dataasmat))
cc <- rainbow(ncol(dataasmat))
heatmap((na.omit(dataasmat)), RowSideColors = cr, ColSideColors = cc,Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus),cexRow=1.5,cexCol=1.5 )
```

```{r face}
#http://www.r-bloggers.com/analyze-face-emotions-with-r/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+RBloggers+%28R+bloggers%29

library("httr")
library("XML")
library("stringr")
library("ggplot2")

img.url     = 'https://i.imgur.com/Njtw8tx.jpg'
URL.emoface = 'https://api.projectoxford.ai/emotion/v1.0/recognize'
emotionKEY = '25eac991d0c8474fb501d704c6454b05'
mybody = list(url = img.url)
faceEMO = POST(
  url = URL.emoface,
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = emotionKEY)),
  body = mybody,
  encode = 'json'
)
faceEMO
trump = httr::content(faceEMO)[[1]]
trump
o<-as.data.frame(as.matrix(trump$scores))
o$V1 <- lapply(strsplit(as.character(o$V1 ), "e"), "[", 1)
o$V1<-as.numeric(o$V1)
colnames(o)[1] <- "Level"
o$Emotion<- rownames(o)
ggplot(data=o, aes(x=Emotion, y=Level)) +
  geom_bar(stat="identity")

img.url = 'https://www.whitehouse.gov/sites/whitehouse.gov/files/images/first-family/44_barack_trump[1].jpg'
faceURL = "https://api.projectoxford.ai/face/v1.0/detect?returnFaceId=true&returnFaceLandmarks=true&returnFaceAttributes=age"
mybody = list(url = img.url)
faceKEY = '6b86f25e09e84291b2bcba8d61d6297e'
faceResponse = POST(
  url = faceURL, 
  content_type('application/json'), add_headers(.headers = c('Ocp-Apim-Subscription-Key' = faceKEY)),
  body = mybody,
  encode = 'json'
)

# Show request results (if Status=200, request is okay)
faceResponse

# Reuqest results from face analysis
trumpR = httr::content(faceResponse)[[1]]

# Define results in data frame
OR<-as.data.frame(as.matrix(trumpR$faceLandmarks))

# Make some transformation to data frame
OR$V2 <- lapply(strsplit(as.character(OR$V1), "\\="), "[", 2)
OR$V2 <- lapply(strsplit(as.character(OR$V2), "\\,"), "[", 1)
colnames(OR)[2] <- "X"
OR$X<-as.numeric(OR$X)

OR$V3 <- lapply(strsplit(as.character(OR$V1), "\\y = "), "[", 2)
OR$V3 <- lapply(strsplit(as.character(OR$V3), "\\)"), "[", 1)
colnames(OR)[3] <- "Y"
OR$Y<-as.numeric(OR$Y)

OR$V1<-NULL
```

```{r Quantile}
dataasmat<-normalize.quantiles(dataasmat)
rn<-rownames(dataasmat)
cn<-colnames(dataasmat)
colnames(dataasmat)<-cn
rownames(dataasmat)<-rn
```

```{r AdjPval}
p.adjust(10^(-data$X.log10P),method = c("hochberg"), n = length(data$X.log10P))
p.adjust(data$pv,method = c("hochberg"), n = length(data$pv))
```

```{r, echo=FALSE}
p <- ggplot(data)
print(p)
#heatmap2(data)
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
---
title: "batch-effect-analysis"
author: "Ani"
date: "Monday, Feb 15, 2016"
output: html_document
---
<http://rmarkdown.rstudio.com>.

>>>>>>> ebfc112f27001e13d78ab738e1c0e1a49f30d70e

---
output: 
  html_notebook: 
    fig_caption: yes
    toc: yes
runtime: shiny
---
```{r msms}
#inpF <- "L:/Animesh/IdaJ/Bafh1resMQ/msmsScans.txt"
#data <- read.delim(inpF, sep = "\t", header = T)
#dataID = data[, grep("Identified", names(data))]
#summary(dataID)
```

```{r protein}
#inpF <- '../../combined/txt/proteinGroups.txt'
inpF <- 'L:/promec/Animesh/Kathleen/181006_HCT_totx_JC_FirstIsoformMap.txt'
#inpF <- 'L:/promec/HF/Lars/2018/november/Barbara/181101_BARBARA_/181101_BARBARA_NUC_1_16926_20181107014533.xlsx'
```

```{r LFQ}
#data <- read.delim(inpF, row.names = 4, sep = "\t", header = T)
data <- read.delim(inpF, sep = "\t", header = T)
install.packages("openxlsx")
install.packages("readxl") 
library("openxlsx")
library("readxl") 
inpF='L:/promec/Animesh/Kathleen/181108_HCT_totPHOS__R1-(1).xlsx'
data <- read_excel(inpF)
data <- read_xlsx(inpF)
dataprot<-data$Checked#=="FALSE"
head(dataprot,2)
summary(data)
plot(data[,23:26])
#data <- read.delim(inpF)
colName = "LFQ.intensity."
dataNorm = data[, grep(colName, names(data))]
#dataNorm = log2(dataNorm+1)
dataNorm = log2(dataNorm)
names(dataNorm) = gsub(pattern = colName, replacement = "", x = names(dataNorm))
summary(as.matrix(data[,23]))
data[is.na(data)]<-0
dNo0=data[data$`Protein FDR Confidence: Combined`!="0",]
summary(as.matrix(dNo0[,23]))
outF='L:/promec/Animesh/Kathleen/181108_HCT_totPHOS__R1-(1)NAto0.xlsx'
write.xlsx(dNo0,outF)
```

```{r Summary}
summary(log2(data))
plot(log2(data))
```

```{r test}
data<-read.table('F:/OneDrive - NTNU/KWS/ribosomerations.txt',comment="D",header=T,sep="\t",row.names = 1)
t.test(data)
t.test(data[1,],data[2,],paired = TRUE)
t.test(as.matrix(data[1,]),as.matrix(data[3,]),paired = TRUE)
t.test(as.matrix(data[2,]),as.matrix(data[3,]),paired = TRUE)
t.test(as.matrix(data[1,]),as.matrix(data[1,]),paired = TRUE)
t.test(data[2,],data[3,])
dDGE <- estimateDisp(dDGE)
dDGE$samples
plotBCV(dDGE)
dTest <- exactTest(dDGE)#, pair = c("Baf","Ctr"))
dTest$table
summary(decideTestsDGE(dTest))
dTestFDR <- topTags(dTest, n = 1000, sort.by = "none")
summary(dTestFDR)
tc = t.test(dataNorm[3,])$p.value
outF = paste0(inpF,"test.csv")
class(tc)
names(tc)
write.csv(tc, file = outF)
pairwise.t.test(data[3,],(factorC))
heatmap.2(as.matrix(data))
pheatmap(as.matrix(data))
```

```{r test}
cor.test(as.matrix(data[1,]),as.matrix(data[3,]))
pv=acf(as.matrix(data[1,]),as.matrix(data[3,]))
t.test(c(as.matrix(data[,1]),as.matrix(data[,2]),as.matrix(data[,3])),(c(2,2,2,1,1,1,3,3,3)),paired = TRUE)
summary(aov(c(as.matrix(data[,1]),as.matrix(data[,2]),as.matrix(data[,3]))~(c(2,2,2,1,1,1,3,3,3))))
summary(aov((c(scale(data[,1]),scale(data[,2]),scale(data[,3])))~(c('c2','c2','c2','c1','c1','c1','c3','c3','c3'))))
TukeyHSD(aov((c(scale(data[,1]),scale(data[,2]),scale(data[,3])))~(c('c2','c2','c2','c1','c1','c1','c3','c3','c3'))))
TukeyHSD(aov(log2(c(as.matrix(data[,1]),as.matrix(data[,2]),as.matrix(data[,3])))~(c('c2','c2','c2','c1','c1','c1','c3','c3','c3'))))
TukeyHSD(aov(c(as.matrix(data[,1]),as.matrix(data[,2]),as.matrix(data[,3]))~(c('c2','c2','c2','c1','c1','c1','c3','c3','c3'))))
t.test(as.matrix(data[1,]),as.matrix(data[2,]),paired = TRUE)
install.packages("lmtest") 
library(lmtest)
lmtest::grangertest(data[1,],data[2,],1)
plot(t(data))
cor(as.matrix(t(data)),method="spearman")
dANOVA=c(data[,1],data[,2],data[,3])
anova(dANOVA,factor(c(1,1,1,2,2,2,3,3,3)))
cor.test(c(2.1,0.1,2.1),c(2.5,1.1,2.1))
plot(dANOVA)
tZscore(data[,1])
cor.test(as.numeric(data[1,]),as.numeric(data[3,]))#,method="spearman")
plot(scale(log2(data[,1])),scale(log2(data[,3])))

```

```{r LIMMA}
#source("https://bioconductor.org/biocLite.R")
#biocLite("limma")
library("limma")
dataNorm[dataNorm == -Inf] <- NaN
boxplot(dataNorm)
plotDensities(dataNorm)
colSums(dataNorm, na.rm = TRUE)
plotMDS(dataNorm)
norm_dN <- mean(colSums(dataNorm,na.rm = TRUE)) / colSums(dataNorm,na.rm = TRUE)
dataN_sl <- sweep(dataNorm, 2, norm_dN, FUN = "*")
plotDensities(dataN_sl)
boxplot(dataN_sl)
```

```{r edgeR}
#source("https://bioconductor.org/biocLite.R")
#biocLite("edgeR")
library("edgeR")
summary(is.na(dataNorm))
dataN_s2 <- dataNorm
dataN_s2[is.na(dataN_s2)] <- 0
dataN_s2 <- sweep(dataN_s2, 2, calcNormFactors(dataN_s2), FUN = "/") 
summary(is.na(dataN_s2))
colSums(dataN_s2)
boxplot(dataN_s2)
plotMDS(dataN_s2)
pairs(dataN_s2)
```


```{r test}
#factorC=c(rep(c("Baf","Ctr"),each=4))
factorC <- c("Baf","Ctr","Baf","Ctr","Ctr","Baf","Baf","Ctr")
t(names(dataNorm))
dDGE <- DGEList(counts = dataN_s2, group = factorC)
dDGE <- calcNormFactors(dDGE)
dDGE <- estimateDisp(dDGE)
dDGE$samples
plotBCV(dDGE)
dTest <- exactTest(dDGE)#, pair = c("Baf","Ctr"))
dTest$table
summary(decideTestsDGE(dTest))
dTestFDR <- topTags(dTest, n = 1000, sort.by = "none")
summary(dTestFDR)
#?TukeyHSD(aov((as.numeric(dataID[160,])~factorC)),"factorC", ordered = TRUE)
#tc=apply(dataID,1,function(x){
#tc=apply(dataNorm, 1, function(x)   tryCatch(TukeyHSD(aov(x~factorC,"factorC", ordered = TRUE)),error=function(x){return(rep(1,20))})}) tryCatch(TukeyHSD(aov(x~factorC,"factorC")),error=function(x){return(rep(1,20))})})
tc = t.test(dataNorm[3,])$p.value
outF = paste0(inpF,"test.csv")
class(tc)
names(tc)
write.csv(tc, file = outF)
pairwise.t.test(dataNorm[3,],(factorC))
```

```{r shiny, echo = FALSE}
#https://rmarkdown.rstudio.com/authoring_shiny.html
#install.packages('shiny')
library(shiny)
dN_shiny <- dataNorm
dN_shiny[is.na(dN_shiny)] <- 0
summary(is.na(dN_shiny))
#ctrl-shift-K
#rmarkdown::run
#shiny::renderUI
```

```{r slider, echo=FALSE}
inputPanel(
  sliderInput("bins", "#bins:", min = 1, max = 100, value = 30),
  textInput("dens", "Density:", "auto",value = 0.6), 
  selectInput("expression", label = "Sample:",choices = names(dN_shiny), selected = names(dN_shiny)[1]))
renderText({paste("Inputs:", input$bins,input$dens,input$expression)})
```

```{r plot, echo = FALSE}
renderPlot({
  hist(dN_shiny[,as.numeric(which(names(dN_shiny) == input$expression))],breaks = as.numeric(input$bins),probability = TRUE,col="orange")
  lines(density(dN_shiny[,as.numeric(which(names(dN_shiny) == input$expression))], adjust = as.numeric(input$dens)),col="blue")
  })
  
```




