##MQ
```{r setup, include=FALSE}
#library/packages
.libPaths( c("L:/promec/Animesh/R/win-library/4.0", .libPaths()) )
.libPaths()
#plot
#par(mfrow=c(1,2))
#data
inpD <-"L:/promec/Animesh/Kristian/"
thr<-20
selThr=0.01
selThrFC=1
hdr<-gsub("[^[:alnum:] ]", "",inpD)
setwd(inpD)
getwd()
#test
chkrVector<-c(22.39459,20.48316,21.87155,NA,20.34495)
chkrDF<-as.data.frame(chkrVector)
plot(chkrDF)
hist(as.matrix(chkrDF))
chkrDF[,1]
apply(chkrDF, 2, function(x) t.test(as.numeric(x[c(1:3)]),as.numeric(x[4:5]),na.rm=T,var.equal=T)$p.value)
t.test(as.numeric(chkrVector[c(1:3)]),as.numeric(chkrVector[4:5]),na.rm=T,var.equal=T)$p.value
t.test(as.numeric(c(22.39459,20.48316,21.87155)),as.numeric(c(NA,20.34495)),na.rm=T,var.equal=T)$p.value
t.test(c(20.39459,19.48316),c(NA,25.34495),na.rm=T,var.equal=T)$p.value
t.test(c(22.39459,20.48316,21.87155),c(NA,17.97),na.rm=T,var.equal=T)$p.value
t.test(c(21.4316,21.77155),c(NA,17.7),na.rm=T,var.equal=T)$p.value
t.test(c(21.4316,21.77155),c(17.6,17.7),na.rm=T,var.equal=F)$p.value
t.test(c(24.00049729, 23.97020359, 24.91145672), c(24.93876407, 24.84723287, 23.98674622),var.equal = T)$p.value
#[1] 0.5304706
t.test(c(24.00049729, 23.97020359, 24.91145672), c(24.93876407, 24.84723287, 23.98674622),var.equal = F)$p.value
#[1] 0.5304824
```

```{r proteinGroups, echo=FALSE}
inpF<-paste0(inpD,"proteinGroups.txt")
data <- read.table(inpF,stringsAsFactors = FALSE, header = TRUE, quote = "", comment.char = "", sep = "\t")
#clean
#data = data[!data$Reverse=="+",]
#data = data[!data$Potential.contaminant=="+",]
#data = data[!data$Only.identified.by.site=="+",]
row.names(data)<-paste(row.names(data),data$Fasta.headers,data$Protein.IDs,data$Protein.names,data$Gene.names,data$Score,data$Peptide.counts..unique.,sep=";;")
summary(data)
dim(data)
hist(as.matrix(log2(data[,grep("Intensity",colnames(data))])))
summary(log2(data[,grep("Intensity",colnames(data))]))
selection<-"LFQ.intensity."
LFQ<-as.matrix(data[,grep(selection,colnames(data))])
#protNum<-1:ncol(LFQ)
#protNum<-"LFQ intensity"#1:ncol(LFQ)
#colnames(LFQ)=paste(protNum,sub(selection,"",colnames(LFQ)),sep=";")
colnames(LFQ)=sub(selection,"",colnames(LFQ))
dim(LFQ)
log2LFQ<-log2(LFQ)
log2LFQ[log2LFQ==-Inf]=NA
log2LFQ[log2LFQ==0]=NA
summary(log2LFQ)
hist(log2LFQ)
inpW<-paste0(inpD,hdr,"log2LFQ.xlsx")
rowName<-paste(sapply(strsplit(paste(sapply(strsplit(data$Majority.protein.IDs, ";",fixed=T), "[", 1)), " "), "[", 1))
writexl::write_xlsx(as.data.frame(cbind(rowName,log2LFQ,rownames(data))),inpW)
```

```{r label}
inpL<-paste0(inpD,"SampleKey.txt")
label<-read.table(inpL,header=T,row.names=1,sep="\t")#, colClasses=c(rep("factor",3)))
label
```

```{r corHC}
#corrplot::corrplot(cor(log2LFQ))
#log2LFQ0=log2LFQ
#log2LFQ0[log2LFQ0==NA]=0
#pheatmap::pheatmap(log2LFQ0,clustering_distance_rows = "binary",clustering_distance_cols = "correlation",fontsize_row=4,cluster_cols=FALSE,cluster_rows=T,color = my_palette,fontsize_col  = 4)
scale=3
log2LFQimp<-matrix(rnorm(dim(log2LFQ)[1]*dim(log2LFQ)[2],mean=mean(log2LFQ,na.rm = T)-scale,sd=sd(log2LFQ,na.rm = T)/(scale)), dim(log2LFQ)[1],dim(log2LFQ)[2])
log2LFQimp[log2LFQimp<0]<-0
bk1 <- c(seq(-3,-0.01,by=0.01))
bk2 <- c(seq(0.01,3,by=0.01))
bk <- c(bk1,bk2)  #combine the break limits for purpose of graphing
my_palette <- c(colorRampPalette(colors = c("darkblue", "white"))(n = length(bk1)-1),
              "gray", "gray",
              c(colorRampPalette(colors = c("white","darkred"))(n = length(bk2)-1)))
colnames(log2LFQimp)<-colnames(log2LFQ)
#pheatmap::pheatmap(log2LFQimp,clustering_distance_rows = "binary",clustering_distance_cols = "correlation",fontsize_row=8,cluster_cols=T,cluster_rows=T,color = my_palette,fontsize_col  = 8)
#pheatmap::pheatmap(log2LFQimp,clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=1,cluster_cols=T,cluster_rows=T,fontsize_col  = 8)
log2LFQimpCorr<-cor(log2LFQ,use="pairwise.complete.obs",method="spearman")
colnames(log2LFQimpCorr)<-colnames(log2LFQ)
rownames(log2LFQimpCorr)<-colnames(log2LFQ)
svgPHC<-pheatmap::pheatmap(log2LFQimpCorr,clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean",fontsize_row=8,cluster_cols=T,cluster_rows=T,fontsize_col  = 8)
#pheatmap::pheatmap(log2LFQimpCorr,clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean",fontsize_row=4,cluster_cols=T,cluster_rows=T,fontsize_col  = 4,labels_col = label$Sample.Number,labels_row = label$Cell.Type_Sample.Group)
#svgPHC<-pheatmap::pheatmap(log2LFQimpCorr,clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean",fontsize_row=6,cluster_cols=T,cluster_rows=T,fontsize_col  = 6,labels_row = paste(label$RawFnum,label$ID_ROLF,label$Tank,label$Diet,label$Fish))
#ggplot2::ggsave(file=paste0(inpF,hdr,selection,"log2spearClust.svg"),plot=svgPHC)#,  
assign(paste0(selection,"clusterData"),log2LFQimpCorr)
write.csv(cbind(rownames(log2LFQimpCorr),log2LFQimpCorr),paste0(inpF,hdr,selection,"log2spearClust.csv"),row.names=F)
```

```{r PCA, echo = FALSE}
options(nwarnings = 1000000)
scale=3
set.seed(scale)
log2LFQimp<-matrix(rnorm(dim(log2LFQ)[1]*dim(log2LFQ)[2],mean=mean(log2LFQ,na.rm = T)-scale,sd=sd(log2LFQ,na.rm = T)/(scale)), dim(log2LFQ)[1],dim(log2LFQ)[2])
hist(log2LFQimp)
log2LFQimp[log2LFQimp<0]<-0
summary(log2LFQimp)
dataHighNormLog2<-log2LFQimp
dataHighNormLog2[!is.na(log2LFQ)]<-log2LFQ[!is.na(log2LFQ)]
summary(dataHighNormLog2)
hist(dataHighNormLog2)
log2LFQt<-t(dataHighNormLog2)
row.names(log2LFQt)<-colnames(log2LFQ)
log2LFQt<-log2LFQt[-grep("PreOC",rownames(log2LFQt)),]
log2LFQtPCA<-prcomp(log2LFQt)
log2LFQtPCAsumm<-summary(log2LFQtPCA)
plot(prcomp(log2LFQt))
plot(log2LFQtPCA$x[,1], log2LFQtPCA$x[,2], pch = 16, col = factor(rownames(log2LFQt)),xlab = paste0("PC1 (", round(100*log2LFQtPCAsumm$importance[2,1],1), "%)"), ylab = paste0("PC2 (", round(100*log2LFQtPCAsumm$importance[2,2],1), "%)"),main=paste("PCA 1/2 with 0 containing proteinGroups imputed","\nProtein groups", dim(log2LFQt)[2],"across samples",dim(log2LFQt)[1]))
op <- par(cex = 0.4)
legend("topright", col = factor(rownames(log2LFQt)), legend = factor(rownames(log2LFQt)), pch = 16)
```

```{r PCAtools}
#git checkout 3bf916b0e2edfb24dae8707dfab9b12269b03462 diffExprANOVA.rmd
#rm gg* in (C:\MR-4.0.2\library, L:\promec\Animesh\R\win-library\4.0)
#http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html
#devtools::install_github('kevinblighe/PCAtools')
#BiocManager::install("airway")
#library(airway)
#https://github.com/kevinblighe/PCAtools#plot-the-entire-project-on-a-single-panel
colnames(log2LFQimp)=colnames(log2LFQ)
rownames(log2LFQimp)=rownames(log2LFQ)
colnames(log2LFQimp)==rownames(label)
p <- PCAtools::pca(log2LFQimp, metadata = label, removeVar = 0.1)
## -- removing the lower 10% of variables based on variance
p.prcomp <- list(sdev = p$sdev,rotation = data.matrix(p$loadings),x = data.matrix(p$rotated),center = TRUE, scale = TRUE)
class(p.prcomp) <- 'prcomp'
biplot(p.prcomp, showLoadings = TRUE, lab = NULL)
#https://rpubs.com/PandulaP/PCA_for_Predictions , for this simple example, just use a chunk of the original data for the prediction
newdata <- t(log2LFQimp[,seq(1,ncol(log2LFQimp))])
predict(p.prcomp, newdata = newdata)[,1:5]
```

```{r Mono}
sel<-"Mono"
hda<-as.matrix(log2LFQ[,grep(sel,label$Cell,ignore.case=T)])
summary(hda)
sum(hda>thr,na.rm=T)
hist(as.matrix(hda))
limma::vennDiagram(hda>thr)
hdaM<-apply(hda,1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{0})
sum(hdaM>thr,na.rm=T)
inpW<-paste0(inpF,hdr,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(paste0(sel,"median"),hdaM)
assign(paste0(sel,"all"),hda)
```

```{r M1}
sel<-"M1"
hda<-as.matrix(log2LFQ[,grep(sel,label$Cell,ignore.case=T)])
summary(hda)
sum(hda>thr,na.rm=T)
hist(as.matrix(hda))
limma::vennDiagram(hda>thr)
hdaM<-apply(hda,1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{0})
sum(hdaM>thr,na.rm=T)
inpW<-paste0(inpF,hdr,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(paste0(sel,"median"),hdaM)
assign(paste0(sel,"all"),hda)
```

```{r MO_D}
sel<-"MO"
hda<-as.matrix(log2LFQ[,grep(sel,label$Cell)])
summary(hda)
sum(hda>thr,na.rm=T)
hist(as.matrix(hda))
limma::vennDiagram(hda>thr)
hdaM<-apply(hda,1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{0})
sum(hdaM>thr,na.rm=T)
inpW<-paste0(inpF,hdr,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(paste0(sel,"median"),hdaM)
assign(paste0(sel,"all"),hda)
```


```{r vennMedians}
hda<-cbind(Monomedian,MOmedian,M1median)
hist(hda)
limma::vennDiagram(hda>thr)
writexl::write_xlsx(as.data.frame(cbind(Fasta=rownames(data),hda,MOmedian,MOmedian,M1median)),paste0(inpF,hdr,selection,"median.xlsx"))
write.csv(cbind(Fasta=rownames(data),hda,MOmedian,MOmedian,M1median),paste0(inpF,hdr,selection,"median.csv"),row.names = F)
```

```{r tM1Mono, echo = FALSE}
dataSellog2grpTtest<-cbind(M1all,Monoall)
comp<-"M1_Mono"
dataSellog2grpTtest[dataSellog2grpTtest==0]=NA
summary(dataSellog2grpTtest)
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
cvThr<-0.05
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-1
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)

pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=T)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&(min(x[c(sCol:mCol)],na.rm=T)==max(x[c(sCol:mCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c((mCol+1):eCol)]))>1&(min(x[c((mCol+1):eCol)],na.rm=T)==max(x[c((mCol+1):eCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1&(sd(x[c(sCol:mCol)],na.rm=T)/mean(x[c(sCol:mCol)],na.rm=T))<cvThr){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1&(sd(x[c((mCol+1):eCol)],na.rm=T)/mean(x[c((mCol+1):eCol)],na.rm=T))<cvThr){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=1){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=T)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=1&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=T)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
summary(pValNA)
dfpValNA<-as.data.frame(ceiling(pValNA))
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
length(pValNA)-(sum(is.na(pValNA))+sum(ceiling(pValNA)==0,na.rm = T))
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
dataSellog2grpTtestNum<-apply(dataSellog2grpTtest, 2,as.numeric)
logFCmedianGrp1 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c(sCol:mCol)],na.rm=T)
logFCmedianGrp2 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c((mCol+1):eCol)],na.rm=T)
logFCmedianGrp1[is.nan(logFCmedianGrp1)]=0
logFCmedianGrp2[is.nan(logFCmedianGrp2)]=0
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hist(log2FCmedianFC)
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,comp,selThr,selThrFC,cvThr,"tTestBH.xlsx"))
write.csv(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,comp,selThr,selThrFC,cvThr,"tTestBH.csv"),row.names = F)
#limma::vennDiagram(cbind(as.matrix(dataSellog2grpTtest[,c(sCol:mCol)]),as.matrix(dataSellog2grpTtest[,c((mCol+1):eCol)])))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
data$RowGeneUniProtScorePeps<-data$Gene.names
ttest.results[is.na(ttest.results)]=selThr
Significance=ttest.results$CorrectedPValueBH<selThr&ttest.results$CorrectedPValueBH>0&abs(ttest.results$Log2MedianChange)>selThrFC
sum(Significance)
dsub <- subset(ttest.results,Significance)
#p <- ggplot2::ggplot(ttest.results,ggplot2::aes(Log2MedianChange,PValueMinusLog10))+ ggplot2::geom_point(ggplot2::aes(color=Significance)) 
#p<-p + ggplot2::theme_bw(base_size=8) + ggplot2::geom_text(data=dsub,ggplot2::aes(label=RowGeneUniProtScorePeps),hjust=0, vjust=0,size=1) + ggplot2::scale_fill_gradient(low="white", high="darkblue") + ggplot2::xlab("Log2 Median Change") + ggplot2::ylab("-Log10 P-value")
#f=paste(file,proc.time()[3],".jpg")
#ggplot2::ggsave(paste0(inpF,hdr,selection,sCol,eCol,comp,selThr,selThrFC,cvThr,"VolcanoTest.svg"), p)
#print(p)
assign(paste0(comp,"_tTest"),dsub)
```

```{r wM1Mono, echo = FALSE}
dataSellog2grpTtest<-cbind(M1all,Monoall)
comp<-"M1_Mono"
dataSellog2grpTtest[dataSellog2grpTtest==0]=NA
summary(dataSellog2grpTtest)
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
qThr<-3
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-1
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)
pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=F)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&(min(x[c(sCol:mCol)],na.rm=T)==max(x[c(sCol:mCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c((mCol+1):eCol)]))>1&(min(x[c((mCol+1):eCol)],na.rm=T)==max(x[c((mCol+1):eCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<=max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>=min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<=max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>=min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=F)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
dataSellog2grpTtestNum<-apply(dataSellog2grpTtest, 2,as.numeric)
logFCmedianGrp1 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c(sCol:mCol)],na.rm=T)
logFCmedianGrp2 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c((mCol+1):eCol)],na.rm=T)
logFCmedianGrp1[is.nan(logFCmedianGrp1)]=0
logFCmedianGrp2[is.nan(logFCmedianGrp2)]=0
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hist(log2FCmedianFC)
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"wTestBH.xlsx"))
write.csv(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"wTestBH.csv"),row.names = F)
#limma::vennDiagram(cbind(as.matrix(dataSellog2grpTtest[,c(sCol:mCol)]),as.matrix(dataSellog2grpTtest[,c((mCol+1):eCol)])))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
data$RowGeneUniProtScorePeps<-data$Gene.names
ttest.results[is.na(ttest.results)]=selThr
Significance=ttest.results$CorrectedPValueBH<selThr&ttest.results$CorrectedPValueBH>0&abs(ttest.results$Log2MedianChange)>selThrFC
sum(Significance)
dsub <- subset(ttest.results,Significance)
#p <- ggplot2::ggplot(ttest.results,ggplot2::aes(Log2MedianChange,PValueMinusLog10))+ ggplot2::geom_point(ggplot2::aes(color=Significance)) 
#p<-p + ggplot2::theme_bw(base_size=8) + ggplot2::geom_text(data=dsub,ggplot2::aes(label=RowGeneUniProtScorePeps),hjust=0, vjust=0,size=1) + ggplot2::scale_fill_gradient(low="white", high="darkblue") + ggplot2::xlab("Log2 Median Change") + ggplot2::ylab("-Log10 P-value")
#f=paste(file,proc.time()[3],".jpg")
#ggplot2::ggsave(paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"WolcanoTest.svg"), p)
#print(p)
assign(paste0(comp,"_wTest"),dsub)
```
```{r vennMedians}
comp<-"M1_Mono"
hda<-merge(M1_Mono_tTest,M1_Mono_wTest,all=T,by="RowGeneUniProtScorePeps")
plot(hda$Log2MedianChange.x,hda$Log2MedianChange.y)
plot(hda$TtestPval.x,hda$TtestPval.y)
assign(paste0("TestsMerge"),hda)
hda[is.na(hda)]=0
limma::vennDiagram(hda[,c(6,18)]>0)
hda[hda$TtestPval.x==0&hda$TtestPval.y>0,]#E9PKB0
writexl::write_xlsx(as.data.frame(hda),paste0(inpF,hdr,comp,"testsM.xlsx"))
write.csv(hda,paste0(inpF,hdr,comp,"testsM.csv"),row.names = F)
```

```{r wMOMono, echo = FALSE}
dataSellog2grpTtest<-cbind(MOall,Monoall)
comp<-"MO_Mono"
dataSellog2grpTtest[dataSellog2grpTtest==0]=NA
summary(dataSellog2grpTtest)
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
qThr<-3
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-1
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)
pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=F)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&(min(x[c(sCol:mCol)],na.rm=T)==max(x[c(sCol:mCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c((mCol+1):eCol)]))>1&(min(x[c((mCol+1):eCol)],na.rm=T)==max(x[c((mCol+1):eCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<=max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>=min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<=max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>=min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=F)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
dataSellog2grpTtestNum<-apply(dataSellog2grpTtest, 2,as.numeric)
logFCmedianGrp1 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c(sCol:mCol)],na.rm=T)
logFCmedianGrp2 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c((mCol+1):eCol)],na.rm=T)
logFCmedianGrp1[is.nan(logFCmedianGrp1)]=0
logFCmedianGrp2[is.nan(logFCmedianGrp2)]=0
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hist(log2FCmedianFC)
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"wTestBH.xlsx"))
write.csv(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"wTestBH.csv"),row.names = F)
#limma::vennDiagram(cbind(as.matrix(dataSellog2grpTtest[,c(sCol:mCol)]),as.matrix(dataSellog2grpTtest[,c((mCol+1):eCol)])))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
data$RowGeneUniProtScorePeps<-data$Gene.names
ttest.results[is.na(ttest.results)]=selThr
Significance=ttest.results$CorrectedPValueBH<selThr&ttest.results$CorrectedPValueBH>0&abs(ttest.results$Log2MedianChange)>selThrFC
sum(Significance)
dsub <- subset(ttest.results,Significance)
#p <- ggplot2::ggplot(ttest.results,ggplot2::aes(Log2MedianChange,PValueMinusLog10))+ ggplot2::geom_point(ggplot2::aes(color=Significance)) 
#p<-p + ggplot2::theme_bw(base_size=8) + ggplot2::geom_text(data=dsub,ggplot2::aes(label=RowGeneUniProtScorePeps),hjust=0, vjust=0,size=1) + ggplot2::scale_fill_gradient(low="white", high="darkblue") + ggplot2::xlab("Log2 Median Change") + ggplot2::ylab("-Log10 P-value")
#f=paste(file,proc.time()[3],".jpg")
#ggplot2::ggsave(paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"WolcanoTest.svg"), p)
#print(p)
assign(paste0(comp,"_wTest"),dsub)
```

```{r wM1MO, echo = FALSE}
dataSellog2grpTtest<-cbind(M1all,MOall)
comp<-"M1_MO"
dataSellog2grpTtest[dataSellog2grpTtest==0]=NA
summary(dataSellog2grpTtest)
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
qThr<-3
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-1
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)
pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=F)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&(min(x[c(sCol:mCol)],na.rm=T)==max(x[c(sCol:mCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c((mCol+1):eCol)]))>1&(min(x[c((mCol+1):eCol)],na.rm=T)==max(x[c((mCol+1):eCol)],na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))<=max(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))==1&(as.numeric(quantile(as.numeric(x[c(sCol:mCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c(sCol:mCol)]),na.rm=T)))>=min(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[1])-(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))<=max(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))==1&sum(!is.na(x[c((mCol+1):eCol)]))>=2&(as.numeric(quantile(as.numeric(x[c((mCol+1):eCol)]),na.rm=T)[5])+(qThr*IQR(as.numeric(x[c((mCol+1):eCol)]),na.rm=T))>=min(as.numeric(x[c(sCol:mCol)]),na.rm=T))){1}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=F)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
dataSellog2grpTtestNum<-apply(dataSellog2grpTtest, 2,as.numeric)
logFCmedianGrp1 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c(sCol:mCol)],na.rm=T)
logFCmedianGrp2 = matrixStats::rowMedians(dataSellog2grpTtestNum[,c((mCol+1):eCol)],na.rm=T)
logFCmedianGrp1[is.nan(logFCmedianGrp1)]=0
logFCmedianGrp2[is.nan(logFCmedianGrp2)]=0
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hist(log2FCmedianFC)
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"wTestBH.xlsx"))
write.csv(ttest.results,paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"wTestBH.csv"),row.names = F)
#limma::vennDiagram(cbind(as.matrix(dataSellog2grpTtest[,c(sCol:mCol)]),as.matrix(dataSellog2grpTtest[,c((mCol+1):eCol)])))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
limma::vennDiagram(cbind(logFCmedianGrp1,logFCmedianGrp2))
data$RowGeneUniProtScorePeps<-data$Gene.names
ttest.results[is.na(ttest.results)]=selThr
Significance=ttest.results$CorrectedPValueBH<selThr&ttest.results$CorrectedPValueBH>0&abs(ttest.results$Log2MedianChange)>selThrFC
sum(Significance)
dsub <- subset(ttest.results,Significance)
#p <- ggplot2::ggplot(ttest.results,ggplot2::aes(Log2MedianChange,PValueMinusLog10))+ ggplot2::geom_point(ggplot2::aes(color=Significance)) 
#p<-p + ggplot2::theme_bw(base_size=8) + ggplot2::geom_text(data=dsub,ggplot2::aes(label=RowGeneUniProtScorePeps),hjust=0, vjust=0,size=1) + ggplot2::scale_fill_gradient(low="white", high="darkblue") + ggplot2::xlab("Log2 Median Change") + ggplot2::ylab("-Log10 P-value")
#f=paste(file,proc.time()[3],".jpg")
#ggplot2::ggsave(paste0(inpF,hdr,selection,sCol,eCol,qThr,comp,selThr,selThrFC,"WolcanoTest.svg"), p)
#print(p)
assign(paste0(comp,"_wTest"),dsub)
```

```{r vennMedians}
comp<-"M1_Mono_MO"
hda<-merge(M1_Mono_wTest,MO_Mono_wTest,all=T,by="RowGeneUniProtScorePeps")
hda<-merge(hda,M1_MO_wTest,all=T,by="RowGeneUniProtScorePeps")
plot(hda$Log2MedianChange.x,hda$Log2MedianChange.y)
plot(hda$TtestPval.x,hda$TtestPval.y)
hda[is.na(hda)]=0
limma::vennDiagram(hda[,c(6,18,30)]>0)
hda[hda$TtestPval.x==0&hda$TtestPval>0,]#Q2M385                
hda[hda$TtestPval.x>0&hda$TtestPval>0,]#P05362                           
writexl::write_xlsx(as.data.frame(hda),paste0(inpF,hdr,comp,"testsM.xlsx"))
write.csv(hda,paste0(inpF,hdr,comp,"testsM.csv"),row.names = F)
assign(paste0("AllTestsMerge"),hda)
```
