```{r setup, include=FALSE}
#library/packages
.libPaths( "L:/promec/Animesh/R/win-library/4.0", c( .libPaths()) )
.libPaths()
#plot
par(mfrow=c(1,2))
#data
inpD <-"L:/promec/USERS/SynnÃ¸ve/20210709_Synnove_6samples/HF/combined/txt/"
thr<-0
selThr=0.3
selThrFC=0.0
hdr<-gsub("[^[:alnum:] ]", "",inpD)
setwd(inpD)
getwd()
#test
chkrVector<-c(22.39459,20.48316,21.87155,NA,20.34495)
chkrDF<-as.data.frame(chkrVector)
plot(chkrDF)
hist(as.matrix(chkrDF))
chkrDF[,1]
apply(chkrDF, 2, function(x) t.test(as.numeric(x[c(1:3)]),as.numeric(x[4:5]),na.rm=T,var.equal=T)$p.value)
t.test(as.numeric(chkrVector[c(1:3)]),as.numeric(chkrVector[4:5]),na.rm=T,var.equal=T)$p.value
t.test(as.numeric(c(22.39459,20.48316,21.87155)),as.numeric(c(NA,20.34495)),na.rm=T,var.equal=T)$p.value
t.test(c(20.39459,19.48316),c(NA,25.34495),na.rm=T,var.equal=T)$p.value
t.test(c(22.39459,20.48316,21.87155),c(NA,17.97),na.rm=T,var.equal=T)$p.value
t.test(c(21.4316,21.77155),c(NA,17.7),na.rm=T,var.equal=T)$p.value
t.test(c(21.4316,21.77155),c(17.6,17.7),na.rm=T,var.equal=F)$p.value
```

```{r proteinGroups, echo=FALSE}
inpF<-paste0(inpD,"proteinGroups.txt")
data <- read.table(inpF,stringsAsFactors = FALSE, header = TRUE, quote = "", comment.char = "", sep = "\t")
#clean
#data = data[!data$Reverse=="+",]
#data = data[!data$Potential.contaminant=="+",]
#data = data[!data$Only.identified.by.site=="+",]
row.names(data)<-paste(row.names(data),data$Fasta.headers,data$Protein.IDs,data$Protein.names,data$Gene.names,data$Score,data$Peptide.counts..unique.,sep=";;")
summary(data)
dim(data)
hist(as.matrix(log2(data[,grep("Intensity",colnames(data))])))
summary(log2(data[,grep("Intensity",colnames(data))]))
selection<-"LFQ.intensity."
LFQ<-as.matrix(data[,grep(selection,colnames(data))])
#protNum<-1:ncol(LFQ)
#protNum<-"LFQ intensity"#1:ncol(LFQ)
#colnames(LFQ)=paste(protNum,sub(selection,"",colnames(LFQ)),sep=";")
colnames(LFQ)=sub(selection,"",colnames(LFQ))
dim(LFQ)
log2LFQ<-log2(LFQ)
log2LFQ[log2LFQ==-Inf]=NA
log2LFQ[log2LFQ==0]=NA
summary(log2LFQ)
hist(log2LFQ)
inpW<-paste0(inpD,"log2LFQ.xlsx")
rowName<-paste(sapply(strsplit(paste(sapply(strsplit(data$Fasta.headers, "|",fixed=T), "[", 2)), "-"), "[", 1))
writexl::write_xlsx(as.data.frame(cbind(rowName,log2LFQ,rownames(data))),inpW)
```

```{r label}
inpL<-paste0(inpD,"label.txt")
label<-read.table(inpL,header=T,row.names=1,sep="\t")#, colClasses=c(rep("factor",3)))
label
```

```{r corHC}
#corrplot::corrplot(cor(log2LFQ))
#log2LFQ0=log2LFQ
#log2LFQ0[log2LFQ0==NA]=0
#pheatmap::pheatmap(log2LFQ0,clustering_distance_rows = "binary",clustering_distance_cols = "correlation",fontsize_row=4,cluster_cols=FALSE,cluster_rows=T,color = my_palette,fontsize_col  = 4)
scale=3
log2LFQimp<-matrix(rnorm(dim(log2LFQ)[1]*dim(log2LFQ)[2],mean=mean(log2LFQ,na.rm = T)-scale,sd=sd(log2LFQ,na.rm = T)/(scale)), dim(log2LFQ)[1],dim(log2LFQ)[2])
log2LFQimp[log2LFQimp<0]<-0
bk1 <- c(seq(-3,-0.01,by=0.01))
bk2 <- c(seq(0.01,3,by=0.01))
bk <- c(bk1,bk2)  #combine the break limits for purpose of graphing
my_palette <- c(colorRampPalette(colors = c("darkblue", "white"))(n = length(bk1)-1),
              "gray", "gray",
              c(colorRampPalette(colors = c("white","darkred"))(n = length(bk2)-1)))
colnames(log2LFQimp)<-colnames(log2LFQ)
#pheatmap::pheatmap(log2LFQimp,clustering_distance_rows = "binary",clustering_distance_cols = "correlation",fontsize_row=8,cluster_cols=T,cluster_rows=T,color = my_palette,fontsize_col  = 8)
#pheatmap::pheatmap(log2LFQimp,clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=1,cluster_cols=T,cluster_rows=T,fontsize_col  = 8)
log2LFQimpCorr<-cor(log2LFQ,use="pairwise.complete.obs",method="spearman")
colnames(log2LFQimpCorr)<-colnames(log2LFQ)
rownames(log2LFQimpCorr)<-colnames(log2LFQ)
svgPHC<-pheatmap::pheatmap(log2LFQimpCorr,clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean",fontsize_row=8,cluster_cols=T,cluster_rows=T,fontsize_col  = 8)
#pheatmap::pheatmap(log2LFQimpCorr,clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean",fontsize_row=4,cluster_cols=T,cluster_rows=T,fontsize_col  = 4,labels_col = label$Sample.Number,labels_row = label$Cell.Type_Sample.Group)
svgPHC<-pheatmap::pheatmap(log2LFQimpCorr,clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean",fontsize_row=6,cluster_cols=T,cluster_rows=T,fontsize_col  = 6,labels_row = paste(label$RawFnum,label$ID_ROLF,label$Tank,label$Diet,label$Fish))
ggplot2::ggsave(file=paste0(inpF,selection,"log2spearClust.svg"),plot=svgPHC)#,  
assign(paste0(selection,"clusterData"),log2LFQimpCorr)
write.csv(cbind(rownames(log2LFQimpCorr),log2LFQimpCorr),paste0(inpF,selection,"log2spearClust.csv"),row.names=F)
```


```{r Group}
sel<-"C"
hda<-as.matrix(log2LFQ[,grep(sel,colnames(log2LFQ))])
summary(hda)
hist(as.matrix(hda))
C<-apply(hda,1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{0})
sum(C>thr,na.rm=T)
sel<-"A"
hda<-as.matrix(log2LFQ[,grep(sel,colnames(log2LFQ))])
summary(hda)
hist(as.matrix(hda))
A<-apply(hda,1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{0})
sum(A>thr,na.rm=T)
hdaM<-cbind(A,C)
limma::vennDiagram(hdaM>thr)
inpW<-paste0(inpF,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(sel,hdaM)
```

```{r Mt, echo = FALSE}
comp<-"A|C"
dataSellog2grpTtest<-as.matrix(log2LFQ[,grep(comp,colnames(log2LFQ))])
dataSellog2grpTtest[dataSellog2grpTtest==0]=NA
summary(dataSellog2grpTtest)
row.names(dataSellog2grpTtest)<-row.names(data)
sCol<-1
eCol<-6
mCol<-3
cvThr<-0.05
t.test(as.numeric(dataSellog2grpTtest[1,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[1,c((mCol+1):eCol)]),na.rm=T)$p.value
chkr<-2
sum(!is.na(dataSellog2grpTtest[chkr,c(1:eCol)]))
t.test(as.numeric(dataSellog2grpTtest[chkr,c(sCol:mCol)]),as.numeric(dataSellog2grpTtest[chkr,c((mCol+1):eCol)]),na.rm=T,var.equal=T)$p.value
dim(dataSellog2grpTtest)
options(nwarnings = 1000000)
pValNA = apply(
  dataSellog2grpTtest, 1, function(x)
    if(sum(!is.na(x[c(sCol:mCol)]))<2&sum(!is.na(x[c((mCol+1):eCol)]))<2){NA}
    else if(sum(is.na(x[c(sCol:mCol)]))==0&sum(is.na(x[c((mCol+1):eCol)]))==0){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),var.equal=T)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>1&sum(!is.na(x[c((mCol+1):eCol)]))<1&(sd(x[c(sCol:mCol)],na.rm=T)/mean(x[c(sCol:mCol)],na.rm=T))<cvThr){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))<1&sum(!is.na(x[c((mCol+1):eCol)]))>1&(sd(x[c((mCol+1):eCol)],na.rm=T)/mean(x[c((mCol+1):eCol)],na.rm=T))<cvThr){0}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=2&sum(!is.na(x[c((mCol+1):eCol)]))>=1){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=T)$p.value}
    else if(sum(!is.na(x[c(sCol:mCol)]))>=1&sum(!is.na(x[c((mCol+1):eCol)]))>=2){
      t.test(as.numeric(x[c(sCol:mCol)]),as.numeric(x[c((mCol+1):eCol)]),na.rm=T,var.equal=T)$p.value}
  else{NA}
  )
summary(warnings())
hist(pValNA)
summary(pValNA)
dfpValNA<-as.data.frame(ceiling(pValNA))
pValNAdm<-cbind(pValNA,dataSellog2grpTtest,row.names(data))
pValNAminusLog10 = -log10(pValNA+.Machine$double.xmin)
hist(pValNAminusLog10)
library(scales)
pValNAminusLog10=squish(pValNAminusLog10,c(0,5))
hist(pValNAminusLog10)
length(pValNA)-(sum(is.na(pValNA))+sum(ceiling(pValNA)==0,na.rm = T))
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBHnaMinusLog10 = -log10(pValBHna+.Machine$double.xmin)
hist(pValBHnaMinusLog10)
logFCmedianGrp1 = apply(dataSellog2grpTtest[,c(sCol:mCol)],1, function(x) median(x,na.rm=T))
logFCmedianGrp1=if(is.null(dim(dataSellog2grpTtest[,c(sCol:mCol)]))){dataSellog2grpTtest[,c(sCol:mCol)]} else{apply(dataSellog2grpTtest[,c(sCol:mCol)],1,function(x) median(x,na.rm=T))}
#summary(logFCmedianGrp11-logFCmedianGrp1)
logFCmedianGrp2=if(is.null(dim(dataSellog2grpTtest[,c((mCol+1):eCol)]))){dataSellog2grpTtest[,c((mCol+1):eCol)]} else{apply(dataSellog2grpTtest[,c((mCol+1):eCol)],1,function(x) median(x,na.rm=T))}
logFCmedianGrp1[is.na(logFCmedianGrp1)]=0
logFCmedianGrp2[is.na(logFCmedianGrp2)]=0
hda<-cbind(logFCmedianGrp1,logFCmedianGrp2)
plot(hda)
limma::vennDiagram(hda>0)
logFCmedian = logFCmedianGrp1-logFCmedianGrp2
logFCmedianFC = 2^(logFCmedian+.Machine$double.xmin)
logFCmedianFC=squish(logFCmedianFC,c(0.01,100))
hist(logFCmedianFC)
log2FCmedianFC=log2(logFCmedianFC)
hist(log2FCmedianFC)

comp<-gsub("[^[:alnum:] ]", "", comp)
ttest.results = data.frame(Uniprot=rowName,PValueMinusLog10=pValNAminusLog10,FoldChanglog2median=logFCmedianFC,CorrectedPValueBH=pValBHna,TtestPval=pValNA,dataSellog2grpTtest,Log2MedianChange=logFCmedian,RowGeneUniProtScorePeps=rownames(dataSellog2grpTtest))
writexl::write_xlsx(ttest.results,paste0(inpF,selection,sCol,eCol,comp,selThr,selThrFC,cvThr,"tTestBH.xlsx"))
write.csv(ttest.results,paste0(inpF,selection,sCol,eCol,comp,selThr,selThrFC,cvThr,"tTestBH.csv"),row.names = F)
data$RowGeneUniProtScorePeps<-data$Gene.names
ttest.results[is.na(ttest.results)]=selThr
Significance=ttest.results$CorrectedPValueBH<selThr&ttest.results$CorrectedPValueBH>0&abs(ttest.results$Log2MedianChange)>selThrFC
sum(Significance)
dsub <- subset(ttest.results,Significance)
p <- ggplot2::ggplot(ttest.results,ggplot2::aes(Log2MedianChange,PValueMinusLog10))+ ggplot2::geom_point(ggplot2::aes(color=Significance)) 
p<-p + ggplot2::theme_bw(base_size=8) + ggplot2::geom_text(data=dsub,ggplot2::aes(label=RowGeneUniProtScorePeps),hjust=0, vjust=0,size=1) + ggplot2::scale_fill_gradient(low="white", high="darkblue") + ggplot2::xlab("Log2 Median Change") + ggplot2::ylab("-Log10 P-value")
#f=paste(file,proc.time()[3],".jpg")
ggplot2::ggsave(paste0(inpF,selection,sCol,eCol,comp,selThr,selThrFC,cvThr,"VolcanoTest.svg"), p)
print(p)
p <- ggplot2::ggplot(ttest.results,ggplot2::aes(Log2MedianChange,PValueMinusLog10))+ ggplot2::geom_point(ggplot2::aes(color=Significance)) 
p<-p + ggplot2::theme_bw(base_size=8) + ggplot2::geom_text(data=dsub,ggplot2::aes(label=""),hjust=0, vjust=0,size=1) + ggplot2::scale_fill_gradient(low="white", high="darkblue") + ggplot2::xlab("Log2 Median Change") + ggplot2::ylab("-Log10 P-value")
#f=paste(file,proc.time()[3],".jpg")
ggplot2::ggsave(paste0(inpF,selection,sCol,eCol,comp,selThr,selThrFC,cvThr,"VolcanoTestNoLabel.svg"), p)
print(p)
```


```{r M12any}
sel<-"M"
hda<-as.matrix(log2LFQ[,grep(sel,colnames(log2LFQ))])
summary(hda)
hist(as.matrix(hda))
hdaM1<-apply(hda[,grep("M1",colnames(hda))],1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{NA})
sum(hdaM1>thr,na.rm=T)
hdaM2<-apply(hda[,grep("M2",colnames(hda))],1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{NA})
sum(hdaM2>thr,na.rm=T)
hdaM<-cbind(hdaM1,hdaM2)
summary(hdaM==thr)
limma::vennDiagram(hdaM>thr)
inpW<-paste0(inpF,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(sel,hdaM)
```

```{r WTany}
sel<-"W"
hda<-as.matrix(log2LFQ[,grep(sel,colnames(log2LFQ))])
summary(hda)
hist(as.matrix(hda))
hdaM<-apply(hda,1,function(x) if(sum(is.na(x))<ncol(hda)){median(x,na.rm=T)} else{0})
sum(hdaM>thr,na.rm=T)
limma::vennDiagram(hdaM>thr)
inpW<-paste0(inpF,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(sel,hdaM)
```

```{r M12all}
sel<-"M"
hda<-as.matrix(log2LFQ[,grep(sel,colnames(log2LFQ))])
summary(hda)
hist(as.matrix(hda))
hdaM1<-apply(hda[,grep("M1",colnames(hda))],1,function(x) if(sum(is.na(x))==0){median(x,na.rm=T)} else{0})
sum(hdaM1>thr,na.rm=T)
hdaM2<-apply(hda[,grep("M2",colnames(hda))],1,function(x) if(sum(is.na(x))==0){median(x,na.rm=T)} else{0})
sum(hdaM2>thr,na.rm=T)
hdaM<-cbind(hdaM1,hdaM2)
limma::vennDiagram(hdaM>thr)
inpW<-paste0(inpF,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(sel,hdaM)
```

```{r WTall}
sel<-"W"
hda<-as.matrix(log2LFQ[,grep(sel,colnames(log2LFQ))])
summary(hda)
hist(as.matrix(hda))
hdaM<-apply(hda,1,function(x) if(sum(is.na(x))==0){median(x,na.rm=T)} else{0})
sum(hdaM>thr,na.rm=T)
limma::vennDiagram(hdaM>thr)
inpW<-paste0(inpF,sel,"LFQ.xlsx")
writexl::write_xlsx(as.data.frame(cbind(hdaM,hda,data$Fasta.headers)),inpW)
assign(sel,hdaM)
```

```{r venn}
hda<-cbind(M,W)
hda[is.na(hda)]=0
hist(hda)
limma::vennDiagram(hda>thr)
comb=data.frame(hda)
```

```{r writeXL}
xW<-comb[comb$hdaM1==0&comb$hdaM2==0&comb$W>0,]
writexl::write_xlsx(cbind(rownames(xW),xW),paste0(inpF,"W42.xlsx"))
xM<-comb[comb$hdaM1>0&comb$hdaM2>0&comb$W==0,]
writexl::write_xlsx(cbind(rownames(xM),xM),paste0(inpF,"M242.xlsx"))
```

```{r mito}
download.file(url="ftp://ftp.broadinstitute.org/distribution/metabolic/papers/Pagliarini/MitoCarta3.0/Human.MitoCarta3.0.xls",destfile = paste0(inpF,"mito.xlsx"))
mito<-readxl::read_xlsx(paste0(inpF,"mito.xlsx"))
```

```{r GOxM}
#https://github.com/Proteomicslab57357/UniprotR
#install.packages("UniprotR")
xM$Fasta <- rownames(xM)
xM$ID<-paste(sapply(strsplit(paste(sapply(strsplit(row.names(xM), "|",fixed=T), "[", 2)), "-"), "[", 1))
GeneOntologyObj <- UniprotR::GetProteinGOInfo(ID) 
GeneOntologyObj$ID <- rownames(GeneOntologyObj)
writexl::write_xlsx(GeneOntologyObj,paste0(inpF,"GeneOntologyObj.M242.xlsx"))
grep("glyco",GeneOntologyObj$Gene.ontology..biological.process.,ignore.case=T)
GeneOntologyObj$glyco <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("glyco",x,ignore.case=T))))
sum(GeneOntologyObj$glyco)
GeneOntologyObj$glycol <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("glycol",x,ignore.case=T))))
sum(GeneOntologyObj$glycol)
GeneOntologyObj$glycog <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("glycog",x,ignore.case=T))))
sum(GeneOntologyObj$glycog)
GeneOntologyObj$TCA <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("TCA",x,ignore.case=T))))
sum(GeneOntologyObj$TCA)
GeneOntologyObj$mitocho <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("mitocho",x,ignore.case=T))))
sum(GeneOntologyObj$mitocho)
combGO<-merge(xM,GeneOntologyObj,by="ID")
writexl::write_xlsx(combGO,paste0(inpF,"M242GO.xlsx"))
```

```{r GOxW}
#https://github.com/Proteomicslab57357/UniprotR
#install.packages("UniprotR")
xW$Fasta <- rownames(xW)
xW$ID<-paste(sapply(strsplit(paste(sapply(strsplit(row.names(xW), "|",fixed=T), "[", 2)), "-"), "[", 1))
GeneOntologyObj <- UniprotR::GetProteinGOInfo(xW$ID) 
GeneOntologyObj$ID <- rownames(GeneOntologyObj)
writexl::write_xlsx(GeneOntologyObj,paste0(inpF,"GeneOntologyObj.W42.xlsx"))
grep("glyco",GeneOntologyObj$Gene.ontology..biological.process.,ignore.case=T)
GeneOntologyObj$glyco <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("glyco",x,ignore.case=T))))
sum(GeneOntologyObj$glyco)
GeneOntologyObj$glycol <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("glycol",x,ignore.case=T))))
sum(GeneOntologyObj$glycol)
GeneOntologyObj$glycog <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("glycog",x,ignore.case=T))))
sum(GeneOntologyObj$glycog)
GeneOntologyObj$TCA <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("TCA",x,ignore.case=T))))
sum(GeneOntologyObj$TCA)
GeneOntologyObj$mitocho <- apply(GeneOntologyObj, 1, function(x)as.integer(any(grep("mitocho",x,ignore.case=T))))
sum(GeneOntologyObj$mitocho)
combGO<-merge(xW,GeneOntologyObj,by="ID")
writexl::write_xlsx(combGO,paste0(inpF,"W42GO.xlsx"))
```

