```{r library}
#BiocManager::install("BSgenome")
library("BSgenome")
#http://bioconductor.org/packages/release/bioc/vignettes/MutationalPatterns/inst/doc/Introduction_to_MutationalPatterns.pdf
#BiocManager::install("MutationalPatterns")
library("MutationalPatterns")
head(available.genomes())
#https://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.NCBI.GRCh38.html
#BiocManager::install("BSgenome.Hsapiens.NCBI.GRCh38")
refGenome<-"BSgenome.Hsapiens.NCBI.GRCh38"
library(refGenome, character.only = TRUE)
save.image("F:/HeLa/vcfR.session.RData")
```

```{r data}
#inpD<-"F:/HeLa/deepvariant/bam/vcf"
inpD<-"F:/HeLa/vcf/"
#vcfFiles <- list.files(inpD, pattern = "R\\*\\.vcf$", full.names = TRUE)
vcfFiles <- Sys.glob(file.path(inpD, "SR*.vcf")) 
#rownames(sampleNames)<-gsub("/"," ",rownames(vcfFiles))
sampleNames<-gsub("/","_",vcfFiles)
sampleNames<-gsub(":","_",sampleNames)
vcfs <- read_vcfs_as_granges(vcfFiles, sampleNames, refGenome)
#warnings()#->email
summary(vcfs)
```

```{r signature}
muts = mutations_from_vcf(vcfs[[1]])
head(muts, 12)
types = mut_type(vcfs[[1]])
head(types, 12)
context = mut_context(vcfs[[1]], refGenome)
head(context, 12)
typeContext = type_context(vcfs[[1]], refGenome)
lapply(typeContext, head, 12)
typeOccurrences <- mut_type_occurrences(vcfs, refGenome)
```

```{r plot}
plot_spectrum(typeOccurrences)
p1 <- plot_spectrum(typeOccurrences)
p2 <- plot_spectrum(typeOccurrences, CT = TRUE)
p3 <- plot_spectrum(typeOccurrences, CT = TRUE, legend = FALSE)
dataSel<-as.matrix(typeOccurrences["",grep("SRR2185899",colnames(typeOccurrences))])
#colnames(dataSel)<-sub("F:/HeLa/deepvariant/bam/vcf/","",colnames(dataSel))
#colnames(dataSel)
summary(dataSel)
```

```{r grid}
library("gridExtra")
grid.arrange(p1, p2, p3, ncol=3, widths=c(3,3,1.75))
p4 <- plot_spectrum(typeOccurrences, CT = TRUE, legend = TRUE)
palette <- c("pink", "orange", "blue", "lightblue", "green", "red", "purple")
p5 <- plot_spectrum(typeOccurrences, CT=TRUE, legend=TRUE, colors=palette)
grid.arrange(p4, p5, ncol=2, widths=c(4,2.3))
```

```{r prof96}
#Make a 96 trinucleodide mutation count matrix: C>A C>G C>T T>A T>C T>G
mutMat <- mut_matrix(vcf_list = vcfs, ref_genome = refGenome)
head(mutMat)
#plot_96_profile(typeOccurrences)
plot_96_profile(mutMat[,c(1,7)])
plot_96_profile(mutMat[,c(1,7)], condensed = TRUE)
```

```{r estimate, echo = FALSE}
#https://www.nature.com/articles/s41467-019-11037-8#code-availability
mutMatEst = mutMat + 0.0001 
estimate = nmf(mutMatEst, rank=2:6, method="brunet", nrun=1000, seed=123456)  
plot(estimate)
```

```{r cosmic, echo = FALSE}
#wget https://cancer.sanger.ac.uk/cancergenome/assets/signatures_probabilities.txt
#sp_url <- paste("http://cancer.sanger.ac.uk/cancergenome/assets/", "signatures_probabilities.txt", sep = "")
sp_url <- "F:/HeLa/signatures_probabilities.txt"
cancer_signatures = read.table(sp_url, sep = "\t", header = TRUE) 
new_order = match(row.names(mutMat), cancer_signatures$Somatic.Mutation.Type) 
cancer_signatures_ro = cancer_signatures[as.vector(new_order),]  
row.names(cancer_signatures_ro) = cancer_signatures_ro$Somatic.Mutation.Type  
cancer_signatures_ro = as.matrix(cancer_signatures_ro[,4:33])
```

```{r mutation_cosmic_overlap, echo = FALSE}
mutMatRO<-mutMat
mutMatRO[,1:ncol(mutMatRO)] = apply(mutMatRO[,1:ncol(mutMatRO)], 2, function(x) as.numeric(as.character(x)))
cancer_signatures_ro_mm<-cancer_signatures_ro
cancer_signatures_ro_mm[,1:ncol(cancer_signatures_ro_mm)] = apply(cancer_signatures_ro_mm[,1:ncol(cancer_signatures_ro_mm)], 2, function(x) as.numeric(as.character(x)))  
fit_res <- fit_to_signatures(mutMatRO, as.matrix(cancer_signatures_ro_mm))
plot_contribution(fit_res$contribution,cancer_signatures_ro_mm[,1:ncol(cancer_signatures_ro_mm)],coord_flip = FALSE,mode = "relative")
```
 
```{r t-test, echo = FALSE}
dataSellog2grpTtest<-dataSellog2[]
pValNA = apply(dataSellog2grpTtest, 1, function(x) if(sum(!is.na(x[c(1:6)]))<2){NaN} else{t.test(as.numeric(x[c(1:6)]),na.rm=T)$p.value})
hist(pValNA)
pVal1 = apply(dataSellog2grpTtest, 1, function(x) if(sum(!is.na(x[c(1:6)]))<2){1} else{t.test(as.numeric(x[c(1:6)]),na.rm=T)$p.value})
hist(pVal1)
pValBHna = p.adjust(pValNA,method = "BH")
hist(pValBHna)
pValBH1 = p.adjust(pVal1,method = "BH")
hist(pValBH1)
plot(pValBH1-pValBHna,pVal1-pValNA)
logFC = rowMeans(dataSellog2grpTtest[,c(1:6)])#-rowMeans(yyt[,c(6,5,4,3,2,1)])
hist(logFC)
plot(logFC,-log10(pVal1))
plot(logFC,-log10(pValNA))
plot(logFC,-log10(pValBHna),col="orange")

logFCmedian = matrixStats::rowMedians(dataSellog2grpTtest[,c(1:6)])#-rowMeans(yyt[,c(6,5,4,3,2,1)])
hist(logFCmedian)

ttest.results = data.frame(logFC=logFC,logFCmedian=logFCmedian,P.Value = pValNA, adj.pval.BH = pValBHna) 
#ttest.results$PSMcount = psm.count.table[ttest.results$gene,"count"]
#ttest.results = ttest.results[with(ttest.results, order(P.Value)), ]
#head(ttest.results)
write.csv(ttest.results,file=paste0(inpF,hdr,"tTestBHgrp.csv"))
```

```{r volcanoPlot}
library(ggplot2)
ttest.results<-ttest.results7
dsub=subset(ttest.results,ttest.results$P.Value<0.05&abs(ttest.results$logFC)>0.58)
rn<-strsplit(rownames(dsub), ';')
row.names(dsub) <- sapply(rn, "[", 1)#rn[[1]]
g = ggplot(ttest.results,aes(logFC,-log10(P.Value)))+geom_point(aes(color=adj.pval.BH),size=0.15) + theme_bw(base_size=10) +geom_text(data=dsub,aes(label=row.names(dsub)), vjust=0.5, size=1.5) + xlab("Log2 Fold Change (NRF2/sNRF2NT)")  + ylab("-Log10 P-value") + ggtitle("Differentially Expressed Protein groups") + scale_size_area()+scale_color_gradient(low="#FF9933", high="#99CC66")
plot(g)
#install.packages("svglite")
library(svglite)
ggsave(file=paste0(inpF,hdr,"volcanoPlotgrp7.svg"),plot=g)#,  width=6, height=6)
```


```{r remNA}
#yyt[is.na(yyt)]<-1 #for equal ratio
dataSellog2remNA<-dataSellog2
dataSellog2remNA[is.na(dataSellog2remNA)]<-0
#y<-yyt[,-(which(colSums(yyt) == 0))] 
dataSellog2remNA<-dataSellog2remNA[-(which(rowSums(dataSellog2remNA) == 0)),] 
summary(dataSellog2remNA)
#yyt<-yyt[,rownames(label)]
#colnames(dataSellog2remNA)
#yyt[,9]=as.factor(yyt[,9])
```

```{r heatmap}
#install.packages('pheatmap')
library(pheatmap)
#?pheatmap
svgPHC<-pheatmap(yyt,scale="row",clustering_distance_rows = "correlation",clustering_distance_cols = "correlation",fontsize_row=6,annotation_col = label,show_rownames=F)
#ggsave(file=paste0(inpF,"corrcoefED.svg"), plot=svgPHC, width=6, height=6)
plot(svgPHC)
```


```{r scale}
#install.packages('quantable')
library(quantable)
y=robustscale(yyt)
y$data[is.na(y$data)]<-0
y$data<-y$data[-(which(rowSums(y$data) == 0)),] 
summary(y$data)
#names(y$data)=sub("X","",names(y$data))
pheatmap(y$data,scale="row",clustering_distance_rows = "manhattan",clustering_distance_cols = "manhattan",fontsize_row=6,annotation_col = label,show_rownames=F)
```

```{r PCA}
library(ggfortify)
#log.yyt=log(yyt[,1:8]+1)
#yyt.pca=prcomp(log.yyt,center=TRUE,scale.=TRUE) 
yyt.pca=prcomp(t(yyt),center=TRUE,scale.=TRUE) 
autoplot(yyt.pca,data=t(yyt),colour=as.numeric(label$Location))
autoplot(yyt.pca,data=t(yyt),colour=as.numeric(label$Treatment))
#biplot(yyt.pca)
```


```{r DEqMS-peptides}
fit3$count = rowMins(as.matrix(df.prot[,grepl("^Peptides\\.[0-9]+", names(df.prot))]))+1
min(fit3$count)

fit4 = spectraCounteBayes(fit3)
VarianceBoxplot(fit4, n=20, main = hdr,xlab="peptide count + 1")
DEqMS.results = outputResult(fit4,coef_col = 1)
DEqMS.results$Gene.name = df.prot[DEqMS.results$gene,]$Gene.names
head(DEqMS.results)
write.table(DEqMS.results,paste0(inpD,hdr,"R-W.DEqMS.pep.results.txt"),sep = "\t",row.names = F,quote=F)
head(DEqMS.results)
VarianceBoxplot(fit4,n=20)
#peptideProfilePlot(dat=df.prot)#,col=2,gene="TGFBR2")
VarianceScatterplot(fit4, xlab="log2(LFQ)")
limma.prior = fit4$s2.prior
abline(h = log(limma.prior),col="green",lwd=3 )
legend("topright",legend=c("DEqMS prior variance","Limma prior variance"),
        col=c("red","green"),lwd=3)
op <- par(mfrow=c(1,2), mar=c(4,4,4,1), oma=c(0.5,0.5,0.5,0))
Residualplot(fit4,  xlab="log2(PSM count)",main="DEqMS")
x = fit3$count
y = log(limma.prior) - log(fit3$sigma^2)
plot(log2(x),y,ylim=c(-6,2),ylab="Variance(estimated-observed)", pch=20, cex=0.5,
     xlab = "log2(PSMcount)",main="Limma")
#install.packages("LSD")
library(LSD)
op <- par(mfrow=c(1,2), mar=c(4,4,4,1), oma=c(0.5,0.5,0.5,0))
x = fit3$count
y = fit3$s2.post
heatscatter(log2(x),log(y),pch=20, xlab = "log2(PSMcount)", 
     ylab="log(Variance)",
     main="Posterior Variance in Limma")

y = fit4$sca.postvar
heatscatter(log2(x),log(y),pch=20, xlab = "log2(PSMcount)",
     ylab="log(Variance)", 
     main="Posterior Variance in DEqMS")

```


```{r randomtree}
set.seed(42)
yy<-rbind(y$data,label$Treatment)
yyt<-t(yy)
summary(yyt)

yyt=as.data.frame(yyt)
colnames(yyt) <- gsub(";", "_", colnames(yyt))
colnames(yyt) <- gsub(":", "__", colnames(yyt))
colnames(yyt) <- gsub("-", "___", colnames(yyt))
summary(yyt)
ind=sample(2,nrow(yyt),replace=TRUE,prob=c(0.75,0.25))
yyt.training=yyt[ind==1,]
yyt.test=yyt[ind==2,]

library("rpart")
library("rpart.plot")
tree=rpart(data=yyt.training,V4500~.,method="class",control=rpart.control(minsplit=10,minbucket=5),parms=list(split="information"))
rpart.plot(tree,main="Classification tree for the yyt data (using 75% of data as training set)",extra=101)


library(randomForest)
set.seed(42)
#yyt=as.data.frame(yyt)
random_forest=randomForest(data=yyt.training,V4500~.,impurity='gini',ntree=200,replace=TRUE)
print(random_forest)

plot(random_forest)
legend("top",cex=0.8,legend=colnames(random_forest$err.rate),lty=c(1,2,3),col=c(1,2,3),horiz=T)

predictions=predict(random_forest,newdata=yyt.training,type="class")
actuals=yyt.training$V4500
table(actuals,predictions)

accuracy=sum(diag(confusion.matrix))/sum(confusion.matrix)
print(accuracy)

sort(importance(random_forest))
varImpPlot(random_forest)
#qplot(RCN1,GALNT2,data=yyt,colour=class,size=I(3))
```

```{r clean, echo = FALSE}
decoyPrefix="REV__"
dfNoRev<-data[-grep(decoyPrefix, rownames(data)),]
dataClean = data[!data$Reverse=="+",]
setdiff(rownames(dataClean),rownames(dfNoRev))
setdiff(rownames(dfNoRev),rownames(dataClean))
summary(dataClean)
```

