```{r data}
%halpdt=data.frame("F:/promec/Qexactive/Linda/raw/PD2/_LINDA_/130405_LINDA_gel-(05)_TargetProtein-HALPDTHighConfGNcompGO.txt")
halpdt=read.delim("F:/promec/Qexactive/Linda/raw/PD2/_LINDA_/130405_LINDA_gel-(05)_TargetProtein-HALPDTHighConfGNcompGO.txt")
summary(halpdt)
gene=as.character(halpdt$T..T..T..FASTA.Title.Lines_)
protein=as.character(halpdt$T..T..Uniprot)
hist(log2(halpdt$T..AreaChkNew))
brapa=read.delim("F:/promec/Results/Ishita//Copy of Log2abs0.5 GO.txt",row.names=1)
brapa=read.csv("F:/promec/Results/Ishita//chk.csv",row.names=1)
```

```{r DAPAR}
source("https://bioconductor.org/biocLite.R")
biocLite("DAPAR")
??DAPAR
#http://bioinformatics.oxfordjournals.org/content/early/2016/09/06/bioinformatics.btw580.short?rss=1
```

```{r MSstats}
source("http://bioconductor.org/biocLite.R")
biocLite("MSstats")
library(MSstats)
#http://msstats.org/msstatsqc/
```

```{r pep-xml}
source("https://bioconductor.org/biocLite.R")
biocLite("pepXMLTab")
library(pepXMLTab)
pepxml <- system.file("extdata/pepxml", "SEQUEST.pepXML", package="pepXMLTab")
pepxml <- "F:/promec/Search/130904_SS+ANBL6_run1.pep.closed.xml"
tttt <- pepXML2tab(pepxml)
tttt[1:2,]
passed <- PSMfilter(tttt, pepFDR=0.01, scorecolumn='xcorr', hitrank=1,minpeplen=6, decoyprefix='rev_')
passed[1, ]
hist(as.double(tttt[,16]))

```

```{r sign-test}
x <- 10; y <- 10; g <- 10000
set.seed(2016)
dat <- matrix(rnorm((x + y) * g), ncol = x + y)
results <- apply(dat, 1, function(dat) {
  wilcox.test(x = dat[1:x], y = dat[(x + 1):(x + y)])$p.value})
cbind(dat, pvals = results)
hist(p.adjust(results,method="BH"))
hist(p.adjust(results,method="BH"))
hist(p.adjust(results,method="holm"))
?p.adjust
hist(results)
min(results)
data=read.delim("C:/Users//animeshs/Google Drive/wilcox_sgn_rnk_wiki.txt",row.names=1)
wilcox.test(data$x2,data$x1,paired = TRUE, alternative = "less")
wilcox.test(data$x2,data$x1,paired = TRUE, alternative = "greater")
wilcox.test(data$x2,data$x1,paired = TRUE)
wilcox.test(c(1,1,1,1,1))
0.5^4
wilcox.test((data$x2-data$x1))$pvalue
binom.test(!is.(data$sgn>0),sum(data$sgn)))
sum(data$sgn>0, na.rm=TRUE)
binom.test(1,3)
t.test(c(data$x1,data$x2),c(data$abs,data$abs))
t.test(extra ~ group, data = sleep)
wilcox.test(extra ~ group, data = sleep)

lapply(matrix(rnorm(10000), 1000),matrix(rnorm(10000), 1000),t.test)$p.value
matrix(rnorm(( + y) * g), ncol = x + y)

x <- 10; y <- 10; g <- 100000
mean(c(1,2,3))
set.seed(1969)
dat <- matrix(rnorm((x + y) * g), ncol = x + y)

results <- apply(dat, 1, function(dat) {
#  t.test(x = dat[1:x], y = dat[(x + 1):(x + y)])$p.value})
  wilcox.test(x = dat[1:x], y = dat[(x + 1):(x + y)])$p.value})

cbind(dat, pvals = results)
hist(results,col="orange")
hist(p.adjust(results,method="BH"))

write.csv2("C:/Users//animeshs/Google Drive/wilcox_sgn_rnk_wiki_test.txt",results)

```

```{r graphite}
source("https://bioconductor.org/biocLite.R")
biocLite("graphite")
### R code from vignette source 'graphite.Rnw'
### Encoding: UTF-8

###################################################
### code chunk number 1: style
###################################################
BiocStyle::latex()


###################################################
### code chunk number 2: startup
###################################################
library(graph)
library(graphite)


###################################################
### code chunk number 3: base1
###################################################
humanReactome <- pathways("hsapiens", "reactome")
names(humanReactome)[1:10]
p <- humanReactome[["ABC-family proteins mediated transport"]]
p


###################################################
### code chunk number 4: base2
###################################################
p <- humanReactome[[21]]
p@title


###################################################
### code chunk number 5: base3
###################################################
head(nodes(p))


###################################################
### code chunk number 6: base4
###################################################
head(edges(p))


###################################################
### code chunk number 7: base5
###################################################
pathwayDatabases()


###################################################
### code chunk number 8: graph1
###################################################
g <- pathwayGraph(p)
g


###################################################
### code chunk number 9: graph2
###################################################
edgeData(g)[1]


###################################################
### code chunk number 10: ident1
###################################################
biocLite('org.Hs.eg.db')
pSymbol <- convertIdentifiers(p, "SYMBOL")
pSymbol
head(nodes(pSymbol))


###################################################
### code chunk number 11: ident2
###################################################
reactomeSymbol <- convertIdentifiers(humanReactome[1:5], "SYMBOL")


###################################################
### code chunk number 12: spia1
###################################################
biocLite('SPIA')
library('SPIA')
data(colorectalcancer)

biocLite('hgu133plus2.db')
library(hgu133plus2.db)
x <- hgu133plus2ENTREZID
top$ENTREZ <- unlist(as.list(x[top$ID]))
top <- top[!is.na(top$ENTREZ), ]
top <- top[!duplicated(top$ENTREZ), ]
tg1 <- top[top$adj.P.Val < 0.05, ]

DE_Colorectal = tg1$logFC
names(DE_Colorectal) <- as.vector(tg1$ENTREZ)
ALL_Colorectal <- top$ENTREZ

prepareSPIA(biocarta[1:2], "biocartaEx")
res <- runSPIA(de=DE_Colorectal, all=ALL_Colorectal, "biocartaEx")
res


###################################################
### code chunk number 13: dg1
###################################################
biocLite('DEGraph')
library(DEGraph)
data("Loi2008_DEGraphVignette")

biocarta <- pathways("hsapiens", "biocarta")
p <- convertIdentifiers(biocarta[["actions of nitric oxide in the heart"]], "ENTREZID")
res <- runDEGraph(p, exprLoi2008, classLoi2008)
res$`1`


###################################################
### code chunk number 14: tg1
###################################################
biocLite('topologyGSA')
library(topologyGSA)
data(examples)

kegg <- pathways("hsapiens", "kegg")
p <- convertIdentifiers(kegg[["Fc epsilon RI signaling pathway"]], "SYMBOL")
runTopologyGSA(p, "var", y1, y2, 0.05)


###################################################
### code chunk number 15: tg2
###################################################
biocLite('clipper')
library(clipper)
biocLite('ALL')
library(ALL)

kegg <- pathways("hsapiens", "kegg")
paths <-convertIdentifiers(kegg[1:5], "ENTREZID")
genes <- unlist(lapply(paths, nodes))
data(ALL)
all <- as.matrix(exprs(ALL[1:length(genes),1:20]))
classes <- c(rep(1,10), rep(2,10))
rownames(all) <- genes
clipped <- runClipper(paths, all, classes, "mean", pathThr=0.1)
resClip <- do.call(rbind,clipped$results)
resClip[,1:5]

```

install.packages(c("devtools"))
source("http://www.bioconductor.org/biocLite.R")
biocLite(c("Biobase","preprocessCore"))
library(preprocessCore)
mat <- matrix(c(5,2,3,4,4,1,4,2,3,4,6,8),ncol=3)
normalize.quantiles(mat)



```{r map}
d3heatmap(brapa, scale = "column")
install.packages("d3heatmap")
library("d3heatmap")
d3heatmap(brapa, scale = "column")
d3heatmap(brapa, scale = "column", dendrogram = "none", color = "Blues")
d3heatmap(brapa, colors = "Blues", scale = "col",  dendrogram = "row", k_row = 3)
d3heatmap(brapa, scale = "column", dendrogram = "none",   color = scales::col_quantile("Blues", NULL, 5))
#install.packages("installr")
#library(installr)
#updateR()
#https://github.com/santhilalsubhash/geneSCF bash commandline tool
#library(org.Hs.eg.db)
#egmap <- revmap(org.Hs.egUNIPROT)[as.character(gene)]
source("https://bioconductor.org/biocLite.R")
#biocLite("BiocUpgrade")
biocLite("UniProt.ws")
library(UniProt.ws)
up <- UniProt.ws(taxId=9606)
egs = keys(up, "ENTREZ_GENE")
setInternet2(TRUE)

biocLite('mygene')
biocLite('IRanges')
library(mygene)
glist <- c('UNG','xrcc1')
glterm <- queryMany(glist, scopes='symbol', fields=c('entrezgene', 'go'), species='human')
glterm[1, 'go.BP'][[1]]

```


```{r install libraries and setup}
#setwd('C:/Users/animeshs/SkyDrive')
#source("http://bioconductor.org/biocLite.R")
#biocLite("GSA")
#biocLite("BiocUpgrade")
#install.packages("pvclust")
#install.packages("knitr")
#rm(list = ls())
#dev.off() 
#reporting
#library("knitr")
#clustering
```


```{r GSA}
library("GSA")
?GSA
######### two class unpaired comparison
# y must take values 1,2

set.seed(100)
x<-matrix(rnorm(1000*20),ncol=20)
dd<-sample(1:1000,size=100)

u<-matrix(2*rnorm(100),ncol=10,nrow=100)
x[dd,11:20]<-x[dd,11:20]+u
y<-c(rep(1,10),rep(2,10))


genenames=paste("g",1:1000,sep="")

#create some random gene sets
genesets=vector("list",50)
for(i in 1:50){
 genesets[[i]]=paste("g",sample(1:1000,size=30),sep="")
}
geneset.names=paste("set",as.character(1:50),sep="")

GSA.obj<-GSA(x,y, genenames=genenames, genesets=genesets,  resp.type="Two class unpaired", nperms=100)
GSA.listsets(GSA.obj, geneset.names=geneset.names,FDRcut=.5)

```


```{r pathview}
library(gage)
library(pathview)
# with MM20CL14 SS data over Hedgehog signaling pathway
hda=readExpData("L:/Elite/LARS/2015/january/Ishita/TestPthviewFA.txt",row.names=1)
hda=as.matrix(hda)
#cancer related hsa05200
?pathview
pathview(hda,pathway.id="ko00061 ",cpd.idtype="kegg",out.suffix="FA")
# http://www.genome.jp/kegg/pathway/map/map00020.html
pathview(hda,pathway.id="hsa00020",gene.idtype="UNIPROT",out.suffix="TCA")
# http://www.genome.jp/kegg/pathway/map/map00030.html
pathview(hda,pathway.id="hsa00030",gene.idtype="UNIPROT",out.suffix="PPP")
# http://www.genome.jp/kegg/pathway/map/map00190.html 
pathview(hda,pathway.id="hsa00190",gene.idtype="UNIPROT",out.suffix="OxPhos")
# http://www.genome.jp/kegg/pathway/map/map00010.html
pathview(hda,pathway.id="hsa00010",gene.idtype="UNIPROT",out.suffix="Glycolysis")

```

```{r check mapping significance}
data=nrow(hda)
dm=10
km=100
kegg=60
#GOmap <-  matrix(c(data-dm, dm,kegg-dm, km), nrow = 2, dimnames = list(S = c("+", "-"), B = c("+", "-")))
GOmap <-  matrix(c(7,9,73, 2034), nrow = 2, dimnames = list(S = c("+", "-"), B = c("+", "-"))) # from ipaAdvMan. C:\Users\animeshs\Google Drive\IPA Training Data
fisher.test(GOmap, alternative = "greater")
GOmap <-  matrix(c(1,2,1,30), nrow = 2, dimnames = list(S = c("+", "-"), B = c("+", "-")))
fisher.test(GOmap) #(2*factorial(32)*factorial(3)*factorial(31))/(2*factorial(30)*factorial(34))
x <- rnorm(1000)
hist(x)
p <- 2*pnorm(sort(-abs(x)))
p.adjust(p, method = "BH")

```

```{r cluster bootstrap}
library("pvclust")
result <- pvclust(hda, method.dist="cor", method.hclust="average", nboot=10)
plot(result)
pvrect(result, alpha=0.95)
```

```{r check pathview for base excision repair pathway}

# PDT over KEGG Apoptosis
hda=readExpData("L:/Qexactive/Linda/TW283.txt",row.names=1)
hda=as.matrix(hda)
pathview(hda,pathway.id="hsa04210",gene.idtype="UNIPROT")
biocLite("goProfiles")
library(goProfiles)
basicProfile(hda, onto = "CC",level = 2, orgPackage = "org.Hs.eg.db")
basicProfile(hda, onto = "MF",level = 3, orgPackage = "org.Hs.eg.db")
basicProfile(hda, onto = "BP",level = 4, orgPackage = "org.Hs.eg.db")
biocLite("GeneAnswers")
library('GeneAnswers')
humga<-geneAnswersBuilder(revmap(org.Hs.egUNIPROT['P13051'])
, "org.Hs.eg.db", categoryType='GO.BP', testType='hyperG',  pvalueT=0.1, FDR.correction=TRUE, geneExpressionProfile=hda)
geneAnswersReadable(humga)
uniprot <- c("A0A183", "A0A5E8", "A0AUZ9","A0AV02","123")

gmap <- revmap(org.Hs.egUNIPROT)[uniprot]
mappedLkeys(gmap)

genes = mget(revmap(org.Hs.egUNIPROT)[uniprot], ifnotfound=NA)
biocLite("AnnotationFuncs")
library(AnnotationFuncs)
translate(rownames(hda),org.Hs.egUNIPROT,return.list=FALSE)
translate(c("A0A183", "A0A5E8", "A0AUZ9","A0AV02"), org.Hs.egUNIPROT, org.Hs.egGENENAME)
 warnings()

hda=readExpData("C:/Users/animeshs/OneDrive/SHBER.txt",row.names=1)
hda=as.matrix(hda)
hda.d=hda[,1:3]-hda[,4:6]
summary(hda.d)
pv.out<-pathview(hda.d,pathway.id="hsa03410",gene.idtype="UNIPROT", limit = list(gene = 5, cpd = 1), out.suffix=proc.time())
str(pv.out)
head(pv.out$plot.data.gene)


# test on http://www.nature.com/nri/journal/v5/n5/fig_tab/nri1604_F1.html 
pathview(hda.d,pathway.id="hsa04630",gene.idtype="UNIPROT",out.suffix="jak")
pathview(hda.d,pathway.id="hsa00020",gene.idtype="UNIPROT",out.suffix="joxphos")



# Trasfection
hda=readExpData("L:/Elite/LARS/2014/mai/transfection 3rd paralell/ListPV1in20FC2.txt",row.names=1)
hda=as.matrix(hda)
# herpes simplex
pathview(hda,pathway.id="hsa03410",gene.idtype="UNIPROT")
pathview(hda,pathway.id="hsa05168",gene.idtype="UNIPROT")


library("org.Hs.eg.db")
frame = toTable(org.Hs.egGO)
goframeData = data.frame(frame$go_id, frame$Evidence, frame$gene_id)
head(goframeData)
goFrame=GOFrame(goframeData,organism="Homo sapiens")
goAllFrame=GOAllFrame(goFrame)
library("GSEABase")
gsc <- GeneSetCollection(goAllFrame, setType = GOCollection())
library("GOstats")
universe = Lkeys(org.Hs.egGO)
genes = universe[1:500]
params <- GSEAGOHyperGParams(name="My Custom GSEA based annot Params",
                              geneSetCollection=gsc,
                              geneIds = genes,
                              universeGeneIds = universe,
                              ontology = "MF",
                              pvalueCutoff = 0.05,
                              conditional = FALSE,
                              testDirection = "over")
Over <- hyperGTest(params)
head(summary(Over))


pv.out <- pathview(hda.d,pathway.id="hsa03410",gene.idtype="UNIPROT", limit = list(gene = 5, cpd = 1), out.suffix="fcnn", kegg.native = TRUE)
str(pv.out)
head(pv.out$plot.data.gene)


## MS to SNP

source("http://bioconductor.org/biocLite.R")
biocLite("sapFinder")
library("sapFinder")
browseVignettes("sapFinder")
vcf <- system.file("extdata/sapFinder_test.vcf",
                   package="sapFinder")
annotation <- system.file("extdata/sapFinder_test_ensGene.txt",
                          package="sapFinder")
refseq <- system.file("extdata/sapFinder_test_ensGeneMrna.fa",
                      package="sapFinder")
outdir <- "db_dir"
prefix <- "sapFinder_test"
db.files <- dbCreator(vcf=vcf, annotation=annotation,
                      refseq=refseq, outdir=outdir,prefix=prefix)


#DEMO

#load data

#KEGG view: gene data only
i <- 1
pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id =
                     demo.paths$sel.paths[i], species = "hsa", out.suffix = "gse16873",
                   kegg.native = TRUE)
str(pv.out)
head(pv.out$plot.data.gene)
#result PNG file in current directory

#Graphviz view: gene data only
data(gse16873.d)
data(demo.paths)
pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id = demo.paths$sel.paths[i], species = "hsa", out.suffix = "gse16873",kegg.native = FALSE, sign.pos = demo.paths$spos[i])
#result PDF file in current directory

png('test.png')
plot(sin(seq(-10,10,0.1)),type = "l", cex = .1, col = "dark red")
dev.off()

?plot

# from Luo
hda=as.matrix(hda)
hda.d=hda[,1:3]-hda[,4:6]
pathview(hda.d,pathway.id="hsa03410",gene.idtype="UNIPROT", limit = list(gene = 5, cpd = 1), out.suffix="fc")

## PCA
data=read.csv("L:/Elite/gaute/test/SHBER.csv")
gene=as.character(data[,1])
data=data[,-1]
prco=prcomp(data)
sco <- prco$x
loa <- prco$rotation
plot(sco[,1], sco[,2], col=row.names(data))
plot(loa[,1], loa[,2], col=row.names(data))
     

library(org.Hs.eg.db)
egmap <- revmap(org.Hs.egUNIPROT)[as.character(gene)]

x <- org.Hs.egGO
mapped_genes <- mappedkeys(x)
xx <- as.list(x[mapped_genes])
goids <- xx[2:3]
names(goids[[1]])

mod <- prcomp(E)

## scores (sample space)
scores <- mod$x
## loadings (gene space)
loads <- mod$rotation

## gene symbols
sym <- as.character(mget(colnames(E), illuminaHumanv3SYMBOL))

# plot
pc1 = 1 # ok, checked
pc2 = 2
# score plot (sample space)
x11()
expVar <- 100 *mod$sdev / sum(mod$sdev) 
hist(expVar)
plot(scores[,pc1], scores[,pc2], col=cl, pch=19, main="Score-plot (Sample map)", xlab=paste("PC:", expVar[pc1]), ylab=paste("PC:", expVar[pc2]))
text(scores[,pc1], scores[,pc2], rownames(E), pos=3)


# loadings (gene space)
plot.factor.name <- "lrT - T"
col <- factor(a[,plot.factor.name])
plot(loads[,pc1], loads[,pc2], col=col, pch='.', main="Gene weights", xlab=paste("PC:", expVar[pc1]), ylab=paste("PC:", expVar[pc2]))
text(loads[,pc1], loads[,pc2], sym, pos=3, cex=0.4)



# extract a quad 
kk <- loads[,pc1] > 0 & loads[,pc2] > 0
plot(kk)


# GO mapping
x <- org.Hs.egGO
mapped_genes <- mappedkeys(x)
xx <- as.list(x[mapped_genes])
goids <- xx[2:3]
names(goids[[1]])
```
