```{r data}
#setwd('L:/Elite/LARS/2014/desember/christiano/serum/')
#file='W2H.txt'
#setwd('L:/Elite/LARS/2014/desember/christiano/plantaris/')
#file='W2C.txt'
#data <- read.table(file, sep = '\t', header = T)
#data <- read.table('L:/Elite/LARS/2014/desember/christiano/serum/W2C.txt', sep = '\t', header = T)
#data <- read.csv('L:/Elite/LARS/2015/january/Ishita/Copy of LFQmedTtestBHcorr.csv', header = T)
#data <- read.table('L:/Elite/LARS/2015/january/Ishita/Copy of LFQmedTtestBHcorr.txt', sep = '\t', header = T)
data <- read.table('L:/Elite/LARS/2015/january/Ishita/uniprot-gomap-expression.txt', sep = '\t', header = T)
data <- read.table('L:/Elite/Celine/Alex/resultdmsocombo/SSandLFQttestPB.txt', sep = '\t', header = T)
"for i in  /cygdrive/f/promec/Elite/kamila/2015-08-08/*.txt ; do awk -F '|' '{print $2}' $i > $i.IDs.txt ; done"

```

```{r Laccase}
source("http://bioconductor.org/biocLite.R")
biocLite("XMLRPC")
library("XMLRPC")
server = "http://localhost:1234"
xml.rpc(server,"PathVisio.importData", "/home/anwesha/combidata.txt","/home/anwesha/PathVisio-Data/gene-databases","/home/anwesha/")
 xml.rpc(server, "PathVisio.createVisualization","/home/anwesha/combidata.txt.pgex","log2FC","blue,white,red","-2,0,2","pvalue","green","[pvalue]<0.05")
xml.rpc(server,"PathVisio.visualizeDataByURI", "WP1","/home/anwesha/combidata.txt.pgex","/home/anwesha/PathVisio-Data/gene-databases","/home/bohler/PathVisio-Data/results")
xml.rpc(server,"PathVisio.createPathway","Test Pathway","Test","Homo Sapiens",".")
xml.rpc(server,"PathVisio.addDataNode", "Test Pathway.gpml", "Glucose","Metabolite"," HMDB00122","HMDB")

xml.rpc(server,"PathVisio.addDataNode", "Test Pathway.gpml", "Glucose-6-P","Metabolite"," HMDB01401","HMDB")
xml.rpc(server,"PathVisio.addInteraction", "Test Pathway.gpml","Glucose->Glucose-6-P","Glucose","Glucose-6-P","","mim-conversion")

```

```{r Laccase}
library('ggplot2')
data <- read.table('F:/promec/Qexactive/LARS/2014/desember/Rami_Morteza/Results/LFQ-pellet-supernatant-RPKM-genome-plasmidPlot.txt', sep = '\t', header = T)
summary(data)
data[1,]
Significance=abs(data$PelletLog2DiffMedian)>1&abs(data$SupernatantLog2DiffMedian)>1&abs(data$RPKMlog2Diff)>1
sum(Significance)
dsub <- subset(data, Significance==TRUE)
plot(data$RPKMlog2Diff,data$SupernatantLog2DiffMedian)

g = ggplot(data=data)+geom_point(aes(RPKMlog2Diff,SupernatantLog2DiffMedian), colour='light green')  + geom_point(aes(RPKMlog2Diff,PelletLog2DiffMedian), colour='orange') + geom_text(data=dsub,aes(RPKMlog2Diff,SupernatantLog2DiffMedian,label=func),hjust=0.5, vjust=-0.5, size=3) 
#g = ggplot(data=data)+geom_point(aes(RPKMlog2Diff,SupernatantLog2DiffMedian), colour='light green')  + geom_point(aes(RPKMlog2Diff,PelletLog2DiffMedian), colour='orange') 
gps<- g  + theme_bw(base_size=10)  + xlab("Transcriptome")  + ylab("Proteome")  + ggtitle("Log2 median ratio change [+/- Laccase Inhibitor]")  + scale_size_area()  #+ geom_text(data=dsub,aes(label=func),hjust=0.5, vjust=-0.5, size=3) 
print(gps)

```

```{r RT}
install.packages('brms')
library(brms)
install.packages('loo')
library(loo)

fit_e <- brm(count ~ log_Age_c + log_Base4_c * Trt_c + (1|patient) + (1|visit),
data = epilepsy, family = "poisson",
prior = c(set_prior("uniform(-100,100)", class = "b"),
set_prior("cauchy(0,05)", class = "sd")))

summary(fit_e)
plot(fit_e)
VarCorr(fit_e)
ranef(fit_e)

players <- group_by(Batting, playerID)
games <- summarise(players, total = sum(G))
head(arrange(games, desc(total)), 5)

```

```{r RT}
#data <- read.delim2('F:/promec/Davi/Christina/Elite/allPeptides.txt',fill=TRUE, sep = '\t', header = T)
summary(data)
plot(data$Intensity,data$Raw.file)
plot(data$Intensity)
data1000=data[sample(nrow(data),size=1000,replace=FALSE),]
hist(log2(as.numeric(as.character(data1000$Intensity))))
typeof(data$Intensity)
ggplot(data1000, aes(x=(data1000$Intensity), fill=data1000$Raw.file)) + geom_histogram(alpha=0.2, position="identity")
data1000[1,1]
install.packages('Lahman')
install.packages('dplyr')
library(Lahman)
library(dplyr)

players <- group_by(Batting, playerID)
games <- summarise(players, total = sum(G))
head(arrange(games, desc(total)), 5)

```
```{r PCL}
#library('ggplot2')
#install.packages('digest')
#install.packages('reshape2')
#library('reshape2')
#library('digest')
data <- read.table('F:/promec/Qexactive/LARS/2014/desember/Rami_Morteza/Results/LFQ-pellet-supernatant-RPKM-genome-plasmidPlot.txt', sep = '\t', header = T)
summary(data)
dataname <- reade.table('G:/felles/PROTEOMICS and XRAY/Articles in prep/sPCL/Data after imputation/Top10upNdn.txt', sep = '\t', header = T)
dataname
data <- read.table('G:/felles/PROTEOMICS and XRAY/Articles in prep/sPCL/Data after imputation/Copy of Copy of SSandLFQttestPB All the collected super-SILAC and LFQ after imputation .txt', sep = '\t', header = T)
data[1,]
Significance=data$PCL2MMpvalue<0.05&abs(data$PCL2MMratio)>0.58&abs(data$PCL2MMratio)<10
sum(Significance)
dsub <- subset(data, Significance==TRUE)
#dsub$Protei
plot(data$medSSlog2diff,-log10(data$medSSlog2pvalue))
g = ggplot(data,aes(medSSlog2diff,-log10(medSSlog2pvalue)))
gps<-g + geom_point(aes(color=Significance)) + theme_bw(base_size=10) + xlab("Log2 Fold Change")  + ylab("-Log10 P-value") + ggtitle("Differentially expressed proteins") + scale_size_area()  + geom_text(data=dataname,aes(label=Gene.names),hjust=0.5, vjust=-0.5, size=3) 
print(gps)
#p=paste(file,proc.time()[3],".jpg")
#ggsave(filename=p, plot=gps)
```

```{r data}
data <- read.table('L:/Elite/LARS/2015/january/Ishita/uniprot-gomap-expression.txt', sep = '\t', header = T)
source("http://bioconductor.org/biocLite.R")
biocLite("STRINGdb")
library(STRINGdb)
install.packages('plyr')
string_db <- STRINGdb$new(version="10",species=3702)
string_db$get_homologs(c("9606.ENSP00000365757", "9606.ENSP00000352336"), target_species_id=10090, bitscore_threshold=60)
string_db$get_homologs(c("9606.ENSP00000365757", "9606.ENSP00000352336"), target_species_id=10090)
#
#

```



```{r GOenrich}
data$Gene.ontology..biological.process.
x=matrix(c(11,101,146,3024),nrow=2,ncol=2)
#http://www.wolframalpha.com/input/?i=%28112%21*3170%21*3125%21*157%21%29%2F%2811%21*146%21*3024%21*3282%21*101%21%29
x=matrix(c(5,1,0,4),nrow=2,ncol=2)
fisher.test(x)
#http://mathworld.wolfram.com/FishersExactTest.html
plot(x)
?fisher.test
```



```{r limma}
source("http://bioconductor.org/biocLite.R")
biocLite("limma")
library('limma')
install.packages('limma')
#https://support.bioconductor.org/p/46747/
class <- factor(c("mine","mine","mine","mine","mine","mine","wt","wt","wt","wt","wt","wt"))
desmat <- model.matrix(~0 + class)
colnames(desmat) <- class
colnames(desmat) <- gsub("class","", colnames(desmat))
fit <- lmFit(data[,8:19],desmat)
contrast.matrix<-makeContrasts(mine-wt, levels=desmat)
fit2<-contrasts.fit(fit,contrast.matrix)
fit2<-eBayes(fit2)
sel.diif<-p.adjust(fit2$F.p.value,method="fdr")<0.05
deg<-data[,8:19][sel.diif,]
topTable(fit2, coef=1, adjust="BH")
plot(fit2$F.p.value,data$pvalueadjusted)

#correction from Arnar
biocLite('impute')
library('impute')
des <- model.matrix(~1 + class)
X <- impute.knn(as.matrix(data[,8:19]))$data
fit2 <- eBayes(lmFit(X, design=des))
tab <- topTable(fit2, coef=2, genelist=data[,1:3], n=Inf, sort.by='none')
plot(data$pvalueadjusted, tab$adj.P.Val)

```


```{r brassica}
#library('ggplot2')
data[1,]
Significance=data$t.test.p.value<0.05&abs(data$t.test.Difference)>1
sum(Significance)
dsub <- subset(data, data$t.test.p.value==max(data$t.test.p.value)|data$t.test.Difference==max(abs(data$t.test.Difference)))
#dsub$Protei
plot(data$t.test.Difference,-log10(data$t.test.p.value))
g = ggplot(data,aes(t.test.Difference,-log10(t.test.p.value)))
gps<-g + geom_point(aes(color=Significance)) + theme_bw(base_size=10) + geom_text(data=dsub,aes(label=Protei),hjust=0, vjust=0) + xlab("Log2 Fold Change")  + ylab("-Log10 P-value") + ggtitle("Differentially expressed proteins") + scale_size_area()
print(gps)
#p=paste(file,proc.time()[3],".jpg")
#ggsave(filename=p, plot=gps)
```


```{r http://stackoverflow.com/questions/17924828/differences-in-heatmap-clustering-defaults-in-r-heatplot-versus-heatmap-2}
library(made4)
data(khan)
data <- as.matrix(khan$train[1:30,])
```

```{r http://sebastianraschka.com/Articles/heatmaps_in_r.html}
data <- read.csv('http://sebastianraschka.com/data/datasets/heatmaps_in_r.csv',header=T)
heatmap(as.numeric(unlist(data[3:12,2:4])))
```

```{r http://stackoverflow.com/questions/8406394/how-to-produce-a-heatmap-with-ggplot2}
library(reshape)
library(ggplot2)
library(scales)
library(plyr)
getwd()
dir()
data <- read.csv("http://dl.dropbox.com/u/2505196/fruit.txt", head=TRUE, sep=",")
 data.m = melt(data)
 data.m <- ddply(data.m, .(variable), transform, rescale = rescale(value))
 p <- ggplot(data.m, aes(variable, people)) + geom_tile(aes(fill = rescale), 
                                                   colour =   "white") 
 p + scale_fill_gradient(low = "white", high = "steelblue")
```
 

```{r https://gist.github.com/dsparks/3710171}
dataa <- read.table(file, sep = '\t', header = T)
zp1 <- ggplot(dataa,
              aes(x = X.Log.P.value., y = Difference, fill = Difference))
zp1 <- zp1 + geom_tile()
zp1 <- zp1 + scale_fill_gradientn(colours = myPalette(100))
zp1 <- zp1 + scale_x_discrete(expand = c(0, 0))
zp1 <- zp1 + scale_y_discrete(expand = c(0, 0))
zp1 <- zp1 + coord_equal()
zp1 <- zp1 + theme_bw()
print(zp1)

```


```{r plot}
#library('ggplot2')
data[1,]
Significance=data$X.Log.P.value.>-log10(0.05)&abs(data$Difference)>log2(1.5)
sum(Significance)
dsub <- subset(data, data$X.Log.P.value.==max(data$X.Log.P.value.)|data$Difference==max(abs(data$Difference)))
dsub$Gene.names
plot(data$Difference,data$X.Log.P.value.)
qplot(Difference,X.Log.P.value.,data=data,color=X.Log.P.value.>-log10(0.05)&abs(Difference)>log2(1.5))
g = ggplot(data,aes(Difference,X.Log.P.value.))
gps<-g + geom_point(aes(color=Significance)) + theme_bw(base_size=10) + geom_text(data=dsub,aes(label=Gene.names),hjust=0, vjust=0) + xlab("Log2 Fold Change")  + ylab("-Log10 P-value") + ggtitle("Differentially expressed proteins") + scale_size_area()
print(gps)
#p=paste(file,proc.time()[3],".jpg")
#ggsave(filename=p, plot=gps)
```

```{r data}
#heatmap(data$Difference)
p <- ggplot(data,aes(Difference,X.Log.P.value.))
p<-p + geom_tile(aes(fill=X.Log.P.value.)) + scale_fill_gradient(low="white", high="darkblue") + xlab("") + ylab("")
f=paste(file,proc.time()[3],".jpg")
ggsave(filename=f, plot=p)
print(p)
```


